
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_amoebae_select_positions &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for module_amoebae_select_positions</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># Copyright 2018 Lael D. Barlow</span>
<span class="c1"># </span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1"># </span>
<span class="c1">#     http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1"># </span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="c1"># </span>

<span class="sd">&quot;&quot;&quot;Contains functions for modifying alignment position/column selection for a</span>
<span class="sd">given phylogenetic tree dataset.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">settings</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">statistics</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="k">import</span> <span class="n">shuffle</span>

<span class="kn">from</span> <span class="nn">module_amoebae_trim_nex</span> <span class="k">import</span> <span class="n">trim_nex</span>
<span class="kn">from</span> <span class="nn">module_paralogue_counter</span> <span class="k">import</span> <span class="n">get_seq_obj_from_srch_res_csv_info</span>
<span class="kn">import</span> <span class="nn">module_amoebae_column_header_lists</span>
<span class="kn">from</span> <span class="nn">module_amoebae_phylo_clas</span> <span class="k">import</span> <span class="n">ModelInfoFromCSV</span><span class="p">,</span>\
<span class="n">get_clade_name_from_model</span><span class="p">,</span> <span class="n">code_names_in_ali</span><span class="p">,</span> <span class="n">quote_tree</span><span class="p">,</span> <span class="n">code_tree</span><span class="p">,</span>\
<span class="n">uncode_tree</span><span class="p">,</span> <span class="n">uncode_tree_obj</span>
<span class="kn">from</span> <span class="nn">module_paralogue_counter</span> <span class="k">import</span> <span class="n">add_seq_to_alignment3</span>

<span class="c1"># Import functions for working with sequences, alignments, and trees.</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span><span class="p">,</span> <span class="n">Gapped</span>
<span class="kn">from</span> <span class="nn">module_afa_to_nex</span> <span class="k">import</span> <span class="n">delete_extra_mesquite_lines</span><span class="p">,</span> <span class="n">afa_to_nex</span><span class="p">,</span> <span class="n">nex_to_afa</span><span class="p">,</span> <span class="n">nex_to_phylip</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="k">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">ete3</span> <span class="k">import</span> <span class="n">Tree</span><span class="p">,</span> <span class="n">faces</span><span class="p">,</span> <span class="n">AttrFace</span><span class="p">,</span> <span class="n">TreeStyle</span><span class="p">,</span> <span class="n">NodeStyle</span><span class="p">,</span> <span class="n">TextFace</span>

<span class="kn">from</span> <span class="nn">module_amoebae_select_seqs</span> <span class="k">import</span> <span class="n">get_clade_name_from_model2</span><span class="p">,</span>\
<span class="n">get_nodes_of_interest</span><span class="p">,</span> <span class="n">get_list_of_leaf_names_for_node</span><span class="p">,</span> <span class="n">TaxonomicInfo</span><span class="p">,</span>\
<span class="n">get_taxonomic_info</span><span class="p">,</span> <span class="n">define_nodestyles_dict_for_colourcoding</span><span class="p">,</span>\
<span class="n">define_textface_for_labeling_stem</span><span class="p">,</span> <span class="n">get_corresponding_node</span><span class="p">,</span>\
<span class="n">get_ml_tree_branch_lengths</span><span class="p">,</span> <span class="n">get_branch_length_info</span><span class="p">,</span> <span class="n">reduce_alignment</span>


<div class="viewcode-block" id="trim_one_column_from_alignment"><a class="viewcode-back" href="../module_amoebae_select_positions.html#module_amoebae_select_positions.trim_one_column_from_alignment">[docs]</a><span class="k">def</span> <span class="nf">trim_one_column_from_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">column_index</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a nexus alignment file path and remove column at a given index</span>
<span class="sd">    (first column has index 0).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">alignment2</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="c1"># Parse input alignment file.</span>
        <span class="n">alignment_obj</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">infh</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
        <span class="c1"># Record input alignment length.</span>
        <span class="n">len1</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_obj</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="c1"># Remove relevant column from alignment object.</span>
        <span class="n">alignment_obj2</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="p">[:,</span> <span class="p">:</span><span class="n">column_index</span><span class="p">]</span> <span class="o">+</span> <span class="n">alignment_obj</span><span class="p">[:,</span> <span class="n">column_index</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:]</span>
        <span class="c1"># Check that the alignment length is shorter by one column.</span>
        <span class="n">len2</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alignment_obj2</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">len1</span> <span class="o">-</span> <span class="n">len2</span> <span class="o">==</span> <span class="mi">1</span>
        <span class="c1"># Write modified alignment object to output file path.</span>
        <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignment_obj2</span><span class="p">,</span> <span class="n">o</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span></div>



<div class="viewcode-block" id="check_whether_all_cols_unique_seqs"><a class="viewcode-back" href="../module_amoebae_select_positions.html#module_amoebae_select_positions.check_whether_all_cols_unique_seqs">[docs]</a><span class="k">def</span> <span class="nf">check_whether_all_cols_unique_seqs</span><span class="p">(</span><span class="n">alignment_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a path to an alignment in nexus format, and return True if all the</span>
<span class="sd">    columns in the alignment have unique sequences, and False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Parse input alignment file and compile a list of column sequences.</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">alignment_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">:</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">infh</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>

        <span class="c1"># Get length of sequences in alignment.</span>
        <span class="n">seq_len</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()</span>

        <span class="c1"># get a list of columns as strings in the original alignment.</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">alignment</span><span class="p">[:,</span> <span class="n">col</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)]</span> 

    <span class="k">assert</span> <span class="n">columns</span> <span class="o">!=</span> <span class="p">[]</span>
        
    <span class="c1"># Determine whether there are redundant column sequences.</span>
    <span class="n">x</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">columns</span><span class="p">)))</span>
    <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="recursively_remove_positions"><a class="viewcode-back" href="../module_amoebae_select_positions.html#module_amoebae_select_positions.recursively_remove_positions">[docs]</a><span class="k">def</span> <span class="nf">recursively_remove_positions</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">,</span>
                                 <span class="n">t1</span><span class="p">,</span>
                                 <span class="n">alignment</span><span class="p">,</span>
                                 <span class="n">subs_model</span><span class="p">,</span>
                                 <span class="n">type_seqs_dict</span><span class="p">,</span>
                                 <span class="n">cols_attempted_to_remove</span><span class="p">,</span>
                                 <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                 <span class="n">log_file_path</span><span class="p">,</span>
                                 <span class="n">optimize_internal_branches</span>
                                 <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a tree, and try to remove positions/columns from the alignment to</span>
<span class="sd">    improve the ratio of the stem lengths of clades to the branch lengths of</span>
<span class="sd">    the branches in the clades, and return the final tree when as many postions</span>
<span class="sd">    as possible have been removed or no sequence can be removed that will</span>
<span class="sd">    result in improving the branch length ratios.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define variable for whether any new tree could be produced.</span>
    <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variable for whether any new tree is better or not.</span>
    <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variables to store info for next recursion, if necessary.</span>
    <span class="n">new_tree</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_alignment</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_subs_model</span> <span class="o">=</span> <span class="n">subs_model</span>
    <span class="n">new_type_seqs_dict</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Define variables for writing to log file.</span>
    <span class="n">action</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="s1">&#39;No output tree&#39;</span>

    <span class="c1"># Determine average of stem length to branch length ratio over clades of</span>
    <span class="c1"># interest.</span>
    <span class="c1">#branch_length_info_dict = get_branch_length_info(t1, type_seqs_dict)</span>
    <span class="n">average_over_clades</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

    <span class="c1"># Determine the clade with the worst ratio.</span>
    <span class="n">clades_by_ascending_ratio</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>\
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">])</span>
    <span class="n">clade_with_lowest_ratio</span> <span class="o">=</span> <span class="n">clades_by_ascending_ratio</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Identify the clade for which to check for improved branch length ratios</span>
    <span class="c1"># for (an alternative would be to iterate over all of them)***</span>
    <span class="n">clade</span> <span class="o">=</span> <span class="n">clade_with_lowest_ratio</span>


    <span class="c1">####################### Key part that is different from the function for</span>
    <span class="c1"># removing sequences instead of positions ########################</span>

    <span class="c1"># Parse input alignment, and find a column to try to remove.</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">:</span>
        <span class="n">alignment_obj</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">infh</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>

        <span class="c1"># Get length of sequences in alignment.</span>
        <span class="n">seq_len</span> <span class="o">=</span> <span class="n">alignment_obj</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()</span>

        <span class="c1"># get a list of columns as strings in the original alignment.</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">alignment_obj</span><span class="p">[:,</span> <span class="n">col</span><span class="p">]</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)]</span> 

        <span class="c1"># Iterate over column sequences.</span>
        <span class="n">column_index</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
        <span class="k">for</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">columns</span><span class="p">:</span>
            <span class="n">column_index</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">column</span> <span class="ow">in</span> <span class="n">cols_attempted_to_remove</span><span class="p">:</span>
                <span class="c1"># Try removing the column from the tree, and see whether it</span>
                <span class="c1"># improves the branch length ratios.</span>

                <span class="c1">## If the sequence is a type sequence, then you have to</span>
                <span class="c1">## choose another sequence in the clade as a replacement</span>
                <span class="c1">## type sequence.</span>
                <span class="c1">#new_type_seqs_dict = type_seqs_dict.copy()</span>
                <span class="c1">#is_type_seq = False</span>
                <span class="c1">#for cladex in new_type_seqs_dict.keys():</span>
                <span class="c1">#    if seqname == new_type_seqs_dict[cladex]:</span>
                <span class="c1">#        is_type_seq = True</span>
                <span class="c1">#        assert clade == cladex</span>
                <span class="c1">#if is_type_seq:</span>
                <span class="c1">#    # Identify an alternative type sequence for the same</span>
                <span class="c1">#    # (current) clade.</span>
                <span class="c1">#    for seqname2 in\</span>
                <span class="c1">#    reversed(branch_length_info_dict[clade]\</span>
                <span class="c1">#    [&#39;seq names by descending length&#39;].copy()):</span>
                <span class="c1">#        if seqname2 != seqname:</span>
                <span class="c1">#            # Change type seq for clade to a different sequence.</span>
                <span class="c1">#            new_type_seqs_dict[clade] = seqname2</span>

                <span class="n">new_type_seqs_dict</span> <span class="o">=</span> <span class="n">type_seqs_dict</span>


                <span class="c1"># Define new alignment path.</span>
                <span class="n">alignment2</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span>\
                <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.nex&#39;</span>

                <span class="c1"># Remove column from alignment.</span>
                <span class="n">trim_one_column_from_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span>
                        <span class="n">column_index</span><span class="p">)</span>
                <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">alignment2</span>

                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">removed column from alignment&#39;</span><span class="p">)</span>

                <span class="c1"># Code alignment (get table file).</span>

                <span class="c1">## Remove seq from a copy of the parsed tree object, and</span>
                <span class="c1">## use that as a constraint tree.</span>
                <span class="c1">## Copy tree.</span>
                <span class="n">t2</span> <span class="o">=</span> <span class="n">t1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                <span class="c1">## Get a list of all node objects with the same name as the</span>
                <span class="c1">## sequence of interest.</span>
                <span class="c1">#seq_nodes = t2.search_nodes(name=seqname)</span>
                <span class="c1">## Check that only one such node was found.</span>
                <span class="c1">#assert len(seq_nodes) == 1</span>
                <span class="c1">## Remove the node from the new tree object.</span>
                <span class="c1">#node_to_remove = seq_nodes[0]</span>
                <span class="c1">#t2_len_before = len(t2.get_leaves())</span>
                <span class="c1">#node_to_remove.delete()</span>

                <span class="c1">## Check that a node was actually removed.</span>
                <span class="c1">#t2_len_after = len(t2.get_leaves())</span>
                <span class="c1">#assert t2_len_before == t2_len_after + 1</span>

                <span class="c1"># Write modified tree to a file.</span>
                <span class="n">tfp</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_newick.tre&#39;</span><span class="p">)</span>
                <span class="n">t2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outfile</span><span class="o">=</span><span class="n">tfp</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>

                <span class="c1"># Run ML tree with IQtree, and get branch length info</span>
                <span class="c1"># dictionary.</span>
                <span class="c1"># Define output directory for files.</span>
                <span class="n">outputdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
                <span class="c1"># Call function to run IQtree and return a parsed tree obj.</span>
                <span class="n">t3</span> <span class="o">=</span> <span class="n">get_ml_tree_branch_lengths</span><span class="p">(</span><span class="n">tfp</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">outputdir</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">ran IQtree&#39;</span><span class="p">)</span>
                <span class="c1"># Set t3 as the new tree.</span>
                <span class="n">new_tree</span> <span class="o">=</span> <span class="n">t3</span>
                <span class="c1"># Define output path for annotated tree image file.</span>
                <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
                <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="n">annotated_tree_outpath</span>
                <span class="c1"># Extract info about branch lengths from tree, and write</span>
                <span class="c1"># annotated tree image file.</span>
                <span class="n">new_branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t3</span><span class="p">,</span>
                                          <span class="n">new_type_seqs_dict</span><span class="p">,</span>
                                          <span class="n">annotated_tree_outpath</span><span class="p">)</span>
                <span class="c1"># Calculate average ratio over all clades from branch</span>
                <span class="c1"># length info. </span>
                <span class="n">new_average_over_clades</span> <span class="o">=</span>\
                        <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
                        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">average ratio over clades: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span>\
                        <span class="nb">str</span><span class="p">(</span><span class="n">new_average_over_clades</span><span class="p">))</span>

                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">parsed IQtree output.&#39;</span><span class="p">)</span>

                <span class="c1"># Write a file with type sequences and clade names for this</span>
                <span class="c1"># iteration.</span>
                <span class="n">new_type_seqs_file</span><span class="o">=</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_type_seqs.csv&#39;</span><span class="p">)</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">new_type_seqs_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>


                <span class="c1"># Should make a separate function for comparing current and</span>
                <span class="c1"># previous trees.</span>

                <span class="c1"># Compare lengths of internal branches.</span>
                <span class="n">previous_internal_branch_lengths</span> <span class="o">=</span>\
                <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
                <span class="n">new_internal_branch_lengths</span> <span class="o">=</span>\
                <span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
                <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">,</span> <span class="n">new_internal_branch_lengths</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                        <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                        <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="nb">min</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                            <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                            <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="c1"># Get sums of previous and new internal branch lengths.</span>
                <span class="n">previous_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
                <span class="nb">sum</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">)</span>
                <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
                <span class="nb">sum</span><span class="p">(</span><span class="n">new_internal_branch_lengths</span><span class="p">)</span>

                <span class="c1"># Indicate that a new tree was produced.</span>
                <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="c1"># Compare the branch length ratios of the new tree to</span>
                <span class="c1"># the current tree, and apply criteria to decide whether</span>
                <span class="c1"># the new tree is better or not.</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">new_average_over_clades</span> <span class="o">&gt;</span> <span class="n">average_over_clades</span><span class="p">:</span>
                        <span class="c1"># Make sure that it does not make the branch length ratio</span>
                        <span class="c1"># worse for any clade.</span>
                        <span class="c1">#made_a_clade_worse = False</span>
                        <span class="c1">#for cladex in branch_length_info_dict.keys():</span>
                        <span class="c1">#    old_clade_ratio = branch_length_info_dict\</span>
                        <span class="c1">#                      [cladex][&#39;stem/branch ratio&#39;]</span>
                        <span class="c1">#    new_clade_ratio = new_branch_length_info_dict\</span>
                        <span class="c1">#                      [cladex][&#39;stem/branch ratio&#39;]</span>
                        <span class="c1">#    if new_clade_ratio &lt; old_clade_ratio:</span>
                        <span class="c1">#        made_a_clade_worse = True</span>
                        <span class="c1">#        break</span>
                        <span class="c1">#if not made_a_clade_worse:</span>
                        <span class="c1">#    print(&#39;\tnew tree is better&#39;)</span>
                        <span class="c1">#    new_tree_better = True</span>

                        <span class="c1"># Make sure that it did not make the ratio worse for</span>
                        <span class="c1"># the clade of interest (current worst clade).</span>
                        <span class="n">old_clade_ratio</span> <span class="o">=</span> <span class="n">branch_length_info_dict</span>\
                                          <span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                        <span class="n">new_clade_ratio</span> <span class="o">=</span> <span class="n">new_branch_length_info_dict</span>\
                                          <span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">new_clade_ratio</span> <span class="o">&gt;=</span> <span class="n">old_clade_ratio</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                            <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>
                        
                        <span class="k">else</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is not better&#39;</span><span class="p">)</span>

                <span class="c1"># Apply different criteria if the</span>
                <span class="c1"># optimize_internal_branches option is selected.</span>
                <span class="k">elif</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                    <span class="c1">#if no_new_internal_branches_are_worse:</span>
                    <span class="c1">#    if at_least_one_internal_branch_longer:</span>
                    <span class="c1">#        print(&#39;\tnew tree is better&#39;)</span>
                    <span class="c1">#        # Temp.</span>
                    <span class="c1">#        assert 2 != 2</span>
                    <span class="c1">#        new_tree_better = True</span>
                    <span class="k">if</span> <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">&gt;</span>\
                    <span class="n">previous_sum_of_internal_branch_lengths</span><span class="p">:</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">shortest_internal_branch_shorter</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                            <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="c1">#if shortest_internal_branch_longer:</span>
                    <span class="c1">#    assert 2 != 2</span>
                    <span class="c1">#if shortest_internal_branch_shorter:</span>
                    <span class="c1">#    assert 2 != 2</span>



                <span class="c1"># Add column sequence to list of column sequences which have</span>
                <span class="c1"># been considered for removal.</span>
                <span class="n">cols_attempted_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">column</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># See if the next longest branch can be removed.</span>
                <span class="k">pass</span>
            
            <span class="c1"># Break the loop so that the function can be run again.</span>
            <span class="k">if</span> <span class="n">new_tree_produced</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="c1">## Break the loop so that the function can be run again.</span>
        <span class="c1">#if new_tree_produced:</span>
        <span class="c1">#    break</span>

    <span class="c1"># If no new tree could be produced, then return the final tree.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">new_tree_produced</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a column&#39;</span><span class="p">,</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">recursion_num</span><span class="p">,</span> <span class="n">t1</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">,</span>\
                <span class="n">cols_attempted_to_remove</span><span class="p">,</span> <span class="n">branch_length_info_dict</span><span class="p">]</span>

    <span class="c1"># Otherwise run the function again.</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">new_tree_better</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a column&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on new tree, alignment ,and type sequences.</span>
            <span class="k">return</span> <span class="n">recursively_remove_positions</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">new_tree</span><span class="p">,</span>
                                           <span class="n">new_alignment</span><span class="p">,</span>
                                           <span class="n">new_subs_model</span><span class="p">,</span>
                                           <span class="n">new_type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">cols_attempted_to_remove</span><span class="p">,</span>
                                           <span class="n">new_branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">True</span>
                                           <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a column&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on same input (except with longer list of seqs that</span>
            <span class="c1"># were considered for removal).</span>
            <span class="k">return</span> <span class="n">recursively_remove_positions</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">t1</span><span class="p">,</span>
                                           <span class="n">alignment</span><span class="p">,</span>
                                           <span class="n">subs_model</span><span class="p">,</span>
                                           <span class="n">type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">cols_attempted_to_remove</span><span class="p">,</span>
                                           <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">True</span>
                                           <span class="p">)</span></div>


<div class="viewcode-block" id="optimize_position_selection"><a class="viewcode-back" href="../module_amoebae_select_positions.html#module_amoebae_select_positions.optimize_position_selection">[docs]</a><span class="k">def</span> <span class="nf">optimize_position_selection</span><span class="p">(</span><span class="n">model_name</span><span class="p">,</span> <span class="n">outdirpath</span><span class="p">,</span> <span class="n">timestamp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Systematically remove position/columns from an alignment to optimize for</span>
<span class="sd">    stem to median branch length ratios for clades of interest as a measure of</span>
<span class="sd">    resolution.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Increase recursion limit for python. Potentially dangerous, but necessary for</span>
    <span class="c1"># large alignments, like for Giantin and lava lamp proteins!</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">setrecursionlimit</span><span class="p">(</span><span class="mi">10000</span><span class="p">)</span>

    <span class="c1"># Parse data from model.</span>
    <span class="c1"># Get info about model that is relevant.</span>
    <span class="n">model_info</span> <span class="o">=</span> <span class="n">ModelInfoFromCSV</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">alignment_file</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">tree_topology_file</span>
    <span class="n">subs_model</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">subs_model</span>
    <span class="n">type_seqs</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">type_seqs_file</span>

    <span class="c1"># Parse tree using ete3.</span>
    <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Print tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Input tree:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>

    <span class="c1"># Define output directory path.</span>
    <span class="n">outputdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdirpath</span><span class="p">,</span>\
            <span class="n">model_name</span> <span class="o">+</span> <span class="s1">&#39;_select_positions_output_&#39;</span> <span class="o">+</span> <span class="n">timestamp</span><span class="p">)</span>
    <span class="c1"># Make output directory.</span>
    <span class="k">assert</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">outputdir</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">outputdir</span><span class="p">)</span>

    <span class="c1"># Construct a dictionary with info from type_seqs_file.</span>
    <span class="n">type_seqs_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">type_seqs</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">infh</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
                <span class="n">spliti</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
                <span class="n">seqname</span> <span class="o">=</span> <span class="n">spliti</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">cladename</span> <span class="o">=</span> <span class="n">spliti</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">cladename</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqname</span>

    <span class="c1"># Copy alignment file to new path.</span>
    <span class="n">new_ali_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_0.nex&#39;</span><span class="p">)</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">new_ali_path</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">new_ali_path</span><span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">new_ali_path</span> 

    <span class="c1"># Run ML search on model tree to get branch lengths.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">get_ml_tree_branch_lengths</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">outputdir</span><span class="p">)</span>

    <span class="c1"># Define output path for annotated tree image file.</span>
    <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
    <span class="c1"># Extract info about branch lengths from tree, and write</span>
    <span class="c1"># annotated tree image file.</span>
    <span class="n">branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t2</span><span class="p">,</span>
                              <span class="n">type_seqs_dict</span><span class="p">,</span>
                              <span class="n">annotated_tree_outpath</span><span class="p">)</span>

    <span class="c1">## Compile list of sequences represented in the input initial tree.</span>
    <span class="c1">#names_of_seqs_in_tree = [x.name for x in t2.get_leaves()]</span>

    <span class="c1">## Parse input sequence file, and construct a list of sequence objects for</span>
    <span class="c1">## potential inclusion in the tree.</span>
    <span class="c1">#seqs = []</span>
    <span class="c1">#for seq in SeqIO.parse(file_with_seqs, &#39;fasta&#39;):</span>
    <span class="c1">#    seqs.append(seq)</span>

    <span class="c1"># Define path to log file (specifically for select_positions).</span>
    <span class="n">log_file_path</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_select_positions_log.txt&#39;</span>
    <span class="n">header_line</span> <span class="o">=</span> <span class="s2">&quot;Step number, Action, Output file</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header_line</span><span class="p">)</span>
        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;Initial version&#39;</span><span class="p">,</span>\
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annotated_tree_outpath</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># Check that all the columns in the input alignment have unique sequences</span>
    <span class="c1"># (so that they can be identified according to their sequence rather than</span>
    <span class="c1"># their order in the alignment).</span>
    <span class="n">all_columns_have_unique_seqs</span> <span class="o">=</span>\
    <span class="n">check_whether_all_cols_unique_seqs</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">all_columns_have_unique_seqs</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Some columns in the input alignment</span>
<span class="s2">    have identical sequences.&quot;&quot;&quot;</span>

    <span class="c1">########</span>
    <span class="c1">## Load information into a recursive or iterative function/loop to remove</span>
    <span class="c1">## sequences.</span>
    <span class="n">cols_attempted_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">best_tree_info</span> <span class="o">=</span> <span class="n">recursively_remove_positions</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span>
                                                  <span class="n">t2</span><span class="p">,</span>
                                                  <span class="n">alignment</span><span class="p">,</span>
                                                  <span class="n">subs_model</span><span class="p">,</span>
                                                  <span class="n">type_seqs_dict</span><span class="p">,</span>
                                                  <span class="n">cols_attempted_to_remove</span><span class="p">,</span>
                                                  <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                                  <span class="n">log_file_path</span><span class="p">,</span>
                                                  <span class="kc">False</span>
                                                  <span class="p">)</span>
    <span class="c1">##(returns: [recursion_num, t1, alignment, subs_model, type_seqs_dict,\</span>
    <span class="c1">##        seqs_attempted_to_remove, branch_length_info_dict] )</span>

    <span class="k">assert</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">2</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Remember to make a mask of the input alignment based on</span>
<span class="s2">    the columns that were selected for removal.&quot;&quot;&quot;</span>
    <span class="c1"># ...</span>

    <span class="k">return</span> <span class="s1">&#39;[main output path here]&#39;</span></div>



</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.6.
    </div>
  </body>
</html>