
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_amoebae_select_seqs &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for module_amoebae_select_seqs</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># Copyright 2018 Lael D. Barlow</span>
<span class="c1"># </span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1"># </span>
<span class="c1">#     http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1"># </span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="c1"># </span>

<span class="sd">&quot;&quot;&quot;Contains functions for modifying sequence selection for a given phylogenetic</span>
<span class="sd">tree.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">settings</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">statistics</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="k">import</span> <span class="n">shuffle</span>

<span class="kn">from</span> <span class="nn">module_amoebae_trim_nex</span> <span class="k">import</span> <span class="n">trim_nex</span>
<span class="kn">from</span> <span class="nn">module_paralogue_counter</span> <span class="k">import</span> <span class="n">get_seq_obj_from_srch_res_csv_info</span>
<span class="kn">import</span> <span class="nn">module_amoebae_column_header_lists</span>
<span class="kn">from</span> <span class="nn">module_amoebae_phylo_clas</span> <span class="k">import</span> <span class="n">ModelInfoFromCSV</span><span class="p">,</span>\
<span class="n">get_clade_name_from_model</span><span class="p">,</span> <span class="n">code_names_in_ali</span><span class="p">,</span> <span class="n">quote_tree</span><span class="p">,</span> <span class="n">code_tree</span><span class="p">,</span>\
<span class="n">uncode_tree</span><span class="p">,</span> <span class="n">uncode_tree_obj</span>
<span class="kn">from</span> <span class="nn">module_paralogue_counter</span> <span class="k">import</span> <span class="n">add_seq_to_alignment3</span>

<span class="c1"># Import functions for working with sequences, alignments, and trees.</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span><span class="p">,</span> <span class="n">Gapped</span>
<span class="kn">from</span> <span class="nn">module_afa_to_nex</span> <span class="k">import</span> <span class="n">delete_extra_mesquite_lines</span><span class="p">,</span> <span class="n">afa_to_nex</span><span class="p">,</span> <span class="n">nex_to_afa</span><span class="p">,</span> <span class="n">nex_to_phylip</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="k">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">ete3</span> <span class="k">import</span> <span class="n">Tree</span><span class="p">,</span> <span class="n">faces</span><span class="p">,</span> <span class="n">AttrFace</span><span class="p">,</span> <span class="n">TreeStyle</span><span class="p">,</span> <span class="n">NodeStyle</span><span class="p">,</span> <span class="n">TextFace</span>


<div class="viewcode-block" id="get_clade_name_from_model2"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_clade_name_from_model2">[docs]</a><span class="k">def</span> <span class="nf">get_clade_name_from_model2</span><span class="p">(</span><span class="n">type_seq_name</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return clade name for given type sequence name and type seqs info file</span>
<span class="sd">    (from models directory).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get clade name.</span>
    <span class="n">clade_name</span> <span class="o">=</span> <span class="s1">&#39;?&#39;</span>
    <span class="k">for</span> <span class="n">clade</span> <span class="ow">in</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">type_seq_name</span> <span class="o">==</span> <span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">]:</span>
                <span class="n">clade_name</span> <span class="o">=</span> <span class="n">clade</span>
                <span class="k">break</span>
        <span class="k">elif</span> <span class="n">type_seq_name</span> <span class="o">==</span> <span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">):</span>
            <span class="n">clade_name</span> <span class="o">=</span> <span class="n">clade</span>
            <span class="k">break</span>
    <span class="k">assert</span> <span class="n">clade_name</span> <span class="o">!=</span> <span class="s1">&#39;?&#39;</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Could not identify clade name for type</span>
<span class="s2">    sequence name </span><span class="si">%s</span><span class="s2">.&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">type_seq_name</span>

    <span class="k">return</span> <span class="n">clade_name</span></div>


<div class="viewcode-block" id="get_nodes_from_actual_tree_obj"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_nodes_from_actual_tree_obj">[docs]</a><span class="k">def</span> <span class="nf">get_nodes_from_actual_tree_obj</span><span class="p">(</span><span class="n">t3</span><span class="p">,</span> <span class="n">nodes_of_interest</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get nodes in the given tree object that correspond to the nodes in a</span>
<span class="sd">    given list of nodes from another tree object, and return as a list.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Get original root node for tree object.</span>
    <span class="n">original_outgroup</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">())[</span><span class="mi">1</span><span class="p">]</span>
    <span class="c1"># Compile a list of corresponding nodes.</span>
    <span class="n">new_node_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nodes_of_interest</span><span class="p">:</span>
        <span class="n">i_leaf_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()])</span>
        <span class="n">corresponding_node</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># Traverse the tree to find node.</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="n">node_leaf_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]</span>
            <span class="c1"># If the node has the same set of leaf names, then it is the</span>
            <span class="c1"># corresponding node.</span>
            <span class="k">if</span> <span class="n">node_leaf_names</span> <span class="o">==</span> <span class="n">i_leaf_names</span><span class="p">:</span>
                <span class="n">corresponding_node</span> <span class="o">=</span> <span class="n">node</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Generate a list of three different rootings for the tree, at</span>
            <span class="c1"># least one of which should allow identification of the clade of</span>
            <span class="c1"># interest.</span>
            <span class="n">roots</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">roots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">get_midpoint_outgroup</span><span class="p">())</span>
            <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">get_midpoint_outgroup</span><span class="p">())</span>
            <span class="n">side1</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">())[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">side2</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">())[</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">side</span> <span class="ow">in</span> <span class="p">[</span><span class="n">side1</span><span class="p">,</span> <span class="n">side2</span><span class="p">]:</span>
                <span class="n">side_longest_length</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">side_longest_leaf</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">side</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
                    <span class="n">length</span> <span class="o">=</span> <span class="n">side</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">leaf</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">length</span> <span class="o">&gt;</span> <span class="n">side_longest_length</span><span class="p">:</span>
                        <span class="n">side_longest_length</span> <span class="o">=</span> <span class="n">length</span>
                        <span class="n">side_longest_leaf</span> <span class="o">=</span> <span class="n">leaf</span>
                <span class="c1"># Check that a leaf was identified.</span>
                <span class="k">assert</span> <span class="n">side_longest_leaf</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="c1"># Add the longest leaf on this side of the midpoint to the list</span>
                <span class="c1"># of nodes on which to try rooting the tree.</span>
                <span class="n">roots</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">side_longest_leaf</span><span class="p">)</span>

            <span class="c1"># Try finding the corresponding node again.</span>
            <span class="k">for</span> <span class="n">root_node</span> <span class="ow">in</span> <span class="n">roots</span><span class="p">:</span>
                <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">root_node</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                    <span class="n">node_leaf_names</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()])</span>
                    <span class="c1"># If the node has the same set of leaf names, then it is the</span>
                    <span class="c1"># corresponding node.</span>
                    <span class="k">if</span> <span class="n">node_leaf_names</span> <span class="o">==</span> <span class="n">i_leaf_names</span><span class="p">:</span>
                        <span class="n">corresponding_node</span> <span class="o">=</span> <span class="n">node</span>
                        <span class="k">break</span>

            <span class="c1"># Check that the corresponding node was found.</span>
            <span class="k">assert</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

            <span class="c1"># Root on the corresponding node identified.</span>
            <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">corresponding_node</span><span class="p">)</span>

        <span class="c1"># Check that the corresponding node was found.</span>
        <span class="k">assert</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

        <span class="c1"># Append the corresponding node to the list.</span>
        <span class="n">new_node_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">corresponding_node</span><span class="p">)</span>

    <span class="c1"># Re-root the tree object on the original root node.</span>
    <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">original_outgroup</span><span class="p">)</span>

    <span class="c1"># Return the list of corresponding nodes.</span>
    <span class="k">return</span> <span class="n">new_node_list</span></div>




<div class="viewcode-block" id="get_nodes_of_interest"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_nodes_of_interest">[docs]</a><span class="k">def</span> <span class="nf">get_nodes_of_interest</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">type_seq_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take an ete3 tree (TreeNode object), a list of type sequence names, and</span>
<span class="sd">    return a list of nodes of interest defined by clades (as ete3 TreeNode</span>
<span class="sd">    objects).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initiate list of orthogroup nodes of interest.</span>
    <span class="n">orthogroup_nodes</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Make a copy of the input tree.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># For each &quot;type&quot; sequence, traverse all nodes and find the node with</span>
    <span class="c1"># the largest number of child nodes that are leaf (terminal) nodes,</span>
    <span class="c1"># containing the &quot;type&quot; sequence of interest, but not containing any of</span>
    <span class="c1"># the other &quot;type&quot; sequences.</span>
    <span class="n">ts_num</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">first_type_seq_node_name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">ts</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
        <span class="n">ts_num</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="n">ts_num</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">first_type_seq_node_name</span> <span class="o">=</span> <span class="n">ts</span>
            <span class="c1"># Root on another &quot;type&quot; sequence for the first type sequence in</span>
            <span class="c1"># the list to get whole clade, then root the tree on the ancestor</span>
            <span class="c1"># node of that first clade.</span>

            <span class="c1"># Get a node name for a node corresponding to a different</span>
            <span class="c1"># &quot;type&quot; sequence.</span>
            <span class="n">other_type_seq_node_name</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">ts</span><span class="p">:</span>
                    <span class="n">other_type_seq_node_name</span> <span class="o">=</span> <span class="n">i</span>
                    <span class="k">break</span>
            <span class="k">assert</span> <span class="n">other_type_seq_node_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

            <span class="c1"># Get node corresponding to a different &quot;type&quot; sequence.</span>
            <span class="n">other_type_seq_node</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t2</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other_type_seq_node_name</span><span class="p">:</span>
                    <span class="n">other_type_seq_node</span> <span class="o">=</span> <span class="n">node</span>
                    <span class="k">break</span>
            <span class="k">assert</span> <span class="n">other_type_seq_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Could not identify</span>
<span class="s2">            a type seq node with name: </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">other_type_seq_node_name</span>

            <span class="c1"># Root on the other &quot;type&quot; sequence node.</span>
            <span class="n">t2</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">other_type_seq_node</span><span class="p">)</span>

            <span class="c1">#print(&#39;\n\n\nTree rooted on a type sequence other than the first type sequence.&#39;)</span>
            <span class="c1">#print(t2)</span>

        <span class="k">elif</span> <span class="n">ts_num</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="c1"># Root on the first &quot;type&quot; sequence node for all subsequent</span>
            <span class="c1"># clades.</span>
            <span class="n">first_type_seq_node</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t2</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="n">leaf_list</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">first_type_seq_node_name</span><span class="p">:</span>
                        <span class="n">first_type_seq_node</span> <span class="o">=</span> <span class="n">node</span>
                        <span class="k">break</span>
            <span class="n">t2</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">first_type_seq_node</span><span class="p">)</span>
            <span class="c1">#print(&#39;\n\n\nTree re-rooted on first type sequence:&#39;)</span>
            <span class="c1">#print(t2)</span>


        <span class="c1"># Make a copy of the tree topology to work with for each run</span>
        <span class="c1"># through this loop.</span>
        <span class="n">t3</span> <span class="o">=</span> <span class="n">t2</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># Make a list of nodes that contain type seq, but not any others. #</span>
        <span class="c1"># SEPARATE</span>
        <span class="n">nodes_of_interest</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="c1"># Search in nodes that contain the type sequence.</span>
            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ts</span><span class="p">):</span>
                <span class="c1"># Search in nodes that don&#39;t contain other type sequences.</span>
                <span class="n">contains_other_type_seqs</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">ts2</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">ts2</span> <span class="o">==</span> <span class="n">ts</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ts2</span><span class="p">):</span>
                            <span class="n">contains_other_type_seqs</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">contains_other_type_seqs</span><span class="p">:</span>
                    <span class="c1"># Add nodes of interest to list.</span>
                    <span class="n">nodes_of_interest</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

        <span class="c1"># Check that a node of interest was found.</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">nodes_of_interest</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span>

        <span class="c1"># find the node with the most child leaf nodes.</span>
        <span class="n">node_w_most_leaves</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">nodes_of_interest</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>\
                <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1">#node_w_most_leaves.name = &#39;X&#39;</span>

        <span class="c1"># Add node to list of nodes of interest.</span>
        <span class="n">orthogroup_nodes</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">ts</span><span class="p">,</span> <span class="n">node_w_most_leaves</span><span class="p">])</span>

    <span class="c1"># Return list of lists containing defining sequence name and node object.</span>
    <span class="k">return</span> <span class="n">orthogroup_nodes</span></div>


<div class="viewcode-block" id="get_list_of_leaf_names_for_node"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_list_of_leaf_names_for_node">[docs]</a><span class="k">def</span> <span class="nf">get_list_of_leaf_names_for_node</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Given an ete3 TreeNode object, return a list of terminal/leaf node</span>
<span class="sd">    names.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">leaf_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">n</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
            <span class="n">leaf_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">leaf_names</span></div>


<div class="viewcode-block" id="TaxonomicInfo"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.TaxonomicInfo">[docs]</a><span class="k">class</span> <span class="nc">TaxonomicInfo</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Class for storing taxonomic info.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">superbranch</span><span class="p">,</span> <span class="n">supergroup</span><span class="p">,</span> <span class="n">group</span><span class="p">,</span> <span class="n">species</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">superbranch</span> <span class="o">=</span> <span class="n">superbranch</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">supergroup</span> <span class="o">=</span> <span class="n">supergroup</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">group</span><span class="o">=</span> <span class="n">group</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species</span> <span class="o">=</span> <span class="n">species</span></div>


<div class="viewcode-block" id="get_taxonomic_info"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_taxonomic_info">[docs]</a><span class="k">def</span> <span class="nf">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Given a species name that appears in the genome info csv file in the</span>
<span class="sd">    genomes folder, return taxonomic information from that csv file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define path to info file.</span>
    <span class="n">info_csv</span> <span class="o">=</span> <span class="n">settings</span><span class="o">.</span><span class="n">db_info_csv</span>

    <span class="c1"># Parse info file with pandas.</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">info_csv</span><span class="p">)</span>

    <span class="c1"># Iterate through rows until one is found with the same species name.</span>
    <span class="n">taxonomic_info_obj</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">found_row</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">species</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">species</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Species (if applicable)&#39;</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">):</span>
            <span class="n">parsed_species_name</span> <span class="o">=</span> <span class="n">species_name</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">x</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">species</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="n">parsed_species_name</span><span class="p">:</span>
                <span class="n">found_row</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="n">superbranch</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Superbranch&#39;</span><span class="p">]</span>
                <span class="n">supergroup</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Supergroup&#39;</span><span class="p">]</span>
                <span class="n">group</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Group&#39;</span><span class="p">]</span>
                <span class="n">taxonomic_info_obj</span> <span class="o">=</span> <span class="n">TaxonomicInfo</span><span class="p">(</span><span class="n">superbranch</span><span class="p">,</span> <span class="n">supergroup</span><span class="p">,</span> <span class="n">group</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">found_row</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="k">if</span> <span class="n">found_row</span><span class="p">:</span>
            <span class="k">break</span>


    <span class="k">assert</span> <span class="n">taxonomic_info_obj</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Could not identify taxonomic</span>
<span class="s2">    information for species </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">species_name</span>

    <span class="k">return</span> <span class="n">taxonomic_info_obj</span></div>


<div class="viewcode-block" id="define_nodestyles_dict_for_colourcoding"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.define_nodestyles_dict_for_colourcoding">[docs]</a><span class="k">def</span> <span class="nf">define_nodestyles_dict_for_colourcoding</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Define node styles for different nodes of interest that need to be</span>
<span class="sd">    colour-coded.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Initiate dict.</span>
    <span class="n">nodestyle_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># General default style for tree (apply using the node.set_style(style)</span>
    <span class="c1"># function).</span>
    <span class="n">style</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#0f0f0f&quot;</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;vt_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#ff0000&quot;</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#ff0000&quot;</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;vt_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">8</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">8</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;vt_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># 0 solid, 1 dashed, 2 dotted</span>
    <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># dashed blue lines</span>
    <span class="n">style1</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#000000&quot;</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;shape&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;circle&quot;</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;vt_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#0000aa&quot;</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;hz_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#0000aa&quot;</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;vt_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;vt_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># 0 solid, 1 dashed, 2 dotted</span>
    <span class="n">style1</span><span class="p">[</span><span class="s2">&quot;hz_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="c1"># stem_node</span>
    <span class="n">stem_node</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
    <span class="n">stem_node</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>

    <span class="c1"># bold lines</span>
    <span class="n">style2</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
    <span class="c1">#style2[&quot;vt_line_width&quot;] = 8</span>
    <span class="n">style2</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>

    <span class="c1"># bold dashed lines</span>
    <span class="n">style3</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
    <span class="n">style3</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>
    <span class="n">style3</span><span class="p">[</span><span class="s2">&quot;hz_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>


    <span class="c1"># Set dashed blue lines in all leaves</span>
    <span class="c1">#style3 = NodeStyle()</span>
    <span class="c1">#style3[&quot;fgcolor&quot;] = &quot;#000000&quot;</span>
    <span class="c1">#style3[&quot;shape&quot;] = &quot;circle&quot;</span>
    <span class="c1">#style3[&quot;vt_line_color&quot;] = &quot;#0000aa&quot;</span>
    <span class="c1">#style3[&quot;hz_line_color&quot;] = &quot;#0000aa&quot;</span>
    <span class="c1">#style3[&quot;vt_line_width&quot;] = 2</span>
    <span class="c1">#style3[&quot;hz_line_width&quot;] = 2</span>
    <span class="c1">#style3[&quot;vt_line_type&quot;] = 1 # 0 solid, 1 dashed, 2 dotted</span>
    <span class="c1">#style3[&quot;hz_line_type&quot;] = 1</span>


    <span class="c1"># Add styles to dictionary.</span>
    <span class="n">nodestyle_dict</span><span class="p">[</span><span class="s1">&#39;default&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">style1</span>
    <span class="n">nodestyle_dict</span><span class="p">[</span><span class="s1">&#39;stem_node&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stem_node</span>
    <span class="n">nodestyle_dict</span><span class="p">[</span><span class="s1">&#39;shortest branch&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">style2</span> <span class="c1"># bold.</span>
    <span class="n">nodestyle_dict</span><span class="p">[</span><span class="s1">&#39;longest branch&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">style3</span> <span class="c1"># dashed.</span>
    <span class="n">nodestyle_dict</span><span class="p">[</span><span class="s1">&#39;default&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">style1</span>

    <span class="c1"># Return dict.</span>
    <span class="k">return</span> <span class="n">nodestyle_dict</span></div>


<div class="viewcode-block" id="define_textface_for_labeling_stem"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.define_textface_for_labeling_stem">[docs]</a><span class="k">def</span> <span class="nf">define_textface_for_labeling_stem</span><span class="p">(</span><span class="n">clade_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Define ete3 textface styles for different nodes of interest that need to be</span>
<span class="sd">    labeled.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define TextFace objects for clade stem nodes/branches.</span>
    <span class="n">stem_clade</span> <span class="o">=</span> <span class="n">TextFace</span><span class="p">(</span><span class="n">clade_name</span><span class="p">)</span>

    <span class="c1"># Set some attributes</span>
    <span class="c1">#stem_clade.margin_top = 10</span>
    <span class="c1">#stem_clade.margin_right = 10</span>
    <span class="c1">#stem_clade.margin_left = 10</span>
    <span class="c1">#stem_clade.margin_bottom = 10</span>
    <span class="c1">#stem_clade.opacity = 0.5 # from 0 to 1</span>
    <span class="c1">#stem_clade.inner_border.width = 1 # 1 pixel border</span>
    <span class="c1">#stem_clade.inner_border.type = 1  # dashed line</span>
    <span class="c1">#stem_clade.border.width = 1</span>
    <span class="c1">#stem_clade.background.color = &quot;LightGreen&quot;</span>

    <span class="c1"># Return dictionary.</span>
    <span class="k">return</span> <span class="n">stem_clade</span></div>


<div class="viewcode-block" id="get_corresponding_node"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_corresponding_node">[docs]</a><span class="k">def</span> <span class="nf">get_corresponding_node</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">tree</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the ete3 TreeNode object in a tree that corresponds to</span>
<span class="sd">    a given node.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Search for a node with child leaf nodes that have the same set of names.</span>
    <span class="n">corresponding_node</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">node_leaves</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]</span>
    <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">([</span><span class="n">y</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">n</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()])</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">node_leaves</span><span class="p">):</span>
            <span class="n">corresponding_node</span> <span class="o">=</span> <span class="n">n</span>
            <span class="k">break</span>
    <span class="c1"># If necessary, re-root input tree on midpoint for parsing.</span>
    <span class="k">if</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Re-root on midpoint.</span>
        <span class="n">tree</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">get_midpoint_outgroup</span><span class="p">())</span>
        <span class="c1"># Search for node of interest again.</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="k">if</span> <span class="nb">set</span><span class="p">([</span><span class="n">y</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">n</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()])</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">node_leaves</span><span class="p">):</span>
                <span class="n">corresponding_node</span> <span class="o">=</span> <span class="n">n</span>
                <span class="k">break</span>

    <span class="c1"># Try rooting on random leaf nodes until the corresponding node can be</span>
    <span class="c1"># found.</span>
    <span class="k">if</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Loop through random leaf nodes in tree, rooting on them, and then</span>
        <span class="c1"># searching for the corresponding node.</span>
        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
            <span class="n">tree</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">l</span><span class="p">)</span>
            <span class="c1"># Search for node of interest again.</span>
            <span class="n">found</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">tree</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="k">if</span> <span class="nb">set</span><span class="p">([</span><span class="n">y</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">n</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()])</span> <span class="o">==</span> <span class="nb">set</span><span class="p">(</span><span class="n">node_leaves</span><span class="p">):</span>
                    <span class="n">corresponding_node</span> <span class="o">=</span> <span class="n">n</span>
                    <span class="n">found</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="n">found</span><span class="p">:</span>
                <span class="k">break</span>

    <span class="c1"># Check that a corresponding node was identified.</span>
    <span class="k">assert</span> <span class="n">corresponding_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Could not find corresponding</span>
<span class="s2">    node.&quot;&quot;&quot;</span>

    <span class="c1"># Return corresponding node.</span>
    <span class="k">return</span> <span class="n">corresponding_node</span></div>



<div class="viewcode-block" id="optimize_sequence_selection"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.optimize_sequence_selection">[docs]</a><span class="k">def</span> <span class="nf">optimize_sequence_selection</span><span class="p">(</span><span class="n">file_with_seqs</span><span class="p">,</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">essential_taxa</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add/remove homologous sequences from a dataset to optimize for long stem</span>
<span class="sd">    lengths and low average branch lengths for clades of interest.</span>

<span class="sd">    The essential_taxa argument is a list of names corresponding to taxonomic</span>
<span class="sd">    groups of eukaryotes (such as &quot;Amorphea&quot; or &quot;Diaphoretickes&quot;) which appear</span>
<span class="sd">    in the genome info csv file in the genomes directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># (make this into a recursive function?)</span>

    <span class="c1"># Define a dictionary of ete3 NodeStyle objects for different branch types</span>
    <span class="c1"># in the tree to be visualized.</span>
    <span class="n">style_dict</span> <span class="o">=</span> <span class="n">define_nodestyles_dict_for_colourcoding</span><span class="p">()</span>

    <span class="c1"># Get info about model that is relevant.</span>
    <span class="n">model_info</span> <span class="o">=</span> <span class="n">ModelInfoFromCSV</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">alignment_file</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">tree_topology_file</span>
    <span class="n">subs_model</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">subs_model</span>
    <span class="n">type_seqs</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">type_seqs_file</span>

    <span class="c1"># Parse tree using ete3.</span>
    <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Print tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Input tree:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>

    <span class="c1"># Get a list of nodes of interest (ancestral nodes for clades of interest).</span>
    <span class="n">orthogroup_nodes</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Get list of &quot;type&quot; sequences from input.</span>
    <span class="n">type_seq_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">type_seqs</span><span class="p">):</span>
        <span class="n">type_seq_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># Copy tree for colour-coding etc.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">t1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1"># Get list of nodes of interest.</span>
    <span class="n">orthogroup_nodes</span> <span class="o">=</span> <span class="n">get_nodes_of_interest</span><span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">type_seq_list</span><span class="p">)</span>

    <span class="c1"># Determine characteristics of clades of interest.</span>
    <span class="n">num</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">seq_node_list</span> <span class="ow">in</span> <span class="n">orthogroup_nodes</span><span class="p">:</span>
        <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">type_seq</span> <span class="o">=</span> <span class="n">seq_node_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">clade_name</span> <span class="o">=</span> <span class="n">get_clade_name_from_model</span><span class="p">(</span><span class="n">type_seq</span><span class="p">,</span> <span class="n">type_seqs</span><span class="p">)</span>
        <span class="n">node</span> <span class="o">=</span> <span class="n">seq_node_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\n</span><span class="s1">Clade defined by sequence &#39;</span> <span class="o">+</span> <span class="n">type_seq</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

        <span class="c1"># Get list of leaf names.</span>
        <span class="n">leaf_names</span> <span class="o">=</span> <span class="n">get_list_of_leaf_names_for_node</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

        <span class="c1"># Get stem length for node.</span>
        <span class="n">stem_len</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">dist</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Stem length: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stem_len</span><span class="p">))</span>

        <span class="c1"># Get branch lengths for all sequences in the clade.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Branch lengths:&#39;</span><span class="p">)</span>
        <span class="n">seq_branch_lengths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">n</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
                <span class="n">length</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">n</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s1">&#39;: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">length</span><span class="p">))</span>
                <span class="n">seq_branch_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">length</span><span class="p">])</span>

        <span class="c1"># Get longest branch length.</span>
        <span class="n">max_len</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">seq_branch_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">Longest branch: &#39;</span> <span class="o">+</span> <span class="n">max_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">max_len</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="c1"># Get shortest branch length.</span>
        <span class="n">min_len</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">seq_branch_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">Shortest branch: &#39;</span> <span class="o">+</span> <span class="n">min_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">min_len</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

        <span class="c1"># Get average branch length relative to ancestral/root node in the</span>
        <span class="c1"># clade for each sequence/terminal/leaf node.</span>
        <span class="n">mean_len</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq_branch_lengths</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">Mean length: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mean_len</span><span class="p">))</span>

        <span class="c1"># Get median branch length relative to ancestral/root node in the</span>
        <span class="c1"># clade for each sequence/terminal/leaf node.</span>
        <span class="n">median_len</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">median</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq_branch_lengths</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">Median length: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">median_len</span><span class="p">))</span>

        <span class="c1"># Calculate ratio of stem length to average branch length.</span>
        <span class="n">stem_branches_ratio</span> <span class="o">=</span> <span class="n">stem_len</span> <span class="o">/</span> <span class="n">mean_len</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\t</span><span class="s1">Ratio of stem len to avg branch len: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stem_branches_ratio</span><span class="p">))</span>


        <span class="c1"># Format tree for visualization.</span>

        <span class="c1"># Root on first clade of interest.</span>
        <span class="k">if</span> <span class="n">num</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">t2</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">t2</span><span class="p">))</span>

        <span class="c1"># Make stem branch for clade of interest bold.</span>
        <span class="n">stem_node</span> <span class="o">=</span> <span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span>
        <span class="n">stem_node</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style_dict</span><span class="p">[</span><span class="s1">&#39;stem_node&#39;</span><span class="p">])</span>

        <span class="c1"># Add clade name to stem branch.</span>
        <span class="c1">#stem_node.name = clade_name</span>
        <span class="n">stem_node</span><span class="o">.</span><span class="n">add_face</span><span class="p">(</span><span class="n">define_textface_for_labeling_stem</span><span class="p">(</span><span class="n">clade_name</span><span class="p">),</span> <span class="n">column</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">position</span> <span class="o">=</span> <span class="s2">&quot;branch-top&quot;</span><span class="p">)</span>

        <span class="c1"># Iterate over branches within clade and customize appearance.</span>
        <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
            <span class="c1"># Set general features for all leaf nodes.</span>
            <span class="n">style</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#0f0f0f&quot;</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;vt_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#000000&quot;</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#000000&quot;</span>

            <span class="c1">#style[&quot;vt_line_width&quot;] = 8</span>
            <span class="c1">#style[&quot;hz_line_width&quot;] = 8</span>
            <span class="c1">#style[&quot;vt_line_type&quot;] = 0 # 0 solid, 1 dashed, 2 dotted</span>
            <span class="c1">#style[&quot;hz_line_type&quot;] = 0</span>

            <span class="c1"># Make shortest branch in clade of interest bold.</span>
            <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">min_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="c1">#shortbranchnode = get_corresponding_node(t1.search_nodes(name=min_len[0])[0], t2)</span>
                <span class="c1">#shortbranchnode.set_style(style_dict[&#39;shortest branch&#39;])</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>

            <span class="c1"># Make longest branch in clade of interest bold.</span>
            <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">max_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="c1">#longbranchnode = get_corresponding_node(t1.search_nodes(name=max_len[0])[0], t2)</span>
                <span class="c1">#longbranchnode.set_style(style_dict[&#39;longest branch&#39;])</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
            
            <span class="c1"># Customize leaf node appearance.</span>
            <span class="c1"># Get species name.</span>
            <span class="n">species_name</span> <span class="o">=</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
            <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">ti</span><span class="o">.</span><span class="n">superbranch</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">ti</span><span class="o">.</span><span class="n">supergroup</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">ti</span><span class="o">.</span><span class="n">group</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">ti</span><span class="o">.</span><span class="n">species</span><span class="p">)]))</span>

            <span class="c1"># Set style for node.</span>
            <span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">leaf</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="p">)</span>


    <span class="c1"># Show tree with colour-coded branches.</span>

    <span class="c1"># Remove underscores from leaf names.</span>
    <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t2</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
        <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>

    <span class="c1"># Stretch branches.</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">TreeStyle</span><span class="p">()</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">show_leaf_name</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">scale</span> <span class="o">=</span>  <span class="mi">120</span> <span class="c1"># 120 pixels per branch length unit</span>

    <span class="c1"># Add title.</span>
    <span class="n">tree_title</span> <span class="o">=</span> <span class="s2">&quot;[Tree title here]&quot;</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">add_face</span><span class="p">(</span><span class="n">TextFace</span><span class="p">(</span><span class="n">tree_title</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="mi">20</span><span class="p">),</span> <span class="n">column</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Show.</span>
    <span class="n">t2</span><span class="o">.</span><span class="n">show</span><span class="p">(</span><span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>

    <span class="c1"># Write tree to file.</span>
    <span class="n">t2</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="s2">&quot;/Users/Lael/Desktop/mytree.png&quot;</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">183</span><span class="p">,</span> <span class="n">units</span><span class="o">=</span><span class="s2">&quot;mm&quot;</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>

    <span class="c1"># Return main output path.</span>
    <span class="k">return</span> <span class="s1">&#39;[no output path defined]&#39;</span></div>


<div class="viewcode-block" id="get_ml_tree_branch_lengths"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_ml_tree_branch_lengths">[docs]</a><span class="k">def</span> <span class="nf">get_ml_tree_branch_lengths</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">outputdir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Run an ML tree with a constrained topology given a tree (for</span>
<span class="sd">    constraining topology), an alignment file, and a substitution model.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Convert nexus alignment to afa.</span>
    <span class="n">temp_fa_1</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputdir</span><span class="p">,</span>\
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_temp1.afa&#39;</span><span class="p">)</span>
    <span class="n">nex_to_afa</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">temp_fa_1</span><span class="p">)</span> 

    <span class="c1"># Code names in alignment.</span>
    <span class="n">outalifp</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.afa&#39;</span>
    <span class="n">outtablefp</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.table&#39;</span>
    <span class="n">code_names_in_ali</span><span class="p">(</span><span class="n">outputdir</span><span class="p">,</span> <span class="n">temp_fa_1</span><span class="p">,</span> <span class="n">outalifp</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span> 

    <span class="c1"># Convert alignment back to nexus.</span>
    <span class="n">outalifpnex</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.nex&#39;</span>
    <span class="n">afa_to_nex</span><span class="p">(</span><span class="n">outalifp</span><span class="p">,</span> <span class="n">outalifpnex</span><span class="p">)</span>

    <span class="c1"># Delete temporary files.</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">temp_fa_1</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">outalifp</span><span class="p">)</span>

    <span class="c1"># Define path for intermediate unrooted tree.</span>
    <span class="n">tree2</span> <span class="o">=</span> <span class="n">outalifpnex</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_unrooted_tree.newick&#39;</span>

    <span class="c1"># Parse tree using ete3.</span>
    <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Unroot tree.</span>
    <span class="n">t1</span><span class="o">.</span><span class="n">unroot</span><span class="p">()</span>

    <span class="c1"># Write unrooted tree to a new file.</span>
    <span class="n">t1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outfile</span><span class="o">=</span><span class="n">tree2</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>

    <span class="c1"># Define path for constraint tree file.</span>
    <span class="n">constraint_tree_fp_coded</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="c1"># Only do the following steps once for all the alignments, because all the</span>
    <span class="c1"># constrained topology analyses will use the same constraint trees.</span>

    <span class="c1"># Put quotation marks around names in constraint tree. # SEPARATE</span>
    <span class="c1">#outtreefp = ali.rsplit(&#39;.&#39;, 1)[0] + &#39;.constraint_tree0.Q.newick&#39;</span>
    <span class="n">constraint_tree_fp</span> <span class="o">=</span> <span class="n">outalifpnex</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_constraint_tree.newick&#39;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tree2</span><span class="p">)</span> <span class="k">as</span> <span class="n">intreefh</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">constraint_tree_fp</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">intreefh</span><span class="p">:</span>
            <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">quote_tree</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">))</span>
    <span class="c1"># Make constraint tree with coded names.</span>
    <span class="n">constraint_tree_fp_coded</span> <span class="o">=</span> <span class="n">constraint_tree_fp</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.tre&#39;</span>
    <span class="n">code_tree</span><span class="p">(</span><span class="n">constraint_tree_fp</span><span class="p">,</span> <span class="n">constraint_tree_fp_coded</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span>

    <span class="c1"># Make a copy of the alignment in phylip format for input to IQtree.</span>
    <span class="n">phy_out</span> <span class="o">=</span> <span class="n">outalifpnex</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.phy&#39;</span>
    <span class="n">nex_to_phylip</span><span class="p">(</span><span class="n">outalifpnex</span><span class="p">,</span> <span class="n">phy_out</span><span class="p">)</span>

    <span class="c1"># Do phylogenetic analysis.</span>

    <span class="c1"># Make subdir for output.</span>
    <span class="n">subdirp</span> <span class="o">=</span> <span class="n">constraint_tree_fp</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_IQ-tree_output&#39;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">subdirp</span><span class="p">)</span>

    <span class="c1"># Use IQtree to do an ML search</span>
    <span class="n">output_file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">subdirp</span><span class="p">,</span> <span class="s1">&#39;iqtree&#39;</span><span class="p">)</span>
    <span class="n">iqtree_command_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;iqtree&#39;</span><span class="p">,</span> <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">phy_out</span><span class="p">,</span> <span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="s1">&#39;-g&#39;</span><span class="p">,</span>
        <span class="n">constraint_tree_fp_coded</span><span class="p">,</span> <span class="s1">&#39;-pre&#39;</span><span class="p">,</span> <span class="n">output_file_prefix</span><span class="p">,</span> <span class="s1">&#39;-nt&#39;</span><span class="p">,</span> <span class="s1">&#39;4&#39;</span><span class="p">]</span>
    <span class="n">stdout_path</span> <span class="o">=</span> <span class="n">output_file_prefix</span> <span class="o">+</span> <span class="s1">&#39;.stdout.txt&#39;</span>
    <span class="n">tree_search_start_time</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">stdout_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">iqtree_command_list</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">o</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">)</span>
    <span class="c1">#tree_search_end_time = time.time()</span>
    <span class="c1">#with open(logfile, &#39;a&#39;) as logh:</span>
    <span class="c1">#    logh.write(&#39; &#39;.join(iqtree_command_list) + &#39;\n&#39;)</span>
    <span class="c1">#    tree_search_elapsed = tree_search_end_time - tree_search_start_time</span>
    <span class="c1">#    logh.write(&#39;Run time: &#39; + str(datetime.timedelta(seconds=tree_search_elapsed)) + &#39;\n&#39;)</span>

    <span class="c1"># Define path to output file.</span>
    <span class="n">tree_file_path</span> <span class="o">=</span> <span class="n">output_file_prefix</span> <span class="o">+</span> <span class="s1">&#39;.treefile&#39;</span>
    <span class="c1"># Check that the output file was actually produced.</span>
    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">tree_file_path</span><span class="p">)</span>

    <span class="c1"># Parse output tree.</span>
    <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>

    <span class="c1"># Uncode tree.</span>
    <span class="n">tree_file_path2</span> <span class="o">=</span> <span class="n">tree_file_path</span> <span class="o">+</span> <span class="s1">&#39;2&#39;</span>
    <span class="n">uncode_tree</span><span class="p">(</span><span class="n">tree_file_path</span><span class="p">,</span> <span class="n">tree_file_path2</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span>

    <span class="c1">#with open(tree_file_path2) as infh:</span>
    <span class="c1">#    for i in infh:</span>
    <span class="c1">#        print(i)</span>


    <span class="c1">#with open(tree_file_path) as infh, open(tree_file_path2, &#39;w&#39;) as o:</span>
    <span class="c1">#    for t in infh:</span>
    <span class="c1">#        # Add quotation marks around node names and write to 2nd file.</span>
    <span class="c1">#        print(&#39;old_tree_text&#39;)</span>
    <span class="c1">#        print(t)</span>
    <span class="c1">#        new_tree_text = quote_tree(t, outtablefp)</span>
    <span class="c1">#        print(&#39;new_tree_text&#39;)</span>
    <span class="c1">#        print(new_tree_text)</span>
    <span class="c1">#        o.write(new_tree_text)</span>
            
    <span class="c1"># Parse new tree file.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree_file_path2</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


    <span class="c1"># Return the tree.</span>
    <span class="k">return</span> <span class="n">t2</span> </div>


<div class="viewcode-block" id="get_branch_length_info"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.get_branch_length_info">[docs]</a><span class="k">def</span> <span class="nf">get_branch_length_info</span><span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">,</span> <span class="n">outpath</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take an ete3 Tree object and a file with type seqs, and return a</span>
<span class="sd">    dictionary with keys as type sequence names and values as branch length</span>
<span class="sd">    ratios for the respective clades in the input tree.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define a dictionary of ete3 NodeStyle objects for different branch types</span>
    <span class="c1"># in the tree to be visualized.</span>
    <span class="n">style_dict</span> <span class="o">=</span> <span class="n">define_nodestyles_dict_for_colourcoding</span><span class="p">()</span>

    <span class="c1"># Get a list of nodes of interest (ancestral nodes for clades of interest).</span>
    <span class="n">orthogroup_nodes</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1"># Get list of &quot;type&quot; sequences from input.</span>
    <span class="c1">#type_seq_list = []</span>
    <span class="c1">#for i in open(type_seqs):</span>
    <span class="c1">#    type_seq_list.append(i.strip().split(&#39;,&#39;)[0])</span>
    <span class="n">type_seq_list</span> <span class="o">=</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>

    <span class="c1"># Copy tree for colour-coding etc.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">t1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="c1"># Root on the midpoint, because ete3 was having trouble identifying clades</span>
    <span class="c1"># in some of the unrooted trees.</span>
    <span class="n">t2</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">t2</span><span class="o">.</span><span class="n">get_midpoint_outgroup</span><span class="p">())</span>

    <span class="c1"># Get list of nodes of interest.</span>
    <span class="n">orthogroup_nodes</span> <span class="o">=</span> <span class="n">get_nodes_of_interest</span><span class="p">(</span><span class="n">t1</span><span class="p">,</span> <span class="n">type_seq_list</span><span class="p">)</span>


    <span class="c1"># Get lengths for internal branches (excluding those within the specific</span>
    <span class="c1"># clades of interest) and add those to the dictionary as a list.</span>

    <span class="c1"># Traverse all the nodes in the tree, excluding those in the clades of</span>
    <span class="c1"># interest.</span>
    <span class="n">internal_branch_lengths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># Copy the tree object.</span>
    <span class="n">t3</span> <span class="o">=</span> <span class="n">t2</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># Record initial number of nodes.</span>
    <span class="n">initial_node_num</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">()))</span>

    <span class="c1"># Record which node is the original outgroup.???</span>
    <span class="n">original_outgroup</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">())[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Find nodes of interest in new tree object for detaching.</span>
    <span class="n">orthogroup_nodes_t3</span> <span class="o">=</span> <span class="n">get_nodes_from_actual_tree_obj</span><span class="p">(</span><span class="n">t3</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">orthogroup_nodes</span><span class="p">])</span>

    <span class="c1"># Re-root on one of the nodes of interest.</span>
    <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">orthogroup_nodes_t3</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1">#print(&#39;\nt3 before detaching nodes of interest:&#39;)</span>
    <span class="c1">#print(t3)</span>

    <span class="c1"># Detach all nodes of interest in the copy of the tree.</span>
    <span class="c1">#for i in to_remove_list:</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">orthogroup_nodes_t3</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">():</span>
            <span class="n">j</span><span class="o">.</span><span class="n">detach</span><span class="p">()</span>
    <span class="c1">#print(&#39;\ntree with nodes removed (is this right???***):&#39;)</span>
    <span class="c1">#print(t3)</span>

    <span class="c1"># Record (reduced) number of nodes.</span>
    <span class="n">reduced_node_num</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">()))</span>

    <span class="c1"># Check that at least one node was actually removed.</span>
    <span class="k">assert</span> <span class="n">reduced_node_num</span> <span class="o">&lt;</span> <span class="n">initial_node_num</span>
    <span class="c1"># Check that the tree has the minimum number of nodes.</span>
    <span class="k">assert</span> <span class="n">reduced_node_num</span> <span class="o">&gt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">orthogroup_nodes</span><span class="p">)</span>

    <span class="c1"># Get internal branch lengths from pruned copy of the tree.</span>
    <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
        <span class="n">internal_branch_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">dist</span><span class="p">)</span>

    <span class="c1"># Check that a minimum number of internal nodes were identified.</span>
    <span class="c1">#assert len(internal_branch_lengths) &gt;= len(orthogroup_nodes), &quot;&quot;&quot;Too few</span>
    <span class="c1">#internal nodes identified.&quot;&quot;&quot;</span>


    <span class="c1"># Initiate output dictionary.</span>
    <span class="n">branch_length_info_dict</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="c1"># Determine characteristics of clades of interest.</span>
    <span class="n">num</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">seq_node_list</span> <span class="ow">in</span> <span class="n">orthogroup_nodes</span><span class="p">:</span>
        <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">type_seq</span> <span class="o">=</span> <span class="n">seq_node_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">clade_name</span> <span class="o">=</span> <span class="n">get_clade_name_from_model2</span><span class="p">(</span><span class="n">type_seq</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">)</span>
        <span class="n">node</span> <span class="o">=</span> <span class="n">seq_node_list</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1">#print(&#39;\n\nClade defined by sequence &#39; + type_seq + &#39;:&#39;)</span>
        <span class="c1">#print(node)</span>

        <span class="c1"># Get list of leaf names.</span>
        <span class="n">leaf_names</span> <span class="o">=</span> <span class="n">get_list_of_leaf_names_for_node</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

        <span class="c1"># Get stem length for node.</span>
        <span class="n">stem_len</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">dist</span>
        <span class="c1">#print(&#39;\nStem length: &#39; + str(stem_len))</span>

        <span class="c1"># Get branch lengths for all sequences in the clade.</span>
        <span class="c1">#print(&#39;\nBranch lengths:&#39;)</span>
        <span class="n">seq_branch_lengths</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">n</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
                <span class="n">length</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
                <span class="c1">#print(&#39;\t&#39; + n.name + &#39;: &#39; + str(length))</span>
                <span class="n">seq_branch_lengths</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">n</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">length</span><span class="p">])</span>

        <span class="c1"># Get longest branch length.</span>
        <span class="n">max_len</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">seq_branch_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#print(&#39;\n\tLongest branch: &#39; + max_len[0] + &#39; &#39; + str(max_len[1]))</span>

        <span class="c1"># Get shortest branch length.</span>
        <span class="n">min_len</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">seq_branch_lengths</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#print(&#39;\n\tShortest branch: &#39; + min_len[0] + &#39; &#39; + str(min_len[1]))</span>

        <span class="c1"># Get average branch length relative to ancestral/root node in the</span>
        <span class="c1"># clade for each sequence/terminal/leaf node.</span>
        <span class="n">mean_len</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq_branch_lengths</span><span class="p">])</span>
        <span class="c1">#print(&#39;\n\tMean length: &#39; + str(mean_len))</span>

        <span class="c1"># Get median branch length relative to ancestral/root node in the</span>
        <span class="c1"># clade for each sequence/terminal/leaf node.</span>
        <span class="n">median_len</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">median</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq_branch_lengths</span><span class="p">])</span>
        <span class="c1">#print(&#39;\n\tMedian length: &#39; + str(median_len))</span>

        <span class="c1"># Calculate ratio of stem length to average branch length.</span>
        <span class="n">stem_branches_ratio</span> <span class="o">=</span> <span class="n">stem_len</span> <span class="o">/</span> <span class="n">mean_len</span>
        <span class="c1">#print(&#39;\n\tRatio of stem len to avg branch len: &#39; + str(stem_branches_ratio))</span>


        <span class="c1"># Format tree for visualization.</span>

        <span class="c1"># Root on first clade of interest.</span>
        <span class="k">if</span> <span class="n">num</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1">#print(&#39;\n\n\n\n&#39;)</span>
            <span class="c1">#print(&#39;t2:&#39;)</span>
            <span class="c1">#print(t2)</span>
            <span class="c1">#print(t2.write())</span>
            <span class="c1">#print(&#39;node:&#39;)</span>
            <span class="c1">#print(node)</span>
            <span class="c1">#print(&#39;\n\n\n\n&#39;)</span>
            <span class="n">t2</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">node</span><span class="p">,</span><span class="n">t2</span><span class="p">))</span>

        <span class="c1"># Make stem branch for clade of interest bold.</span>
        <span class="n">stem_node</span> <span class="o">=</span> <span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">node</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span>
        <span class="n">stem_node</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style_dict</span><span class="p">[</span><span class="s1">&#39;stem_node&#39;</span><span class="p">])</span>

        <span class="c1"># Add clade name to stem branch.</span>
        <span class="c1">#stem_node.name = clade_name</span>
        <span class="n">stem_node</span><span class="o">.</span><span class="n">add_face</span><span class="p">(</span><span class="n">define_textface_for_labeling_stem</span><span class="p">(</span><span class="n">clade_name</span><span class="p">),</span> <span class="n">column</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">position</span> <span class="o">=</span> <span class="s2">&quot;branch-top&quot;</span><span class="p">)</span>

        <span class="c1"># Iterate over branches within clade and customize appearance.</span>
        <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
            <span class="c1"># Set general features for all leaf nodes.</span>
            <span class="n">style</span> <span class="o">=</span> <span class="n">NodeStyle</span><span class="p">()</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;fgcolor&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#0f0f0f&quot;</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;vt_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#000000&quot;</span>
            <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_color&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;#000000&quot;</span>

            <span class="c1">#style[&quot;vt_line_width&quot;] = 8</span>
            <span class="c1">#style[&quot;hz_line_width&quot;] = 8</span>
            <span class="c1">#style[&quot;vt_line_type&quot;] = 0 # 0 solid, 1 dashed, 2 dotted</span>
            <span class="c1">#style[&quot;hz_line_type&quot;] = 0</span>

            <span class="c1"># Make shortest branch in clade of interest bold.</span>
            <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">min_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="c1">#shortbranchnode = get_corresponding_node(t1.search_nodes(name=min_len[0])[0], t2)</span>
                <span class="c1">#shortbranchnode.set_style(style_dict[&#39;shortest branch&#39;])</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>

            <span class="c1"># Make longest branch in clade of interest bold.</span>
            <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">max_len</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="c1">#longbranchnode = get_corresponding_node(t1.search_nodes(name=max_len[0])[0], t2)</span>
                <span class="c1">#longbranchnode.set_style(style_dict[&#39;longest branch&#39;])</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_width&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">5</span>
                <span class="n">style</span><span class="p">[</span><span class="s2">&quot;hz_line_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
            
            <span class="c1"># Customize leaf node appearance.</span>
            <span class="c1"># Get species name.</span>
            <span class="n">species_name</span> <span class="o">=</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="c1">#print(species_name)</span>

            <span class="c1"># Check that the name does not end with &#39;.copy&#39;.</span>
            <span class="k">assert</span> <span class="ow">not</span> <span class="n">species_name</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.copy&#39;</span><span class="p">),</span> <span class="s2">&quot;&quot;&quot;A sequence with an</span>
<span class="s2">            identical name to a sequence already in the alignment/tree was</span>
<span class="s2">            apparently added: </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span>

            <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
            <span class="c1">#print(&#39;\t&#39; + &#39; &#39;.join([str(ti.superbranch), str(ti.supergroup),</span>
            <span class="c1">#    str(ti.group), str(ti.species)]))</span>

            <span class="c1"># Set style for node.</span>
            <span class="n">get_corresponding_node</span><span class="p">(</span><span class="n">leaf</span><span class="p">,</span> <span class="n">t2</span><span class="p">)</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="n">style</span><span class="p">)</span>

        <span class="c1"># Sort seq names by descending branch lengths.</span>
        <span class="n">seq_names_by_br_len</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">(),</span>\
                <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">n</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">n</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)]</span>

        <span class="c1"># Identify most &quot;basal&quot; sequence in the clade, and the length of the</span>
        <span class="c1"># stem for the remaining seqs.</span>
        <span class="c1"># The node that is sister to the most basal sequence would be the one</span>
        <span class="c1"># that has the most daughter nodes.</span>
        <span class="n">basal_node</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">node_sister_to_basal_node</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">iter_descendants</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">node_sister_to_basal_node</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">node_sister_to_basal_node</span> <span class="o">=</span> <span class="n">n</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">n</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">())</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">node_sister_to_basal_node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()):</span>
                <span class="n">node_sister_to_basal_node</span> <span class="o">=</span> <span class="n">n</span>
        <span class="n">sister_leaf_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node_sister_to_basal_node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]</span>
        <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">node</span><span class="p">:</span>
            <span class="n">in_sister</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">n</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">sister_leaf_names</span><span class="p">:</span> 
                    <span class="n">in_sister</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">in_sister</span><span class="p">:</span>
                <span class="n">basal_node</span> <span class="o">=</span> <span class="n">n</span>
                <span class="k">break</span>
        <span class="c1"># Check whether the basal node was identified.</span>
        <span class="k">assert</span> <span class="n">basal_node</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="c1"># Define name of basal node.</span>
        <span class="n">basal_node_name</span> <span class="o">=</span> <span class="s1">&#39;clade&#39;</span>
        <span class="k">if</span> <span class="n">basal_node</span><span class="o">.</span><span class="n">is_leaf</span><span class="p">():</span>
            <span class="n">basal_node_name</span> <span class="o">=</span> <span class="n">basal_node</span><span class="o">.</span><span class="n">name</span>
        <span class="c1"># Define stem length of the clade that is sister to the &quot;basal&quot; clade.</span>
        <span class="n">basal_node_depth</span> <span class="o">=</span> <span class="n">node</span><span class="o">.</span><span class="n">get_distance</span><span class="p">(</span><span class="n">node_sister_to_basal_node</span><span class="p">)</span>


        <span class="c1"># Add info to output dictionary.</span>
        <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">clade_name</span><span class="p">]</span> <span class="o">=</span>\
                <span class="p">{</span><span class="s1">&#39;stem length&#39;</span><span class="p">:</span> <span class="n">stem_len</span><span class="p">,</span>
                 <span class="s1">&#39;average branch length&#39;</span><span class="p">:</span> <span class="n">mean_len</span><span class="p">,</span>
                 <span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">:</span> <span class="n">stem_branches_ratio</span><span class="p">,</span>
                 <span class="s1">&#39;seq names by descending length&#39;</span><span class="p">:</span> <span class="n">seq_names_by_br_len</span><span class="p">,</span>
                 <span class="s1">&#39;basal node name&#39;</span><span class="p">:</span> <span class="n">basal_node_name</span><span class="p">,</span>
                 <span class="s1">&#39;basal node depth&#39;</span><span class="p">:</span> <span class="n">basal_node_depth</span><span class="p">,</span>
                 <span class="s1">&#39;internal branch lengths&#39;</span><span class="p">:</span> <span class="n">internal_branch_lengths</span>
                 <span class="p">}</span>




    <span class="c1"># Show tree annotations.</span>

    <span class="c1"># Remove underscores from leaf names.</span>
    <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">t2</span><span class="o">.</span><span class="n">iter_leaves</span><span class="p">():</span>
        <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span>

    <span class="c1"># Stretch branches.</span>
    <span class="n">ts</span> <span class="o">=</span> <span class="n">TreeStyle</span><span class="p">()</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">show_leaf_name</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">scale</span> <span class="o">=</span>  <span class="mi">120</span> <span class="c1"># 120 pixels per branch length unit</span>

    <span class="c1"># Add title.</span>
    <span class="n">tree_title</span> <span class="o">=</span> <span class="s2">&quot;[Tree title here]&quot;</span>
    <span class="n">ts</span><span class="o">.</span><span class="n">title</span><span class="o">.</span><span class="n">add_face</span><span class="p">(</span><span class="n">TextFace</span><span class="p">(</span><span class="n">tree_title</span><span class="p">,</span> <span class="n">fsize</span><span class="o">=</span><span class="mi">20</span><span class="p">),</span> <span class="n">column</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Show.</span>
    <span class="c1">#t2.show(tree_style=ts)</span>

    <span class="c1"># Write tree to file.</span>
    <span class="k">if</span> <span class="n">outpath</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">t2</span><span class="o">.</span><span class="n">render</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">183</span><span class="p">,</span> <span class="n">units</span><span class="o">=</span><span class="s2">&quot;mm&quot;</span><span class="p">,</span> <span class="n">tree_style</span><span class="o">=</span><span class="n">ts</span><span class="p">)</span>

    <span class="c1"># Return main output path.</span>
    <span class="k">return</span> <span class="n">branch_length_info_dict</span></div>


<div class="viewcode-block" id="reduce_alignment"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.reduce_alignment">[docs]</a><span class="k">def</span> <span class="nf">reduce_alignment</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">,</span> <span class="n">output_alignment_file</span><span class="p">,</span> <span class="n">removal_name_list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take file path for an existing alignment and an output alignment, and</span>
<span class="sd">    a list of sequence names for removal. Write the reduced alignment to the</span>
<span class="sd">    output path in nexus format. Note: This function assumes that the names in</span>
<span class="sd">    the removal lit and the alignment are all either coded or uncoded.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Remove extra mesquite lines if present.</span>
    <span class="n">delete_extra_mesquite_lines</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">)</span>

    <span class="c1"># Make a copy of the input alignment file with the listed sequences</span>
    <span class="c1"># removed.</span>
    <span class="n">coded_seqs_to_not_remove_from_dataset</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">alignment_file</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_alignment_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">infh</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
        <span class="c1">#print(&#39;number of seqs in original alignment&#39;)</span>
        <span class="c1">#print(len(alignment))</span>
        <span class="n">alignment2</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([])</span>
        <span class="c1">#print(&#39;number of seqs in alignment2:&#39;)</span>
        <span class="c1">#print(len(alignment2))</span>
        <span class="c1">#print(&#39;Removing sequences from alignment.&#39;)</span>
        <span class="c1">#ali_len = len(alignment)</span>
        <span class="n">inum</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> 
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">alignment</span><span class="p">:</span>
            <span class="n">inum</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">removal_name_list</span><span class="p">:</span>
                <span class="n">alignment2</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alignment</span><span class="p">[</span><span class="n">inum</span><span class="p">])</span>

        <span class="n">alignment3</span> <span class="o">=</span>\
        <span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">alignment2</span><span class="p">[:,:])</span>
        <span class="c1">#print(&#39;number of seqs in alignment3:&#39;)</span>
        <span class="c1">#print(len(alignment3))</span>

        <span class="c1"># Get list of sequences in new alignment.</span>
        <span class="n">ids_in_alignment3</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">alignment3</span><span class="p">:</span>
            <span class="n">ids_in_alignment3</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

        <span class="c1"># Write reduced alignment.</span>
        <span class="c1">#print(&#39;Writing modified alignment file:\n\t%s&#39; % o.name)</span>
        <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">alignment3</span><span class="p">,</span> <span class="n">o</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="recursively_remove_seqs"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.recursively_remove_seqs">[docs]</a><span class="k">def</span> <span class="nf">recursively_remove_seqs</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">,</span>
                            <span class="n">t1</span><span class="p">,</span>
                            <span class="n">alignment</span><span class="p">,</span>
                            <span class="n">subs_model</span><span class="p">,</span>
                            <span class="n">type_seqs_dict</span><span class="p">,</span>
                            <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                            <span class="n">essential_taxa</span><span class="p">,</span>
                            <span class="n">branch_length_info_dict</span><span class="p">,</span>
                            <span class="n">log_file_path</span><span class="p">,</span>
                            <span class="n">optimize_internal_branches</span>
                            <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a tree, and try to remove sequences to improve the ratio of the</span>
<span class="sd">    stem lengths of clades to the branch lengths of the branches in the clades,</span>
<span class="sd">    and return the final tree when all the non-taxonomically-essential</span>
<span class="sd">    sequences have been removed or no sequence can be removed that will result</span>
<span class="sd">    in improving the branch length ratios.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define variable for whether any new tree could be produced.</span>
    <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variable for whether any new tree is better or not.</span>
    <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variables to store info for next recursion, if necessary.</span>
    <span class="n">new_tree</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_alignment</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_subs_model</span> <span class="o">=</span> <span class="n">subs_model</span>
    <span class="n">new_type_seqs_dict</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Define variables for writing to log file.</span>
    <span class="n">action</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="s1">&#39;No output tree&#39;</span>

    <span class="c1"># Determine average of stem length to branch length ratio over clades of</span>
    <span class="c1"># interest.</span>
    <span class="c1">#branch_length_info_dict = get_branch_length_info(t1, type_seqs_dict)</span>
    <span class="n">average_over_clades</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

    <span class="c1"># Determine what sequence to remove.</span>

    <span class="c1"># Determine the clade with the worst ratio.</span>
    <span class="n">clades_by_ascending_ratio</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>\
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">])</span>
    <span class="c1">#clade_with_lowest_ratio = clades_by_ascending_ratio[0]</span>

    <span class="k">for</span> <span class="n">clade</span> <span class="ow">in</span> <span class="n">clades_by_ascending_ratio</span><span class="p">:</span>
        <span class="c1"># Make a list of taxonomic terms currently represented in the clade.</span>
        <span class="n">taxa_represented</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">seqname</span> <span class="ow">in</span>\
        <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;seq names by descending length&#39;</span><span class="p">]:</span>
            <span class="c1"># Get species name from sequence name.</span>
            <span class="n">species_name</span> <span class="o">=</span> <span class="n">seqname</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="c1"># Get taxonomic information for species name from info csv file.</span>
            <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
            <span class="c1"># Add taxon names to taxa represented.</span>
            <span class="n">taxa_represented</span> <span class="o">=</span> <span class="n">taxa_represented</span>\
                    <span class="o">+</span> <span class="p">[</span><span class="n">ti</span><span class="o">.</span><span class="n">superbranch</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">supergroup</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">group</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">species</span><span class="p">]</span>

        <span class="c1"># Determine which seqs in the clade are dispensable, considering their</span>
        <span class="c1"># taxonomic placement.</span>
        <span class="k">for</span> <span class="n">seqname</span> <span class="ow">in</span>\
        <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;seq names by descending length&#39;</span><span class="p">]:</span>
            <span class="c1"># Get species name from sequence name.</span>
            <span class="n">species_name</span> <span class="o">=</span> <span class="n">seqname</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="c1"># Get taxonomic information for species name from info csv file.</span>
            <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
            <span class="n">ti_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">ti</span><span class="o">.</span><span class="n">superbranch</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">supergroup</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">group</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">species</span><span class="p">]</span>
            <span class="c1"># Determine whether any of the taxonomy terms are in the essential</span>
            <span class="c1"># list.</span>
            <span class="n">seq_is_essential</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">essential_terms</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ti_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">essential_taxa</span><span class="p">:</span>
                    <span class="n">essential_terms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
            <span class="c1"># Check whether any of the terms appear in the taxa_represented list</span>
            <span class="c1"># only once (if so then the sequence is essential).</span>
            <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">essential_terms</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">taxa_represented</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">term</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">seq_is_essential</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Determine whether a sequence can be removed, if not then check the</span>
            <span class="c1"># next clade.</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">seq_is_essential</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">seqname</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seqs_attempted_to_remove</span><span class="p">:</span>
                    <span class="c1"># Try removing the corresponding sequence from the</span>
                    <span class="c1"># tree, and see whether it improves the branch length</span>
                    <span class="c1"># ratios.</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">removing sequence </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">seqname</span><span class="p">)</span>


                    <span class="c1"># If the sequence is a type sequence, then you have to</span>
                    <span class="c1"># choose another sequence in the clade as a replacement</span>
                    <span class="c1"># type sequence.</span>
                    <span class="n">new_type_seqs_dict</span> <span class="o">=</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="n">is_type_seq</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">cladex</span> <span class="ow">in</span> <span class="n">new_type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">seqname</span> <span class="o">==</span> <span class="n">new_type_seqs_dict</span><span class="p">[</span><span class="n">cladex</span><span class="p">]:</span>
                            <span class="n">is_type_seq</span> <span class="o">=</span> <span class="kc">True</span>
                            <span class="k">assert</span> <span class="n">clade</span> <span class="o">==</span> <span class="n">cladex</span>
                    <span class="k">if</span> <span class="n">is_type_seq</span><span class="p">:</span>
                        <span class="c1"># Identify an alternative type sequence for the same</span>
                        <span class="c1"># (current) clade.</span>
                        <span class="k">for</span> <span class="n">seqname2</span> <span class="ow">in</span>\
                        <span class="nb">reversed</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">]</span>\
                        <span class="p">[</span><span class="s1">&#39;seq names by descending length&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()):</span>
                            <span class="k">if</span> <span class="n">seqname2</span> <span class="o">!=</span> <span class="n">seqname</span><span class="p">:</span>
                                <span class="c1"># Change type seq for clade to a different sequence.</span>
                                <span class="n">new_type_seqs_dict</span><span class="p">[</span><span class="n">clade</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqname2</span>


                    <span class="c1"># Define new alignment path.</span>
                    <span class="n">alignment2</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span>\
                    <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.nex&#39;</span>
                    <span class="c1"># Remove seq from alignment.</span>
                    <span class="n">reduce_alignment</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="p">[</span><span class="n">seqname</span><span class="p">])</span>
                    <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">alignment2</span>

                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">removed seq from alignment&#39;</span><span class="p">)</span>

                    <span class="c1"># Code alignment (get table file).</span>

                    <span class="c1"># Remove seq from a copy of the parsed tree object, and</span>
                    <span class="c1"># use that as a constraint tree.</span>
                    <span class="c1"># Copy tree.</span>
                    <span class="n">t2</span> <span class="o">=</span> <span class="n">t1</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                    <span class="c1"># Get a list of all node objects with the same name as the</span>
                    <span class="c1"># sequence of interest.</span>
                    <span class="n">seq_nodes</span> <span class="o">=</span> <span class="n">t2</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">seqname</span><span class="p">)</span>
                    <span class="c1"># Check that only one such node was found.</span>
                    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_nodes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
                    <span class="c1"># Remove the node from the new tree object.</span>
                    <span class="n">node_to_remove</span> <span class="o">=</span> <span class="n">seq_nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">t2_len_before</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">t2</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">())</span>
                    <span class="n">node_to_remove</span><span class="o">.</span><span class="n">delete</span><span class="p">()</span>

                    <span class="c1"># Check that a node was actually removed.</span>
                    <span class="n">t2_len_after</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">t2</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">())</span>
                    <span class="k">assert</span> <span class="n">t2_len_before</span> <span class="o">==</span> <span class="n">t2_len_after</span> <span class="o">+</span> <span class="mi">1</span>

                    <span class="c1"># Write modified tree to a file.</span>
                    <span class="n">tfp</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_newick.tre&#39;</span><span class="p">)</span>
                    <span class="n">t2</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outfile</span><span class="o">=</span><span class="n">tfp</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>

                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">removed seq from tree object&#39;</span><span class="p">)</span>

                    <span class="c1"># Run ML tree with IQtree, and get branch length info</span>
                    <span class="c1"># dictionary.</span>
                    <span class="c1"># Define output directory for files.</span>
                    <span class="n">outputdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
                    <span class="c1"># Call function to run IQtree and return a parsed tree obj.</span>
                    <span class="n">t3</span> <span class="o">=</span> <span class="n">get_ml_tree_branch_lengths</span><span class="p">(</span><span class="n">tfp</span><span class="p">,</span> <span class="n">alignment2</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">outputdir</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">ran IQtree&#39;</span><span class="p">)</span>
                    <span class="c1"># Set t3 as the new tree.</span>
                    <span class="n">new_tree</span> <span class="o">=</span> <span class="n">t3</span>
                    <span class="c1"># Define output path for annotated tree image file.</span>
                    <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
                    <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="n">annotated_tree_outpath</span>
                    <span class="c1"># Extract info about branch lengths from tree, and write</span>
                    <span class="c1"># annotated tree image file.</span>
                    <span class="n">new_branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t3</span><span class="p">,</span>
                                              <span class="n">new_type_seqs_dict</span><span class="p">,</span>
                                              <span class="n">annotated_tree_outpath</span><span class="p">)</span>
                    <span class="c1"># Calculate average ratio over all clades from branch</span>
                    <span class="c1"># length info. </span>
                    <span class="n">new_average_over_clades</span> <span class="o">=</span>\
                            <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
                            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">average ratio over clades: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span>\
                            <span class="nb">str</span><span class="p">(</span><span class="n">new_average_over_clades</span><span class="p">))</span>

                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">parsed IQtree output.&#39;</span><span class="p">)</span>

                    <span class="c1"># Write a file with type sequences and clade names for this</span>
                    <span class="c1"># iteration.</span>
                    <span class="n">new_type_seqs_file</span><span class="o">=</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_type_seqs.csv&#39;</span><span class="p">)</span>
                    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">new_type_seqs_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
                        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                            <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

                    <span class="c1"># Compare lengths of internal branches.</span>
                    <span class="n">previous_internal_branch_lengths</span> <span class="o">=</span>\
                    <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
                    <span class="n">new_internal_branch_lengths</span> <span class="o">=</span>\
                    <span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
                    <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">,</span> <span class="n">new_internal_branch_lengths</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                            <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                            <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="nb">min</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">):</span>
                            <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                                <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                            <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                                <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="c1"># Get sums of previous and new internal branch lengths.</span>
                    <span class="n">previous_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
                    <span class="nb">sum</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">)</span>
                    <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
                    <span class="nb">sum</span><span class="p">(</span><span class="n">new_internal_branch_lengths</span><span class="p">)</span>

                    <span class="c1"># Indicate that a new tree was produced.</span>
                    <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="c1"># Compare the branch length ratios of the new tree to</span>
                    <span class="c1"># the current tree, and apply criteria to decide whether</span>
                    <span class="c1"># the new tree is better or not.</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">new_average_over_clades</span> <span class="o">&gt;</span> <span class="n">average_over_clades</span><span class="p">:</span>
                            <span class="c1"># Make sure that it does not make the branch length ratio</span>
                            <span class="c1"># worse for any clade.</span>
                            <span class="c1">#made_a_clade_worse = False</span>
                            <span class="c1">#for cladex in branch_length_info_dict.keys():</span>
                            <span class="c1">#    old_clade_ratio = branch_length_info_dict\</span>
                            <span class="c1">#                      [cladex][&#39;stem/branch ratio&#39;]</span>
                            <span class="c1">#    new_clade_ratio = new_branch_length_info_dict\</span>
                            <span class="c1">#                      [cladex][&#39;stem/branch ratio&#39;]</span>
                            <span class="c1">#    if new_clade_ratio &lt; old_clade_ratio:</span>
                            <span class="c1">#        made_a_clade_worse = True</span>
                            <span class="c1">#        break</span>
                            <span class="c1">#if not made_a_clade_worse:</span>
                            <span class="c1">#    print(&#39;\tnew tree is better&#39;)</span>
                            <span class="c1">#    new_tree_better = True</span>

                            <span class="c1"># Make sure that it did not make the ratio worse for</span>
                            <span class="c1"># the clade of interest (current worst clade).</span>
                            <span class="n">old_clade_ratio</span> <span class="o">=</span> <span class="n">branch_length_info_dict</span>\
                                              <span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                            <span class="n">new_clade_ratio</span> <span class="o">=</span> <span class="n">new_branch_length_info_dict</span>\
                                              <span class="p">[</span><span class="n">clade</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                            <span class="k">if</span> <span class="n">new_clade_ratio</span> <span class="o">&gt;=</span> <span class="n">old_clade_ratio</span><span class="p">:</span>
                                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                                <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>
                            
                            <span class="k">else</span><span class="p">:</span>
                                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is not better&#39;</span><span class="p">)</span>

                    <span class="c1"># Apply different criteria if the</span>
                    <span class="c1"># optimize_internal_branches option is selected.</span>
                    <span class="k">elif</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                        <span class="c1">#if no_new_internal_branches_are_worse:</span>
                        <span class="c1">#    if at_least_one_internal_branch_longer:</span>
                        <span class="c1">#        print(&#39;\tnew tree is better&#39;)</span>
                        <span class="c1">#        # Temp.</span>
                        <span class="c1">#        assert 2 != 2</span>
                        <span class="c1">#        new_tree_better = True</span>
                        <span class="k">if</span> <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">&gt;</span>\
                        <span class="n">previous_sum_of_internal_branch_lengths</span><span class="p">:</span>
                            <span class="k">if</span> <span class="ow">not</span> <span class="n">shortest_internal_branch_shorter</span><span class="p">:</span>
                                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                                <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>

                        <span class="c1">#if shortest_internal_branch_longer:</span>
                        <span class="c1">#    assert 2 != 2</span>
                        <span class="c1">#if shortest_internal_branch_shorter:</span>
                        <span class="c1">#    assert 2 != 2</span>



                    <span class="c1"># Add sequence to list of sequences which have been considered for</span>
                    <span class="c1"># removal.</span>
                    <span class="n">seqs_attempted_to_remove</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqname</span><span class="p">)</span>

                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># See if the next longest branch can be removed.</span>
                    <span class="k">pass</span>
            
            <span class="c1"># Break the loop so that the function can be run again.</span>
            <span class="k">if</span> <span class="n">new_tree_produced</span><span class="p">:</span>
                <span class="k">break</span>
        <span class="c1"># Break the loop so that the function can be run again.</span>
        <span class="k">if</span> <span class="n">new_tree_produced</span><span class="p">:</span>
            <span class="k">break</span>

    <span class="c1"># If no new tree could be produced, then return the final tree.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">new_tree_produced</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a sequence&#39;</span><span class="p">,</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">recursion_num</span><span class="p">,</span> <span class="n">t1</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">,</span>\
                <span class="n">seqs_attempted_to_remove</span><span class="p">,</span> <span class="n">branch_length_info_dict</span><span class="p">]</span>

    <span class="c1"># Otherwise run the function again.</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">new_tree_better</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a sequence&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on new tree, alignment ,and type sequences.</span>
            <span class="k">return</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">new_tree</span><span class="p">,</span>
                                           <span class="n">new_alignment</span><span class="p">,</span>
                                           <span class="n">new_subs_model</span><span class="p">,</span>
                                           <span class="n">new_type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">new_branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">True</span>
                                           <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Removed a sequence&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on same input (except with longer list of seqs that</span>
            <span class="c1"># were considered for removal).</span>
            <span class="k">return</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">t1</span><span class="p">,</span>
                                           <span class="n">alignment</span><span class="p">,</span>
                                           <span class="n">subs_model</span><span class="p">,</span>
                                           <span class="n">type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">True</span>
                                           <span class="p">)</span></div>


<div class="viewcode-block" id="recursively_add_seqs"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.recursively_add_seqs">[docs]</a><span class="k">def</span> <span class="nf">recursively_add_seqs</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">,</span>
                         <span class="n">t1</span><span class="p">,</span>
                         <span class="n">alignment</span><span class="p">,</span>
                         <span class="n">subs_model</span><span class="p">,</span>
                         <span class="n">type_seqs_dict</span><span class="p">,</span>
                         <span class="n">seqs_attempted_to_add</span><span class="p">,</span>
                         <span class="n">essential_taxa</span><span class="p">,</span>
                         <span class="n">branch_length_info_dict</span><span class="p">,</span>
                         <span class="n">all_sequences</span><span class="p">,</span>
                         <span class="n">log_file_path</span><span class="p">,</span>
                         <span class="n">optimize_internal_branches</span>
                         <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a tree, and try to add sequences to improve the ratio of the</span>
<span class="sd">    stem lengths of clades to the branch lengths of the branches in the clades,</span>
<span class="sd">    and return the final tree when all the sequences have been added to the</span>
<span class="sd">    dataset (at least temporarily).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define variable for whether any new tree could be produced.</span>
    <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variable for whether any new tree is better or not.</span>
    <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Define variables to store info for next recursion, if necessary.</span>
    <span class="n">new_tree</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_alignment</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">new_subs_model</span> <span class="o">=</span> <span class="n">subs_model</span>

    <span class="c1"># Define variables for writing to log file.</span>
    <span class="n">action</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="s1">&#39;No output tree&#39;</span>

    <span class="c1"># Determine average of stem length to branch length ratio over clades of</span>
    <span class="c1"># interest.</span>
    <span class="c1">#branch_length_info_dict = get_branch_length_info(t1, type_seqs_dict)</span>
    <span class="n">average_over_clades</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>


    <span class="c1"># Determine the clade with the worst ratio.</span>
    <span class="n">clades_by_ascending_ratio</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span>\
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">])</span>
    <span class="n">clade_with_lowest_ratio</span> <span class="o">=</span> <span class="n">clades_by_ascending_ratio</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Get a list of all sequence names for tree.</span>
    <span class="n">seq_names_in_tree</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">t1</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]</span>

    <span class="n">dblundrscr</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;__&#39;</span><span class="p">)</span>
    <span class="n">uniform_spaces</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">dblundrscr</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span>\
        <span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">))</span>

    <span class="c1"># Randomly pick a sequence that has not already been added, and does not</span>
    <span class="c1"># have the same name as any of the sequences already in the tree.</span>
    <span class="n">seq_to_add</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">shuffle</span><span class="p">(</span><span class="n">all_sequences</span><span class="p">)</span>
    <span class="n">all_potential_seqs_to_add</span> <span class="o">=</span> <span class="n">seqs_attempted_to_add</span> <span class="o">+</span> <span class="n">seq_names_in_tree</span>
    <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">all_sequences</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">uniform_spaces</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">uniform_spaces</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>\
                <span class="n">all_potential_seqs_to_add</span><span class="p">]:</span>
            <span class="n">seq_to_add</span> <span class="o">=</span> <span class="n">seq</span>
            <span class="k">break</span>

    <span class="k">if</span> <span class="n">seq_to_add</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Add to list of sequences added to to the tree.</span>
        <span class="n">seqs_attempted_to_add</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_to_add</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>

        <span class="c1"># Indicate that a new tree was produced.</span>
        <span class="n">new_tree_produced</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="c1"># Define logfile path to write to.</span>
        <span class="n">logfile</span> <span class="o">=</span> <span class="n">log_file_path</span>

        <span class="c1">#ali = os.path.join(outdirpath1, &#39;1.nex&#39;) </span>
        <span class="n">ali_num</span> <span class="o">=</span> <span class="n">recursion_num</span>
        <span class="n">outdirpath1</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
        <span class="n">ali</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;.nex&#39;</span>
        <span class="c1">#add_seq_to_alignment2(seq_to_add, alignment, ali)</span>
        <span class="n">add_seq_to_alignment3</span><span class="p">(</span><span class="n">seq_to_add</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">ali</span><span class="p">)</span>
        <span class="n">new_alignment</span> <span class="o">=</span> <span class="n">ali</span>

        <span class="c1"># Trim the alignment (any positions where only the new sequence has</span>
        <span class="c1"># residues). ????</span>

        <span class="c1"># Write input tree to a file.</span>
        <span class="n">tree</span> <span class="o">=</span> <span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_original_newick.tre&#39;</span>
        <span class="n">t1</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">outfile</span><span class="o">=</span><span class="n">tree</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Convert nexus alignment to afa.</span>
        <span class="n">temp_fa_1</span> <span class="o">=</span> <span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_temp1.afa&#39;</span>
        <span class="n">nex_to_afa</span><span class="p">(</span><span class="n">ali</span><span class="p">,</span> <span class="n">temp_fa_1</span><span class="p">)</span> 

        <span class="c1"># Code names in alignment.</span>
        <span class="n">outalifp</span> <span class="o">=</span> <span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.afa&#39;</span>
        <span class="n">outtablefp</span> <span class="o">=</span> <span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.table&#39;</span>
        <span class="n">code_names_in_ali</span><span class="p">(</span><span class="n">outdirpath1</span><span class="p">,</span> <span class="n">temp_fa_1</span><span class="p">,</span> <span class="n">outalifp</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span> 

        <span class="c1"># Convert alignment back to nexus.</span>
        <span class="n">outalifpnex</span> <span class="o">=</span> <span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.nex&#39;</span>
        <span class="n">afa_to_nex</span><span class="p">(</span><span class="n">outalifp</span><span class="p">,</span> <span class="n">outalifpnex</span><span class="p">)</span>

        <span class="c1"># Delete temporary files.</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">temp_fa_1</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">outalifp</span><span class="p">)</span>

        <span class="c1"># Make constraint tree.</span>
        <span class="c1"># Put quotation marks around names in constraint tree.</span>
        <span class="c1">#outtreefp = ali.rsplit(&#39;.&#39;, 1)[0] + &#39;.constraint_tree0.Q.newick&#39;</span>
        <span class="n">constraint_tree_fp</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">ali</span><span class="p">),</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_constraint_tree.newick&#39;</span><span class="p">))</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">tree</span><span class="p">)</span> <span class="k">as</span> <span class="n">intreefh</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">constraint_tree_fp</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">intreefh</span><span class="p">:</span>
                <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">quote_tree</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">))</span>
        <span class="c1"># Make constraint tree with coded names.</span>
        <span class="n">constraint_tree_fp_coded</span> <span class="o">=</span> <span class="n">constraint_tree_fp</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.C.tre&#39;</span>
        <span class="n">code_tree</span><span class="p">(</span><span class="n">constraint_tree_fp</span><span class="p">,</span> <span class="n">constraint_tree_fp_coded</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span>


        <span class="c1"># Run an ML search with the alignment (with additional sequence) and</span>
        <span class="c1"># constraint tree defined above.</span>

        <span class="c1"># Make a copy of the alignment in phylip format for input to IQtree.</span>
        <span class="n">phy_out</span> <span class="o">=</span> <span class="n">outalifpnex</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.phy&#39;</span>
        <span class="n">nex_to_phylip</span><span class="p">(</span><span class="n">outalifpnex</span><span class="p">,</span> <span class="n">phy_out</span><span class="p">)</span>

        <span class="c1"># Remove nexus format version of alignment, because it is redundant with</span>
        <span class="c1"># the phylip version.</span>
        <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">outalifpnex</span><span class="p">)</span>

        <span class="c1"># Do phylogenetic analysis. # SEPARATE</span>
        <span class="c1"># Make subdir for output.</span>
        <span class="n">subdirp</span> <span class="o">=</span> <span class="n">constraint_tree_fp</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_IQ-tree_output&#39;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">subdirp</span><span class="p">)</span>

        <span class="c1"># ***Use the get_ml_tree_branch_lengths function here to run IQtree</span>
        <span class="c1"># instead???</span>

        <span class="c1"># Use IQtree to do an ML search</span>
        <span class="n">output_file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">subdirp</span><span class="p">,</span> <span class="s1">&#39;iqtree&#39;</span><span class="p">)</span>
        <span class="n">stdout_path</span> <span class="o">=</span> <span class="n">output_file_prefix</span> <span class="o">+</span> <span class="s1">&#39;.stdout.txt&#39;</span>
        <span class="n">iqtree_command_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;iqtree&#39;</span><span class="p">,</span> <span class="s1">&#39;-s&#39;</span><span class="p">,</span> <span class="n">phy_out</span><span class="p">,</span> <span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="s1">&#39;-g&#39;</span><span class="p">,</span>
            <span class="n">constraint_tree_fp_coded</span><span class="p">,</span> <span class="s1">&#39;-pre&#39;</span><span class="p">,</span> <span class="n">output_file_prefix</span><span class="p">,</span> <span class="s1">&#39;-nt&#39;</span><span class="p">,</span> <span class="s1">&#39;4&#39;</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">stdout_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">iqtree_command_list</span><span class="p">,</span> <span class="n">stdout</span><span class="o">=</span><span class="n">o</span><span class="p">,</span> <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">STDOUT</span><span class="p">)</span>

        <span class="c1"># Define path to output file with tree.</span>
        <span class="n">tree_file_path</span> <span class="o">=</span> <span class="n">output_file_prefix</span> <span class="o">+</span> <span class="s1">&#39;.treefile&#39;</span>
        <span class="c1"># Check that the output file was actually produced.</span>
        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">tree_file_path</span><span class="p">)</span>

        <span class="c1"># Parse the resulting tree to determine in which clade of interest, if any,</span>
        <span class="c1"># the sequence was placed.</span>

        <span class="c1"># Parse tree using ete3.</span>
        <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>
        <span class="n">t2</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree_file_path</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Convert names in tree back to original names.</span>
        <span class="n">uncode_tree_obj</span><span class="p">(</span><span class="n">t2</span><span class="p">,</span> <span class="n">outtablefp</span><span class="p">)</span>

        <span class="c1"># Define new tree object for further recursions of this function.</span>
        <span class="n">new_tree</span> <span class="o">=</span> <span class="n">t2</span>

        <span class="c1">## Define name for tree without branch lengths.</span>
        <span class="c1">#simple_tree = tree_file_path.rsplit(&#39;_&#39;, 1)[0] + &#39;_TEMP2&#39; </span>

        <span class="c1">## Write simple tree to a new file </span>
        <span class="c1">#t1.write(format=9, outfile=simple_tree)</span>

        <span class="c1">## Parse simple tree.</span>
        <span class="c1">#t2 = Tree(simple_tree)</span>

        <span class="c1"># Print simple tree.</span>
        <span class="c1">#print(&#39;ML tree:&#39;)</span>
        <span class="c1">#print(t2)</span>

        <span class="c1"># Make a copy of the tree object.</span>
        <span class="n">t3</span> <span class="o">=</span> <span class="n">t2</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1"># Get list of &quot;type&quot; sequences from input.</span>
        <span class="n">type_seq_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="c1"># Determine whether the additional sequence was placed in one of the</span>
        <span class="c1"># clades of interest, or not.</span>

        <span class="c1"># For each &quot;type&quot; sequence, traverse all nodes and find the node with</span>
        <span class="c1"># the largest number of child nodes that are leaf (terminal) nodes,</span>
        <span class="c1"># containing the &quot;type&quot; sequence of interest, but not containing any of</span>
        <span class="c1"># the other &quot;type&quot; sequences.</span>
        <span class="n">ts_num</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">first_type_seq_node_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">ts_that_additional_seq_was_placed_in</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">ts</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
            <span class="n">ts_num</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">ts_num</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">first_type_seq_node_name</span> <span class="o">=</span> <span class="n">ts</span>
                <span class="c1"># Root on another &quot;type&quot; sequence for the first type sequence in</span>
                <span class="c1"># the list to get whole clade, then root the tree on the ancestor</span>
                <span class="c1"># node of that first clade.</span>

                <span class="c1"># Get a node name for a node corresponding to a different</span>
                <span class="c1"># &quot;type&quot; sequence.</span>
                <span class="n">other_type_seq_node_name</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="n">ts</span><span class="p">:</span>
                        <span class="n">other_type_seq_node_name</span> <span class="o">=</span> <span class="n">i</span>
                        <span class="k">break</span>

                <span class="c1"># Get node corresponding to a different &quot;type&quot; sequence.</span>
                <span class="n">other_type_seq_node</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">other_type_seq_node_name</span><span class="p">:</span>
                        <span class="n">other_type_seq_node</span> <span class="o">=</span> <span class="n">node</span>
                        <span class="k">break</span>

                <span class="c1"># Root on the other &quot;type&quot; sequence node.</span>
                <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">other_type_seq_node</span><span class="p">)</span>

                <span class="c1">#print(&#39;\n\n\nTree rooted on a type sequence other than the first type sequence.&#39;)</span>
                <span class="c1">#print(t3)</span>

            <span class="k">elif</span> <span class="n">ts_num</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                <span class="c1"># Root on the first &quot;type&quot; sequence node for all subsequent</span>
                <span class="c1"># clades.</span>
                <span class="n">first_type_seq_node</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t3</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                    <span class="n">leaf_list</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">leaf</span> <span class="ow">in</span> <span class="n">node</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">leaf</span><span class="o">.</span><span class="n">name</span> <span class="o">==</span> <span class="n">first_type_seq_node_name</span><span class="p">:</span>
                            <span class="n">first_type_seq_node</span> <span class="o">=</span> <span class="n">node</span>
                            <span class="k">break</span>
                <span class="n">t3</span><span class="o">.</span><span class="n">set_outgroup</span><span class="p">(</span><span class="n">first_type_seq_node</span><span class="p">)</span>
                <span class="c1">#print(&#39;\n\n\nTree re-rooted on first type sequence:&#39;)</span>
                <span class="c1">#print(t3)</span>


            <span class="c1"># Make a copy of the tree topology to work with for each run</span>
            <span class="c1"># through this loop.</span>
            <span class="n">t4</span> <span class="o">=</span> <span class="n">t3</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

            <span class="c1"># Make a list of nodes that contain type seq, but not any others. #</span>
            <span class="c1"># SEPARATE</span>
            <span class="n">nodes_of_interest</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">node</span> <span class="ow">in</span> <span class="n">t4</span><span class="o">.</span><span class="n">traverse</span><span class="p">():</span>
                <span class="c1"># Search in nodes that contain the type sequence.</span>
                <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ts</span><span class="p">):</span>
                    <span class="c1"># Search in nodes that don&#39;t contain other type sequences.</span>
                    <span class="n">contains_other_type_seqs</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">for</span> <span class="n">ts2</span> <span class="ow">in</span> <span class="n">type_seq_list</span><span class="p">:</span>
                        <span class="k">if</span> <span class="ow">not</span> <span class="n">ts2</span> <span class="o">==</span> <span class="n">ts</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">node</span><span class="o">.</span><span class="n">search_nodes</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">ts2</span><span class="p">):</span>
                                <span class="n">contains_other_type_seqs</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">contains_other_type_seqs</span><span class="p">:</span>
                        <span class="c1"># Add nodes of interest to list.</span>
                        <span class="n">nodes_of_interest</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

            <span class="c1"># find the node with the most child leaf nodes.</span>
            <span class="c1">#node_num = 0</span>
            <span class="c1">#for node in nodes_of_interest:</span>
            <span class="c1">#    node_num += 1 </span>
            <span class="c1">#    print(&#39;Node &#39; + str(node_num) + &#39; Number of leaves:  &#39; + str(len(node.get_leaves())))</span>
            <span class="c1">#    print(node)</span>
            <span class="n">node_w_most_leaves</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">nodes_of_interest</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>\
                    <span class="nb">len</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="c1">#node_w_most_leaves.name = &#39;X&#39;</span>
            <span class="c1">#print(&#39;\n\nClade of interest defined by sequence &#39; + ts + &#39;:&#39;)</span>
            <span class="c1">#print(node_w_most_leaves)</span>

            <span class="n">name_of_additional_seq</span> <span class="o">=</span> <span class="n">seq_to_add</span><span class="o">.</span><span class="n">description</span> <span class="c1"># should be .id ?</span>
            <span class="k">if</span> <span class="n">name_of_additional_seq</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">node_w_most_leaves</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]:</span>
                <span class="n">ts_that_additional_seq_was_placed_in</span> <span class="o">=</span> <span class="n">ts</span>
                <span class="c1">#print(&#39;\n\n\n%s is in this clade!!!!\n\n\n&#39; % name_of_additional_seq)</span>
                <span class="k">break</span>

        <span class="c1"># Check that the clade that the additional sequence was placed in was</span>
        <span class="c1"># identified.</span>
        <span class="c1">#assert ts_that_additional_seq_was_placed_in is not None, &quot;&quot;&quot;Sequence was</span>
        <span class="c1">#not placed in any of the clades of interest: %s&quot;&quot;&quot; % record.description</span>

        <span class="c1"># Define output path for annotated tree image file.</span>
        <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
        <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="n">annotated_tree_outpath</span>
        <span class="c1"># Extract info about branch lengths from tree, and write</span>
        <span class="c1"># annotated tree image file.</span>
        <span class="n">new_branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t2</span><span class="p">,</span>
                                      <span class="n">type_seqs_dict</span><span class="p">,</span>
                                      <span class="n">annotated_tree_outpath</span>
                                      <span class="p">)</span>

        <span class="c1"># Write a file with type sequences and clade names for this</span>
        <span class="c1"># iteration.</span>
        <span class="n">new_type_seqs_file</span><span class="o">=</span>\
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_type_seqs.csv&#39;</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">new_type_seqs_file</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> <span class="o">+</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>


        <span class="c1"># If the sequence was placed in one of the clades of interest, then</span>
        <span class="c1"># proceed with collecting further information. If not, then assume it</span>
        <span class="c1"># is of no use, and move on to the next sequence.</span>
        <span class="k">if</span> <span class="n">ts_that_additional_seq_was_placed_in</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># Get clade name.</span>
            <span class="n">clade_that_seq_was_placed_in</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">type_seqs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">==</span> <span class="n">ts_that_additional_seq_was_placed_in</span><span class="p">:</span>
                    <span class="n">clade_that_seq_was_placed_in</span> <span class="o">=</span> <span class="n">c</span>
            <span class="k">assert</span> <span class="n">clade_that_seq_was_placed_in</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;Could not find</span>
<span class="s2">            corresponding clade for sequence </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">ts_that_additional_seq_was_placed_in</span>

            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Sequence placed in clade: &#39;</span> <span class="o">+</span>\
                    <span class="n">clade_that_seq_was_placed_in</span><span class="p">)</span>

            <span class="c1"># Get branch length info for new tree.</span>

            <span class="c1"># Define output path for annotated tree image file.</span>
            <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ali</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
            <span class="n">annot_tree_path</span> <span class="o">=</span> <span class="n">annotated_tree_outpath</span>
            <span class="c1"># Extract info about branch lengths from tree, and write</span>
            <span class="c1"># annotated tree image file.</span>
            <span class="n">new_branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t2</span><span class="p">,</span>
                                          <span class="n">type_seqs_dict</span><span class="p">,</span>
                                          <span class="n">annotated_tree_outpath</span>
                                          <span class="p">)</span>
            <span class="c1"># Calculate average ratio over all clades from branch</span>
            <span class="c1"># length info. </span>
            <span class="n">new_average_over_clades</span> <span class="o">=</span>\
                    <span class="n">statistics</span><span class="o">.</span><span class="n">mean</span><span class="p">([</span><span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>\
                    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>

            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">New average ratio over clades: &#39;</span> <span class="o">+</span>\
                    <span class="nb">str</span><span class="p">(</span><span class="n">new_average_over_clades</span><span class="p">))</span>

            <span class="c1"># Compare lengths of internal branches.</span>
            <span class="n">previous_internal_branch_lengths</span> <span class="o">=</span>\
            <span class="n">branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
            <span class="n">new_internal_branch_lengths</span> <span class="o">=</span>\
            <span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">branch_length_info_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;internal branch lengths&#39;</span><span class="p">]</span>
            <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">,</span> <span class="n">new_internal_branch_lengths</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                    <span class="n">no_new_internal_branches_are_worse</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                    <span class="n">at_least_one_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="nb">min</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">j</span> <span class="o">&gt;</span> <span class="n">i</span><span class="p">:</span>
                        <span class="n">shortest_internal_branch_longer</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">if</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                        <span class="n">shortest_internal_branch_shorter</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="c1"># Get sums of previous and new internal branch lengths.</span>
            <span class="n">previous_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
            <span class="nb">sum</span><span class="p">(</span><span class="n">previous_internal_branch_lengths</span><span class="p">)</span>
            <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">=</span>\
            <span class="nb">sum</span><span class="p">(</span><span class="n">new_internal_branch_lengths</span><span class="p">)</span>


            <span class="c1"># Compare branch length ratios between the new tree and the old tree.</span>
            <span class="c1"># Including overall average, specifically for the worst clade, and</span>
            <span class="c1"># specifically for the clade that the new sequence was placed into.</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">new_average_over_clades</span> <span class="o">&gt;=</span> <span class="n">average_over_clades</span><span class="p">:</span>
                    <span class="c1"># Compare for the clade that the sequence was added to.</span>
                    <span class="n">old_clade_ratio</span> <span class="o">=</span> <span class="n">branch_length_info_dict</span>\
                                      <span class="p">[</span><span class="n">clade_that_seq_was_placed_in</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                    <span class="n">new_clade_ratio</span> <span class="o">=</span> <span class="n">new_branch_length_info_dict</span>\
                                      <span class="p">[</span><span class="n">clade_that_seq_was_placed_in</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                    <span class="c1">#print(&#39;\tnew ratio for clade %s: &#39; + str(new_clade_ratio) % clade_that_seq_was_placed_in)</span>
                    <span class="k">if</span> <span class="n">new_clade_ratio</span> <span class="o">&gt;=</span> <span class="n">old_clade_ratio</span><span class="p">:</span>
                        <span class="c1">#print(&#39;\tnew tree is better&#39;)</span>
                        <span class="c1">#new_tree_better = True</span>

                        <span class="c1"># Compare for the clade that was the worst clade in the</span>
                        <span class="c1"># original tree.</span>
                        <span class="n">old_worst_clade_ratio</span> <span class="o">=</span> <span class="n">branch_length_info_dict</span>\
                                          <span class="p">[</span><span class="n">clade_with_lowest_ratio</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                        <span class="n">new_worst_clade_ratio</span> <span class="o">=</span> <span class="n">new_branch_length_info_dict</span>\
                                          <span class="p">[</span><span class="n">clade_with_lowest_ratio</span><span class="p">][</span><span class="s1">&#39;stem/branch ratio&#39;</span><span class="p">]</span>
                        <span class="c1">#print(&#39;\tnew ratio for worst clade (%s): &#39; +\</span>
                        <span class="c1">#        str(new_worst_clade_ratio) %\</span>
                        <span class="c1">#        clade_with_lowest_ratio)</span>
                        <span class="k">if</span> <span class="n">new_worst_clade_ratio</span> <span class="o">&gt;=</span> <span class="n">old_worst_clade_ratio</span><span class="p">:</span>
                            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                            <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="c1"># Apply different criteria if optimizing for internal branches.</span>
            <span class="k">elif</span> <span class="n">optimize_internal_branches</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">new_sum_of_internal_branch_lengths</span> <span class="o">&gt;</span>\
                <span class="n">previous_sum_of_internal_branch_lengths</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">shortest_internal_branch_shorter</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is better&#39;</span><span class="p">)</span>
                        <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>


            <span class="c1"># Make a list of taxonomic terms currently represented in the clade.</span>
            <span class="n">taxa_represented</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">seqname</span> <span class="ow">in</span>\
            <span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="n">clade_that_seq_was_placed_in</span><span class="p">][</span><span class="s1">&#39;seq names by descending length&#39;</span><span class="p">]:</span>
                <span class="c1"># Get species name from sequence name.</span>
                <span class="n">species_name</span> <span class="o">=</span> <span class="n">seqname</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="c1"># Get taxonomic information for species name from info csv file.</span>
                <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
                <span class="c1"># Add taxon names to taxa represented.</span>
                <span class="n">taxa_represented</span> <span class="o">=</span> <span class="n">taxa_represented</span>\
                        <span class="o">+</span> <span class="p">[</span><span class="n">ti</span><span class="o">.</span><span class="n">superbranch</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">supergroup</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">group</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">species</span><span class="p">]</span>

            <span class="c1"># Determine which seqs in the clade are dispensable, considering their</span>
            <span class="c1"># taxonomic placement.</span>
            <span class="n">essential_seq_names</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">seqname</span> <span class="ow">in</span>\
            <span class="n">new_branch_length_info_dict</span><span class="p">[</span><span class="n">clade_that_seq_was_placed_in</span><span class="p">][</span><span class="s1">&#39;seq names by descending length&#39;</span><span class="p">]:</span>
                <span class="c1"># Get species name from sequence name.</span>
                <span class="n">species_name</span> <span class="o">=</span> <span class="n">seqname</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="c1"># Get taxonomic information for species name from info csv file.</span>
                <span class="n">ti</span> <span class="o">=</span> <span class="n">get_taxonomic_info</span><span class="p">(</span><span class="n">species_name</span><span class="p">)</span>
                <span class="n">ti_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">ti</span><span class="o">.</span><span class="n">superbranch</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">supergroup</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">group</span><span class="p">,</span> <span class="n">ti</span><span class="o">.</span><span class="n">species</span><span class="p">]</span>
                <span class="c1"># Determine whether any of the taxonomy terms are in the essential</span>
                <span class="c1"># list.</span>
                <span class="n">seq_is_essential</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">essential_terms</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ti_list</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">essential_taxa</span><span class="p">:</span>
                        <span class="n">essential_terms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
                <span class="c1"># Check whether any of the terms appear in the taxa_represented list</span>
                <span class="c1"># only once (if so then the sequence is essential).</span>
                <span class="k">for</span> <span class="n">term</span> <span class="ow">in</span> <span class="n">essential_terms</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">taxa_represented</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">term</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">seq_is_essential</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="c1"># Add to list of essential seq names.</span>
                <span class="k">if</span> <span class="n">seq_is_essential</span><span class="p">:</span>
                    <span class="n">essential_seq_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seqname</span><span class="p">)</span>

            <span class="c1"># If the added sequence is in the list of essential taxa, then</span>
            <span class="c1"># assume the tree is better.</span>
            <span class="k">if</span> <span class="n">name_of_additional_seq</span> <span class="ow">in</span> <span class="n">essential_seq_names</span><span class="p">:</span>
                <span class="n">new_tree_better</span> <span class="o">=</span> <span class="kc">True</span>


            <span class="k">if</span> <span class="ow">not</span> <span class="n">new_tree_better</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">new tree is not better&#39;</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># The additional sequence was not placed in a clade of interest</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Sequence was not placed in a clade of interest.&#39;</span><span class="p">)</span>

    <span class="c1"># If no new tree could be produced, then return the final tree.</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">new_tree_produced</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;No sequences to add&#39;</span><span class="p">,</span>\
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">[</span><span class="n">recursion_num</span><span class="p">,</span> <span class="n">t1</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">type_seqs_dict</span><span class="p">,</span>\
                <span class="n">seqs_attempted_to_add</span><span class="p">,</span> <span class="n">branch_length_info_dict</span><span class="p">]</span>

    <span class="c1"># Otherwise run the function again.</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">new_tree_better</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Added a sequence&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on new tree and alignment.</span>
            <span class="k">return</span> <span class="n">recursively_add_seqs</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">new_tree</span><span class="p">,</span>
                                           <span class="n">new_alignment</span><span class="p">,</span>
                                           <span class="n">new_subs_model</span><span class="p">,</span>
                                           <span class="n">type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">seqs_attempted_to_add</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">new_branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">all_sequences</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="n">optimize_internal_branches</span>
                                           <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">log</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">recursion_num</span><span class="p">),</span> <span class="s1">&#39;Added a sequence&#39;</span><span class="p">,</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annot_tree_path</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1"># Run again on same input (except with longer list of seqs that</span>
            <span class="c1"># were considered for addition).</span>
            <span class="k">return</span> <span class="n">recursively_add_seqs</span><span class="p">(</span><span class="n">recursion_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span>
                                           <span class="n">t1</span><span class="p">,</span>
                                           <span class="n">alignment</span><span class="p">,</span>
                                           <span class="n">subs_model</span><span class="p">,</span>
                                           <span class="n">type_seqs_dict</span><span class="p">,</span>
                                           <span class="n">seqs_attempted_to_add</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                           <span class="n">all_sequences</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="n">optimize_internal_branches</span>
                                           <span class="p">)</span></div>


<div class="viewcode-block" id="optimize_sequence_selection2"><a class="viewcode-back" href="../module_amoebae_select_seqs.html#module_amoebae_select_seqs.optimize_sequence_selection2">[docs]</a><span class="k">def</span> <span class="nf">optimize_sequence_selection2</span><span class="p">(</span><span class="n">file_with_seqs</span><span class="p">,</span> <span class="n">model_name</span><span class="p">,</span> <span class="n">essential_taxa</span><span class="p">,</span>
        <span class="n">timestamp</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Add/remove homologous sequences from a dataset to optimize for long stem</span>
<span class="sd">    lengths and low average branch lengths for clades of interest.</span>

<span class="sd">    The essential_taxa argument is a list of names corresponding to taxonomic</span>
<span class="sd">    groups of eukaryotes (such as &quot;Amorphea&quot; or &quot;Diaphoretickes&quot;) which appear</span>
<span class="sd">    in the genome info csv file in the genomes directory.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Increase recursion limit for python. Potentially dangerous, but necessary for</span>
    <span class="c1"># large alignments, like for Giantin and lava lamp proteins!</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">setrecursionlimit</span><span class="p">(</span><span class="mi">10000</span><span class="p">)</span>

    <span class="c1"># Parse data from model.</span>
    <span class="c1"># Get info about model that is relevant.</span>
    <span class="n">model_info</span> <span class="o">=</span> <span class="n">ModelInfoFromCSV</span><span class="p">(</span><span class="n">model_name</span><span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">alignment_file</span>
    <span class="n">tree</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">tree_topology_file</span>
    <span class="n">subs_model</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">subs_model</span>
    <span class="n">type_seqs</span> <span class="o">=</span> <span class="n">model_info</span><span class="o">.</span><span class="n">type_seqs_file</span>

    <span class="c1"># Parse tree using ete3.</span>
    <span class="c1"># Note: parentheses and commas get replaced with underscores.</span>
    <span class="n">t1</span> <span class="o">=</span> <span class="n">Tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">quoted_node_names</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="c1"># Print tree.</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Input tree:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">t1</span><span class="p">)</span>

    <span class="c1"># Define output directory path.</span>
    <span class="n">outputdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">file_with_seqs</span><span class="p">),</span>\
            <span class="s1">&#39;select_seqs_output_&#39;</span> <span class="o">+</span> <span class="n">timestamp</span><span class="p">)</span>
    <span class="c1"># Make output directory.</span>
    <span class="k">assert</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">outputdir</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">outputdir</span><span class="p">)</span>

    <span class="c1"># Construct a dictionary with info from type_seqs_file.</span>
    <span class="n">type_seqs_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">type_seqs</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">infh</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">):</span>
                <span class="n">spliti</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
                <span class="n">seqname</span> <span class="o">=</span> <span class="n">spliti</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">cladename</span> <span class="o">=</span> <span class="n">spliti</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">type_seqs_dict</span><span class="p">[</span><span class="n">cladename</span><span class="p">]</span> <span class="o">=</span> <span class="n">seqname</span>

    <span class="c1"># Copy alignment file to new path.</span>
    <span class="n">new_ali_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_0.nex&#39;</span><span class="p">)</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">new_ali_path</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">new_ali_path</span><span class="p">)</span>
    <span class="n">alignment</span> <span class="o">=</span> <span class="n">new_ali_path</span> 

    <span class="c1"># Run ML search on model tree to get branch lengths.</span>
    <span class="n">t2</span> <span class="o">=</span> <span class="n">get_ml_tree_branch_lengths</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">subs_model</span><span class="p">,</span> <span class="n">outputdir</span><span class="p">)</span>

    <span class="c1"># Define output path for annotated tree image file.</span>
    <span class="n">annotated_tree_outpath</span> <span class="o">=</span>\
    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_annotated_result.png&#39;</span><span class="p">)</span>
    <span class="c1"># Extract info about branch lengths from tree, and write</span>
    <span class="c1"># annotated tree image file.</span>
    <span class="n">branch_length_info_dict</span> <span class="o">=</span> <span class="n">get_branch_length_info</span><span class="p">(</span><span class="n">t2</span><span class="p">,</span>
                              <span class="n">type_seqs_dict</span><span class="p">,</span>
                              <span class="n">annotated_tree_outpath</span><span class="p">)</span>

    <span class="c1"># Compile list of sequences represented in the input initial tree.</span>
    <span class="n">names_of_seqs_in_tree</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">t2</span><span class="o">.</span><span class="n">get_leaves</span><span class="p">()]</span>

    <span class="c1"># Parse input sequence file, and construct a list of sequence objects for</span>
    <span class="c1"># potential inclusion in the tree.</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">file_with_seqs</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
        <span class="n">seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

    <span class="c1"># Define path to log file (specifically for select_seqs).</span>
    <span class="n">log_file_path</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_select_seqs_log.txt&#39;</span>
    <span class="n">header_line</span> <span class="o">=</span> <span class="s2">&quot;Step number, Action, Output file</span><span class="se">\n</span><span class="s2">&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header_line</span><span class="p">)</span>
        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;Initial version&#39;</span><span class="p">,</span>\
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">annotated_tree_outpath</span><span class="p">)])</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>


    <span class="c1">########</span>
    <span class="c1">## Load information into a recursive or iterative function/loop to remove</span>
    <span class="c1">## sequences.</span>
    <span class="n">seqs_attempted_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">best_tree_info</span> <span class="o">=</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span>
                                             <span class="n">t2</span><span class="p">,</span>
                                             <span class="n">alignment</span><span class="p">,</span>
                                             <span class="n">subs_model</span><span class="p">,</span>
                                             <span class="n">type_seqs_dict</span><span class="p">,</span>
                                             <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                             <span class="n">essential_taxa</span><span class="p">,</span>
                                             <span class="n">branch_length_info_dict</span><span class="p">,</span>
                                             <span class="n">log_file_path</span><span class="p">,</span>
                                             <span class="kc">False</span>
                                             <span class="p">)</span>
    <span class="c1">##(returns: [recursion_num, t1, alignment, subs_model, type_seqs_dict,\</span>
    <span class="c1">##        seqs_attempted_to_remove, branch_length_info_dict] )</span>

    <span class="c1">## Try to remove even more sequences.</span>
    <span class="c1">#seqs_attempted_to_remove = []</span>
    <span class="n">best_tree_info</span> <span class="o">=</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="n">best_tree_info</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                              <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                              <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                              <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                              <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                                              <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                              <span class="n">essential_taxa</span><span class="p">,</span>
                                              <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                                              <span class="n">log_file_path</span><span class="p">,</span>
                                              <span class="kc">True</span>
                                              <span class="p">)</span>
    <span class="c1">#(returns: [recursion_num, t1, alignment, subs_model, type_seqs_dict,\</span>
    <span class="c1">#        seqs_attempted_to_remove, branch_length_info_dict] )</span>


    <span class="c1">## Now that all the non-essential sequences have been removed, try adding</span>
    <span class="c1">## sequences.</span>
    <span class="n">seqs_attempted_to_add</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">best_tree_info1</span> <span class="o">=</span> <span class="n">recursively_add_seqs</span><span class="p">(</span><span class="n">best_tree_info</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                           <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                           <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                           <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                           <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                                           <span class="n">seqs_attempted_to_add</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">best_tree_info</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                                           <span class="n">seqs</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">False</span>
                                           <span class="p">)</span>
    <span class="n">seqs_attempted_to_add</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">best_tree_info1</span> <span class="o">=</span> <span class="n">recursively_add_seqs</span><span class="p">(</span><span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                           <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                           <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                           <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                           <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                                           <span class="n">seqs_attempted_to_add</span><span class="p">,</span>
                                           <span class="n">essential_taxa</span><span class="p">,</span>
                                           <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                                           <span class="n">seqs</span><span class="p">,</span>
                                           <span class="n">log_file_path</span><span class="p">,</span>
                                           <span class="kc">True</span>
                                           <span class="p">)</span>

    <span class="c1">#seqs_attempted_to_add = []</span>
    <span class="c1">#best_tree_info1 = recursively_add_seqs(1,</span>
    <span class="c1">#                                       t2,</span>
    <span class="c1">#                                       alignment,</span>
    <span class="c1">#                                       subs_model,</span>
    <span class="c1">#                                       type_seqs_dict,</span>
    <span class="c1">#                                       seqs_attempted_to_add,</span>
    <span class="c1">#                                       essential_taxa,</span>
    <span class="c1">#                                       branch_length_info_dict,</span>
    <span class="c1">#                                       seqs,</span>
    <span class="c1">#                                       log_file_path</span>
    <span class="c1">#                                       )</span>

    <span class="c1">#return [recursion_num, t1, alignment, subs_model, type_seqs_dict,\</span>
    <span class="c1">#        seqs_attempted_to_add, branch_length_info_dict]</span>

    <span class="n">seqs_attempted_to_remove</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">best_tree_info2</span> <span class="o">=</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                             <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                             <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                             <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                             <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                                             <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                             <span class="n">essential_taxa</span><span class="p">,</span>
                                             <span class="n">best_tree_info1</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                                             <span class="n">log_file_path</span><span class="p">,</span>
                                             <span class="kc">False</span>
                                             <span class="p">)</span>
    <span class="c1">##(returns: [recursion_num, t1, alignment, subs_model, type_seqs_dict,\</span>
    <span class="c1">##        seqs_attempted_to_remove, branch_length_info_dict] )</span>

    <span class="n">best_tree_info3</span> <span class="o">=</span> <span class="n">recursively_remove_seqs</span><span class="p">(</span><span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                              <span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                              <span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                              <span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                              <span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span>
                                              <span class="n">seqs_attempted_to_remove</span><span class="p">,</span>
                                              <span class="n">essential_taxa</span><span class="p">,</span>
                                              <span class="n">best_tree_info2</span><span class="p">[</span><span class="mi">6</span><span class="p">],</span>
                                              <span class="n">log_file_path</span><span class="p">,</span>
                                              <span class="kc">True</span>
                                              <span class="p">)</span>

    <span class="k">return</span> <span class="s1">&#39;[main output path here]&#39;</span></div>








</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>