
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>replace_seqs &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for replace_seqs</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># Copyright 2018 Lael D. Barlow</span>
<span class="c1"># </span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1"># </span>
<span class="c1">#     http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1"># </span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="c1"># </span>
<span class="sd">&quot;&quot;&quot;Module for amoebae script with functions for the replace_seqs command.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="c1"># Import built-in modules.</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">copy</span>

<span class="c1"># Import major package modules.</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Import amoebae modules.</span>
<span class="kn">import</span> <span class="nn">settings</span>
<span class="kn">from</span> <span class="nn">module_afa_to_nex</span> <span class="k">import</span> <span class="n">nex_to_afa</span><span class="p">,</span> <span class="n">afa_to_nex</span><span class="p">,</span> <span class="n">determine_alphabet</span>
<span class="kn">from</span> <span class="nn">module_add_to_db</span> <span class="k">import</span> <span class="n">make_blast_db</span>
<span class="kn">from</span> <span class="nn">misc_functions</span> <span class="k">import</span> <span class="n">get_fa_record_text_from_obj</span>
<span class="kn">from</span> <span class="nn">module_paralogue_counter</span> <span class="k">import</span> <span class="n">add_seq_to_alignment3</span>
<span class="kn">from</span> <span class="nn">module_align_to_profile_iter</span> <span class="k">import</span> <span class="n">do_align_iteratively</span>
<span class="kn">from</span> <span class="nn">module_mask_nex</span> <span class="k">import</span> <span class="n">mask_nex</span>
<span class="kn">from</span> <span class="nn">module_amoebae_trim_nex</span> <span class="k">import</span> <span class="n">trim_nex</span>
<span class="kn">from</span> <span class="nn">module_afa_to_nex</span> <span class="k">import</span> <span class="n">delete_extra_mesquite_lines</span>
<span class="kn">from</span> <span class="nn">module_amoebae</span> <span class="k">import</span> <span class="n">get_seqs_from_fasta_db</span>

<span class="c1"># Import functions for working with sequences, alignments, and trees.</span>
<span class="c1">#from ete3 import Tree, TreeStyle, Tree, TextFace, add_face_to_node</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SearchIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">AlignIO</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span><span class="p">,</span> <span class="n">Gapped</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="k">import</span> <span class="n">MultipleSeqAlignment</span>


<span class="c1"># Define functions to be used in amoebae.</span>

<div class="viewcode-block" id="afa_to_fa"><a class="viewcode-back" href="../replace_seqs.html#replace_seqs.afa_to_fa">[docs]</a><span class="k">def</span> <span class="nf">afa_to_fa</span><span class="p">(</span><span class="n">infilepath</span><span class="p">,</span> <span class="n">outfilepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes an aligned fasta file (protein) and writes a .fa file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">inhandle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">infilepath</span><span class="p">)</span>
    <span class="n">outhandle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfilepath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">afa_file</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">inhandle</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">afa_file</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
        <span class="n">record</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">outhandle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">get_fa_record_text_from_obj</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>
    <span class="n">inhandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">outhandle</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="replace_seqs_in_alignment_with_seqs_from_fasta"><a class="viewcode-back" href="../replace_seqs.html#replace_seqs.replace_seqs_in_alignment_with_seqs_from_fasta">[docs]</a><span class="k">def</span> <span class="nf">replace_seqs_in_alignment_with_seqs_from_fasta</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">fasta</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes sequences from an alignment, finds the top blastp hit in a given</span>
<span class="sd">    fasta file for each one, then aligns the top hits to the alignment, then</span>
<span class="sd">    removes the original sequences from the alignment, writes the alignemnt,</span>
<span class="sd">    and returns the path to the output alignment file.</span>

<span class="sd">    Note: Might be better to explicitly mask the new sequences to only the</span>
<span class="sd">    columns that align with the original trimmed sequences instead of applying</span>
<span class="sd">    general masking criteria, because some input alignments might be masked</span>
<span class="sd">    more stringently.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Make a folder to contain temporary files.</span>
    <span class="n">tempdirpath</span> <span class="o">=</span> <span class="n">alignment</span> <span class="o">+</span> <span class="s1">&#39;_temp&#39;</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">):</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">)</span>

    <span class="c1"># Copy nexus alignment into temporary directory.</span>
    <span class="c1">#nex_ali_copy = os.path.join(tempdirpath, os.path.basename(alignment))</span>
    <span class="c1">#shutil.copyfile(alignment, nex_ali_copy) </span>

    <span class="c1"># Remove extra mesquite lines if present.</span>
    <span class="n">delete_extra_mesquite_lines</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>

    <span class="c1"># Get number of sequences and accessions in the input alignment.</span>
    <span class="n">alignment_seq_num</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">original_seq_accs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">):</span>
        <span class="n">alignment_seq_num</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">original_seq_accs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="c1"># Convert alignment to fasta.</span>
    <span class="n">afa_alignment</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">,</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.afaa&#39;</span><span class="p">)</span>
    <span class="n">nex_to_afa</span><span class="p">(</span><span class="n">alignment</span><span class="p">,</span> <span class="n">afa_alignment</span><span class="p">)</span> 

    <span class="c1"># Remove &#39;?&#39; characters from alignment, and replace space characters in</span>
    <span class="c1"># fasta headers with underscores..</span>
    <span class="n">modified_original_seq_accs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">afa_alignment2</span> <span class="o">=</span> <span class="n">afa_alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>\
        <span class="o">+</span> <span class="s1">&#39;_withoutquestionmarks.afaa&#39;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">afa_alignment</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">afa_alignment2</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">infh</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">):</span>
                <span class="c1">#new_header_line = i.replace(&#39; &#39;, &#39;_&#39;)</span>
                <span class="n">new_header_line</span> <span class="o">=</span> <span class="n">i</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;__ORIGNALSEQ</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_header_line</span><span class="p">)</span>
                <span class="c1"># Assumes that there are no multi-line fasta headers.</span>
                <span class="n">modified_original_seq_accs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_header_line</span><span class="o">.</span><span class="n">strip</span><span class="p">()[</span><span class="mi">1</span><span class="p">:])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;?&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">))</span>

    <span class="c1"># Check that the correct number of headers was found.</span>

    <span class="c1"># Make a copy of the alignment in nexus format.</span>
    <span class="n">nex_ali_copy</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">alignment</span><span class="p">))</span>
    <span class="n">afa_to_nex</span><span class="p">(</span><span class="n">afa_alignment2</span><span class="p">,</span> <span class="n">nex_ali_copy</span><span class="p">)</span>

    <span class="c1"># Initiate a variable to store the path to a fasta file with top hits or</span>
    <span class="c1"># full-length sequences with the same IDs from the same genomes.</span>
    <span class="n">top_hit_fasta</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Initiate variable to store a list of all headers for sequences to be</span>
    <span class="c1"># added to the alignment.</span>
    <span class="n">nonredun_top_hit_headers</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># If there is a fasta file provided, then find top hits for each sequence</span>
    <span class="c1"># in the input alignment.</span>
    <span class="k">if</span> <span class="n">fasta</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Remove hyphens from fasta file (make it an unaligned fasta).</span>
        <span class="n">fa_query_file</span> <span class="o">=</span> <span class="n">afa_alignment2</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.faa&#39;</span>
        <span class="n">afa_to_fa</span><span class="p">(</span><span class="n">afa_alignment2</span><span class="p">,</span> <span class="n">fa_query_file</span><span class="p">)</span>

        <span class="c1"># Copy the input fasta file into the temporary directory.</span>
        <span class="n">fasta_copy</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fasta</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
            <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.faa&#39;</span><span class="p">))</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">fasta</span><span class="p">,</span> <span class="n">fasta_copy</span><span class="p">)</span> 

        <span class="c1"># Make input fasta file a blastable database.</span>
        <span class="n">make_blast_db</span><span class="p">(</span><span class="n">fasta_copy</span><span class="p">)</span>

        <span class="c1"># BLASTP search with the sequences from the alignment as the query into the</span>
        <span class="c1"># input fasta file.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Running blastp search&#39;</span><span class="p">)</span>
        <span class="n">blast_outpath</span> <span class="o">=</span> <span class="n">fasta_copy</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_blastp_out.txt&#39;</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">([</span><span class="s1">&#39;blastp&#39;</span><span class="p">,</span> <span class="s1">&#39;-query&#39;</span><span class="p">,</span> <span class="n">fa_query_file</span><span class="p">,</span> <span class="s1">&#39;-db&#39;</span><span class="p">,</span> <span class="n">fasta_copy</span><span class="p">,</span>
        <span class="s1">&#39;-out&#39;</span><span class="p">,</span> <span class="n">blast_outpath</span><span class="p">,</span> <span class="s1">&#39;-outfmt&#39;</span><span class="p">,</span> <span class="s1">&#39;5&#39;</span><span class="p">,</span> <span class="s1">&#39;-evalue&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="mf">0.000000001</span><span class="p">),</span>
        <span class="s1">&#39;-num_threads&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="mi">4</span><span class="p">)])</span> 

        <span class="c1"># Parse the output, and make a list of all the top hits for each</span>
        <span class="c1"># sequence/query/searchrecord.</span>
        <span class="c1"># Iterate through each search record (one for each query sequence) and</span>
        <span class="c1"># compile a list of headers for top hits.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Parsing blastp output file</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">top_hit_headers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">queries_retrieving_no_hits</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">modified_original_seq_accs</span><span class="p">)</span>
        <span class="n">log_file_path</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_with_newseqs.csv&#39;</span> 
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">log_file_path</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
            <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Header of original sequence,Header of replacement sequence,E-value of top HSP</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">idx</span> <span class="o">=</span> <span class="n">SearchIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">blast_outpath</span><span class="p">,</span> <span class="s1">&#39;blast-xml&#39;</span><span class="p">)</span>
            <span class="n">qnum</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">idx</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                <span class="n">qnum</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">query_id</span> <span class="o">=</span> <span class="n">idx</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">id</span>
                <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">query_id</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span><span class="p">)</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">Query &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">qnum</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">query_id</span><span class="p">)</span>
                <span class="n">found_a_hit</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">idx</span><span class="p">[</span><span class="n">key</span><span class="p">]:</span>
                    <span class="n">found_a_hit</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">Top hit:&#39;</span> <span class="o">+</span> <span class="n">hit</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">Top HSP E-value:&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">hit</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">evalue</span><span class="p">))</span>
                    <span class="n">top_hit_headers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hit</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                    <span class="c1">#if query_id in queries_retrieving_no_hits:</span>
                    <span class="c1">#    queries_retrieving_no_hits.remove(query_id)</span>
                    <span class="n">queries_retrieving_no_hits</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">query_id</span><span class="p">)</span>
                    <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">hit</span><span class="o">.</span><span class="n">id</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">hit</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">evalue</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="k">break</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">found_a_hit</span><span class="p">:</span>
                    <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;N/A,N/A</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">idx</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

        <span class="c1"># Remove redundant headers from the list.</span>
        <span class="n">nonredun_top_hit_headers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">top_hit_headers</span><span class="p">))</span>

        <span class="c1"># Calculate number of redundant sequences among top hits.</span>
        <span class="n">num_redun_top_hits</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">top_hit_headers</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">nonredun_top_hit_headers</span><span class="p">)</span>

        <span class="c1"># Report stats.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of sequences in input alignment:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">alignment_seq_num</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">top_hit_headers</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of nonredundant top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nonredun_top_hit_headers</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of redundant sequences among top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">num_redun_top_hits</span><span class="p">)</span>

        <span class="c1"># Check that the number of top hits is not below 90 percent of the number</span>
        <span class="c1"># of query sequences.</span>
        <span class="n">percent_of_queries_without_hits</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">queries_retrieving_no_hits</span><span class="p">)</span> <span class="o">/</span>\
            <span class="n">alignment_seq_num</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span><span class="se">\n</span><span class="s2">Total queries that did not retrieve any hits in the input fasta</span>
<span class="s2">        file: </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">queries_retrieving_no_hits</span><span class="p">)))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span><span class="se">\n</span><span class="s2">Percent of queries that did not retrieve any hits in the input</span>
<span class="s2">        fasta file: </span><span class="si">%s</span><span class="s2">&quot;&quot;&quot;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">percent_of_queries_without_hits</span><span class="p">))</span>
        <span class="k">assert</span> <span class="n">percent_of_queries_without_hits</span> <span class="o">&lt;=</span> <span class="mi">10</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;More than 10</span><span class="si">% o</span><span class="s2">f queries</span>
<span class="s2">        retrieved no hits in the input fasta file.&quot;&quot;&quot;</span> 

        <span class="c1">## Check that the number of top hits is the same as the number of query</span>
        <span class="c1">## sequences.</span>
        <span class="c1">#assert len(top_hit_headers) == alignment_seq_num, &quot;&quot;&quot;Different number</span>
        <span class="c1">#of top hits than query sequences. There may be no hits for some sequences</span>
        <span class="c1">#in the input alignment.&quot;&quot;&quot;</span>

        <span class="c1"># Retrieve the top hit sequences from the input fasta file, and align them</span>
        <span class="c1"># to the input alignment file.</span>
        <span class="c1">#print(&#39;\nAligning top hits to input alignment&#39;)</span>
        <span class="c1">#with open(fasta_copy) as infp:</span>
        <span class="c1">#    for seq in SeqIO.parse(infp, &#39;fasta&#39;):</span>
        <span class="c1">#        if seq.id in nonredun_top_hit_headers:</span>
        <span class="c1">#            # Remove &#39;?&#39; characters from sequence?</span>
        <span class="c1">#            #seq.seq = str(seq.seq).replace(&#39;?&#39;, &#39;&#39;)</span>
        <span class="c1">#            assert not &#39;?&#39; in str(seq.seq)</span>
        <span class="c1">#            # Align to input alignment.</span>
        <span class="c1">#            new_nex_path = nex_ali_copy + &#39;_temp.nex&#39; </span>
        <span class="c1">#            add_seq_to_alignment3(seq, nex_ali_copy, new_nex_path)</span>
        <span class="c1">#            # Delete old nex.</span>
        <span class="c1">#            os.remove(nex_ali_copy)</span>
        <span class="c1">#            # Rename new nex as old nex.</span>
        <span class="c1">#            os.rename(new_nex_path, nex_ali_copy)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Getting top hits from input fasta file&#39;</span><span class="p">)</span>
        <span class="n">top_hit_fasta</span> <span class="o">=</span> <span class="n">fasta_copy</span> <span class="o">+</span> <span class="s1">&#39;_top_hits.faa&#39;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_copy</span><span class="p">)</span> <span class="k">as</span> <span class="n">infp</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">top_hit_fasta</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">infp</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">nonredun_top_hit_headers</span><span class="p">:</span>
                    <span class="c1"># Write to file.</span>
                    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">o</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

    <span class="c1"># Otherwise look in relevant files in the Genomes directory.</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># Iterate through sequences in the input alignment and compile a list</span>
        <span class="c1"># of sequence objects corresponding to each sequence.</span>
        <span class="n">seq_objects</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">nonredun_top_hit_headers</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">afa_alignment</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">infh</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">):</span>
                <span class="c1"># Extract the species name and sequence ID from the sequence header.</span>
                <span class="n">header</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">id</span>
                <span class="c1">#print(&#39;\nFinding full-length sequence for sequence: %s&#39; % header)</span>
                <span class="n">species_name</span> <span class="o">=</span> <span class="n">header</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="c1">#print(&#39;species name: &#39; + species_name)</span>
                <span class="n">seq_id</span> <span class="o">=</span> <span class="n">header</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                <span class="c1">#print(&#39;sequence identifier: &#39; + seq_id)</span>

                <span class="c1"># Determine which databases could correspond to the species name,</span>
                <span class="c1"># given information in the genome info spreadsheet.</span>
                <span class="n">db_names_with_species_name</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">settings</span><span class="o">.</span><span class="n">db_info_csv</span><span class="p">)</span> <span class="k">as</span> <span class="n">infh2</span><span class="p">:</span>
                    <span class="n">dfx</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">infh2</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">dfx</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">species_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Species (if applicable)&#39;</span><span class="p">]:</span>
                            <span class="n">filename</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Filename&#39;</span><span class="p">]</span>
                            <span class="k">if</span> <span class="n">filename</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.faa&#39;</span><span class="p">):</span>
                                <span class="c1">#print(filename)</span>
                                <span class="n">db_names_with_species_name</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>

                <span class="c1"># Check that at least one database might have the full-length sequence.</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_names_with_species_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;&quot;&quot;No databases</span>
<span class="s2">                identified for species name </span><span class="si">%s</span><span class="s2">.&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">species_name</span>

                <span class="c1"># For each possible database/file, try to retrieve a sequence with</span>
                <span class="c1"># the ID.</span>
                <span class="n">full_length_seq_obj</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">for</span> <span class="n">db_name</span> <span class="ow">in</span> <span class="n">db_names_with_species_name</span><span class="p">:</span>
                    <span class="c1"># Call a function to retrieve the sequence from a database</span>
                    <span class="c1"># given an ID.</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1"># Try fast retrieval, and if that doesn&#39;t work then try</span>
                        <span class="c1"># the slow method.</span>
                        <span class="n">full_length_seq_obj</span> <span class="o">=</span>\
                        <span class="n">get_seqs_from_fasta_db</span><span class="p">(</span><span class="n">db_name</span><span class="p">,</span> <span class="p">[</span><span class="n">seq_id</span><span class="p">],</span> <span class="kc">False</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

                    <span class="k">except</span><span class="p">:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Looking in another file.&#39;</span><span class="p">)</span>
                        <span class="k">continue</span>

                <span class="c1"># Check that a sequence was found.</span>
                <span class="k">assert</span> <span class="n">full_length_seq_obj</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>

                <span class="c1"># Change sequence header to be identical to that for the</span>
                <span class="c1"># sequence that it is replacing.</span>
                <span class="n">full_length_seq_obj</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
                <span class="n">full_length_seq_obj</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">header</span>

                <span class="c1"># Add sequence object to list. </span>
                <span class="n">seq_objects</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">full_length_seq_obj</span><span class="p">)</span>

            <span class="c1"># Append each sequence object in the list to a fasta file.</span>
            <span class="n">top_hit_fasta</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_full_length_seqs.faa&#39;</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">top_hit_fasta</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">seq_object</span> <span class="ow">in</span> <span class="n">seq_objects</span><span class="p">:</span>
                    <span class="c1"># Add header to list of headers for sequences to add.</span>
                    <span class="n">nonredun_top_hit_headers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_object</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

                    <span class="c1"># Write to file.</span>
                    <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_object</span><span class="p">,</span> <span class="n">o</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>

            <span class="c1"># Make list of sequence headers for sequences to add non-redundant.</span>
            <span class="n">nonredun_top_hit_headers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">nonredun_top_hit_headers</span><span class="p">))</span>

    <span class="c1"># Align top hits to copy of input alignment.</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Aligning replacement sequences to input alignment&#39;</span><span class="p">)</span>
    <span class="n">new_nex_path</span> <span class="o">=</span> <span class="n">alignment</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_with_newseqs.nex&#39;</span> 
    <span class="n">do_align_iteratively</span><span class="p">(</span><span class="n">nex_ali_copy</span><span class="p">,</span> <span class="n">top_hit_fasta</span><span class="p">,</span> <span class="n">new_nex_path</span><span class="p">)</span>
    
    <span class="c1"># Mask the alignment and write to the directory that the input alignment is</span>
    <span class="c1"># in.</span>
    <span class="n">masked_nex_path</span> <span class="o">=</span> <span class="n">new_nex_path</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.mask.nex&#39;</span>
    <span class="n">mask_nex</span><span class="p">(</span><span class="n">new_nex_path</span><span class="p">,</span> <span class="n">masked_nex_path</span><span class="p">)</span>

    <span class="c1"># Trim the alignment.</span>
    <span class="n">trimmed_nex_path</span> <span class="o">=</span> <span class="n">masked_nex_path</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.trim.nex&#39;</span>
    <span class="n">trim_nex</span><span class="p">(</span><span class="n">masked_nex_path</span><span class="p">,</span> <span class="n">trimmed_nex_path</span><span class="p">)</span>
    
    <span class="c1"># Remove all the original sequences from the trimmed alignment.</span>
    <span class="n">new_ali_obj</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">trimmed_nex_path</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>
    <span class="c1">#new_ali_obj_copy = copy.deepcopy(new_ali_obj)</span>
    <span class="n">alphabet_to_use</span> <span class="o">=</span> <span class="n">determine_alphabet</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
    <span class="n">new_ali_obj_reduced</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([],</span> <span class="n">alphabet</span><span class="o">=</span><span class="n">alphabet_to_use</span><span class="p">)</span>
    <span class="n">inum</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">new_ali_obj</span><span class="p">:</span>
        <span class="n">inum</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">nonredun_top_hit_headers</span><span class="p">:</span>
            <span class="n">new_ali_obj_reduced</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_ali_obj</span><span class="p">[</span><span class="n">inum</span><span class="p">])</span>

    <span class="c1"># Write reduced alignment to directory that input alignment is in.</span>
    <span class="n">final_output_nex</span> <span class="o">=</span> <span class="n">trimmed_nex_path</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.newseqs.nex&#39;</span>
    <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_ali_obj_reduced</span><span class="p">,</span> <span class="n">final_output_nex</span><span class="p">,</span> <span class="s1">&#39;nexus&#39;</span><span class="p">)</span>

    <span class="c1"># Remove temporary directory and contents.</span>
    <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">tempdirpath</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">fasta</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Report number of sequences for which top hits were not identified.</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">queries_retrieving_no_hits</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;&quot;</span><span class="se">\n</span><span class="s2">The following sequences (</span><span class="si">%s</span><span class="s2"> percent)from the input alignment retrieved no</span>
<span class="s2">            blastp hits from the input fasta file:&quot;&quot;&quot;</span> <span class="o">%</span> <span class="n">percent_of_queries_without_hits</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">queries_retrieving_no_hits</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="p">)</span>

        <span class="c1"># Report stats again.</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of sequences in input alignment:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">alignment_seq_num</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">top_hit_headers</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of nonredundant top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">nonredun_top_hit_headers</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Number of redundant sequences among top hits:&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">num_redun_top_hits</span><span class="p">)</span>

    <span class="c1"># Print paths to main output files.</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Fasta file with identified replacement sequences:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">top_hit_fasta</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Trimmed nexus alignment with only replacement sequences:&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">final_output_nex</span><span class="p">)</span>


    <span class="c1"># Return the output alignment file path.</span>
    <span class="k">return</span> <span class="n">final_output_nex</span></div>

</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>