
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_amoebae_select_seqs module &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="module_amoebae_srchresfile module" href="module_amoebae_srchresfile.html" />
    <link rel="prev" title="module_amoebae_select_positions module" href="module_amoebae_select_positions.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_amoebae_srchresfile.html" title="module_amoebae_srchresfile module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="module_amoebae_select_positions.html" title="module_amoebae_select_positions module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-module_amoebae_select_seqs">
<span id="module-amoebae-select-seqs-module"></span><h1>module_amoebae_select_seqs module<a class="headerlink" href="#module-module_amoebae_select_seqs" title="Permalink to this headline">¶</a></h1>
<p>Contains functions for modifying sequence selection for a given phylogenetic
tree.</p>
<dl class="class">
<dt id="module_amoebae_select_seqs.TaxonomicInfo">
<em class="property">class </em><code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">TaxonomicInfo</code><span class="sig-paren">(</span><em>superbranch</em>, <em>supergroup</em>, <em>group</em>, <em>species</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#TaxonomicInfo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.TaxonomicInfo" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Class for storing taxonomic info.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.define_nodestyles_dict_for_colourcoding">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">define_nodestyles_dict_for_colourcoding</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#define_nodestyles_dict_for_colourcoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.define_nodestyles_dict_for_colourcoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Define node styles for different nodes of interest that need to be
colour-coded.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.define_textface_for_labeling_stem">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">define_textface_for_labeling_stem</code><span class="sig-paren">(</span><em>clade_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#define_textface_for_labeling_stem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.define_textface_for_labeling_stem" title="Permalink to this definition">¶</a></dt>
<dd><p>Define ete3 textface styles for different nodes of interest that need to be
labeled.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_branch_length_info">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_branch_length_info</code><span class="sig-paren">(</span><em>t1</em>, <em>type_seqs_dict</em>, <em>outpath=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_branch_length_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_branch_length_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Take an ete3 Tree object and a file with type seqs, and return a
dictionary with keys as type sequence names and values as branch length
ratios for the respective clades in the input tree.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_clade_name_from_model2">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_clade_name_from_model2</code><span class="sig-paren">(</span><em>type_seq_name</em>, <em>type_seqs_dict</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_clade_name_from_model2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_clade_name_from_model2" title="Permalink to this definition">¶</a></dt>
<dd><p>Return clade name for given type sequence name and type seqs info file
(from models directory).</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_corresponding_node">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_corresponding_node</code><span class="sig-paren">(</span><em>node</em>, <em>tree</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_corresponding_node"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_corresponding_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ete3 TreeNode object in a tree that corresponds to
a given node.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_list_of_leaf_names_for_node">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_list_of_leaf_names_for_node</code><span class="sig-paren">(</span><em>node</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_list_of_leaf_names_for_node"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_list_of_leaf_names_for_node" title="Permalink to this definition">¶</a></dt>
<dd><p>Given an ete3 TreeNode object, return a list of terminal/leaf node
names.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_ml_tree_branch_lengths">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_ml_tree_branch_lengths</code><span class="sig-paren">(</span><em>tree</em>, <em>alignment</em>, <em>subs_model</em>, <em>outputdir</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_ml_tree_branch_lengths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_ml_tree_branch_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Run an ML tree with a constrained topology given a tree (for
constraining topology), an alignment file, and a substitution model.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_nodes_from_actual_tree_obj">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_nodes_from_actual_tree_obj</code><span class="sig-paren">(</span><em>t3</em>, <em>nodes_of_interest</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_nodes_from_actual_tree_obj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_nodes_from_actual_tree_obj" title="Permalink to this definition">¶</a></dt>
<dd><p>Get nodes in the given tree object that correspond to the nodes in a
given list of nodes from another tree object, and return as a list.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_nodes_of_interest">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_nodes_of_interest</code><span class="sig-paren">(</span><em>tree</em>, <em>type_seq_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_nodes_of_interest"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_nodes_of_interest" title="Permalink to this definition">¶</a></dt>
<dd><p>Take an ete3 tree (TreeNode object), a list of type sequence names, and
return a list of nodes of interest defined by clades (as ete3 TreeNode
objects).</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.get_taxonomic_info">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">get_taxonomic_info</code><span class="sig-paren">(</span><em>species_name</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#get_taxonomic_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.get_taxonomic_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a species name that appears in the genome info csv file in the
genomes folder, return taxonomic information from that csv file.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.optimize_sequence_selection">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">optimize_sequence_selection</code><span class="sig-paren">(</span><em>file_with_seqs</em>, <em>model_name</em>, <em>essential_taxa</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#optimize_sequence_selection"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.optimize_sequence_selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Add/remove homologous sequences from a dataset to optimize for long stem
lengths and low average branch lengths for clades of interest.</p>
<p>The essential_taxa argument is a list of names corresponding to taxonomic
groups of eukaryotes (such as “Amorphea” or “Diaphoretickes”) which appear
in the genome info csv file in the genomes directory.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.optimize_sequence_selection2">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">optimize_sequence_selection2</code><span class="sig-paren">(</span><em>file_with_seqs</em>, <em>model_name</em>, <em>essential_taxa</em>, <em>timestamp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#optimize_sequence_selection2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.optimize_sequence_selection2" title="Permalink to this definition">¶</a></dt>
<dd><p>Add/remove homologous sequences from a dataset to optimize for long stem
lengths and low average branch lengths for clades of interest.</p>
<p>The essential_taxa argument is a list of names corresponding to taxonomic
groups of eukaryotes (such as “Amorphea” or “Diaphoretickes”) which appear
in the genome info csv file in the genomes directory.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.recursively_add_seqs">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">recursively_add_seqs</code><span class="sig-paren">(</span><em>recursion_num</em>, <em>t1</em>, <em>alignment</em>, <em>subs_model</em>, <em>type_seqs_dict</em>, <em>seqs_attempted_to_add</em>, <em>essential_taxa</em>, <em>branch_length_info_dict</em>, <em>all_sequences</em>, <em>log_file_path</em>, <em>optimize_internal_branches</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#recursively_add_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.recursively_add_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a tree, and try to add sequences to improve the ratio of the
stem lengths of clades to the branch lengths of the branches in the clades,
and return the final tree when all the sequences have been added to the
dataset (at least temporarily).</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.recursively_remove_seqs">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">recursively_remove_seqs</code><span class="sig-paren">(</span><em>recursion_num</em>, <em>t1</em>, <em>alignment</em>, <em>subs_model</em>, <em>type_seqs_dict</em>, <em>seqs_attempted_to_remove</em>, <em>essential_taxa</em>, <em>branch_length_info_dict</em>, <em>log_file_path</em>, <em>optimize_internal_branches</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#recursively_remove_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.recursively_remove_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a tree, and try to remove sequences to improve the ratio of the
stem lengths of clades to the branch lengths of the branches in the clades,
and return the final tree when all the non-taxonomically-essential
sequences have been removed or no sequence can be removed that will result
in improving the branch length ratios.</p>
</dd></dl>

<dl class="function">
<dt id="module_amoebae_select_seqs.reduce_alignment">
<code class="descclassname">module_amoebae_select_seqs.</code><code class="descname">reduce_alignment</code><span class="sig-paren">(</span><em>alignment_file</em>, <em>output_alignment_file</em>, <em>removal_name_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_amoebae_select_seqs.html#reduce_alignment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_amoebae_select_seqs.reduce_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Take file path for an existing alignment and an output alignment, and
a list of sequence names for removal. Write the reduced alignment to the
output path in nexus format. Note: This function assumes that the names in
the removal lit and the alignment are all either coded or uncoded.</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="module_amoebae_select_positions.html"
                        title="previous chapter">module_amoebae_select_positions module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="module_amoebae_srchresfile.html"
                        title="next chapter">module_amoebae_srchresfile module</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_amoebae_srchresfile.html" title="module_amoebae_srchresfile module"
             >next</a> |</li>
        <li class="right" >
          <a href="module_amoebae_select_positions.html" title="module_amoebae_select_positions module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.6.
    </div>
  </body>
</html>