
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_nhmmer_scaffolds module &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="module_paralogue_counter module" href="module_paralogue_counter.html" />
    <link rel="prev" title="module_get_redun_hits module" href="module_get_redun_hits.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_paralogue_counter.html" title="module_paralogue_counter module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="module_get_redun_hits.html" title="module_get_redun_hits module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-module_nhmmer_scaffolds">
<span id="module-nhmmer-scaffolds-module"></span><h1>module_nhmmer_scaffolds module<a class="headerlink" href="#module-module_nhmmer_scaffolds" title="Permalink to this headline">¶</a></h1>
<p>This module is for using nhmmer to find the location of nucleotide sequences
matching an HMM, and outputs to a fasta file.</p>
<dl class="class">
<dt id="module_nhmmer_scaffolds.NhmmerHitAlig">
<em class="property">class </em><code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">NhmmerHitAlig</code><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#NhmmerHitAlig"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.NhmmerHitAlig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<dl class="method">
<dt id="module_nhmmer_scaffolds.NhmmerHitAlig.db_name">
<code class="descname">db_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#NhmmerHitAlig.db_name"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.NhmmerHitAlig.db_name" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="module_nhmmer_scaffolds.NhmmerHitAlig.hit_range">
<code class="descname">hit_range</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#NhmmerHitAlig.hit_range"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.NhmmerHitAlig.hit_range" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="module_nhmmer_scaffolds.NhmmerHitAlig.sequ">
<code class="descname">sequ</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#NhmmerHitAlig.sequ"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.NhmmerHitAlig.sequ" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="module_nhmmer_scaffolds.NhmmerHitAlig.sequprot">
<code class="descname">sequprot</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#NhmmerHitAlig.sequprot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.NhmmerHitAlig.sequprot" title="Permalink to this definition">¶</a></dt>
<dd><p>Translate the sequence based on how it aligns to the HMM.</p>
<p><a href="#id1"><span class="problematic" id="id2">**</span></a><a href="#id3"><span class="problematic" id="id4">*</span></a>Assumes that the HMM starts with the start codon of the aligned
homologues, and that the rest of the alignment is in frame.</p>
<p><a href="#id5"><span class="problematic" id="id6">**</span></a><a href="#id7"><span class="problematic" id="id8">*</span></a>Assumes indels keep the whole hit in frame.</p>
<p><a href="#id9"><span class="problematic" id="id10">**</span></a><a href="#id11"><span class="problematic" id="id12">*</span></a>Assumes standard genetic code.</p>
<p><a href="#id13"><span class="problematic" id="id14">**</span></a><a href="#id15"><span class="problematic" id="id16">*</span></a>Does not always to a good job of translating in the right frame…</p>
<p>May not be sufficient. May need to align to whole query sequence to
determine frame along length.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.check_if_hits_overlapping">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">check_if_hits_overlapping</code><span class="sig-paren">(</span><em>hit_objs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#check_if_hits_overlapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.check_if_hits_overlapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of nhmmer hit objects, and returns true if they overlap.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.cluster_ali_recur">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">cluster_ali_recur</code><span class="sig-paren">(</span><em>sublists</em>, <em>origlist</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#cluster_ali_recur"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.cluster_ali_recur" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively cluster each object in the list with those that are within
10,000 bp. Return a dict.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.display_dict">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">display_dict</code><span class="sig-paren">(</span><em>nhmmer_hit_objs_dict2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#display_dict"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.display_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints relevant info.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.frameify_seq">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">frameify_seq</code><span class="sig-paren">(</span><em>inseq</em>, <em>hmmfrom</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#frameify_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.frameify_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a string, and return a modified string that is in fram with the
HMM.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_concat_seq">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_concat_seq</code><span class="sig-paren">(</span><em>nhmmer_hit_objs</em>, <em>dtype='nucl'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_concat_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_concat_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of nhmmer hit alignment objects that should correspond to
the same gene, and returns a concatenated sequence.</p>
<p>May not be sufficient if the hit alignments do overlap.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_db_fasta_filepath">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_db_fasta_filepath</code><span class="sig-paren">(</span><em>dir_string</em>, <em>dbname</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_db_fasta_filepath"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_db_fasta_filepath" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the name of the appropriate fasta file.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_gene_region">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_gene_region</code><span class="sig-paren">(</span><em>nhmmer_hit_objs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_gene_region"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_gene_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of nhmmer hit alignment objects that should correspond to
the same gene, and returns a concatenated sequence (of appropriate strand).</p>
<p>So, if the gene region is on the negative strand, then it returns the
sequence of the negative strand (5’ to 3’) by getting the reverse
complement of the forward strand.</p>
<p><a href="#id17"><span class="problematic" id="id18">**</span></a><a href="#id19"><span class="problematic" id="id20">*</span></a>Does not account for paralogous genes possibly being very close to each
other on the same strand of the same chromosome…</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_nhmmer_hit_gene_region_seqs">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_nhmmer_hit_gene_region_seqs</code><span class="sig-paren">(</span><em>nhmmer_hit_objs</em>, <em>seq_filepath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_nhmmer_hit_gene_region_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_nhmmer_hit_gene_region_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a list of nhmmer hit alignment objects and return one or more
nucleotide subsequences that may correspond to the coding sequence for a
single gene.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_three_frame_transl">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_three_frame_transl</code><span class="sig-paren">(</span><em>seq_filepath</em>, <em>three_frame_translation_dirpath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_three_frame_transl"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_three_frame_transl" title="Permalink to this definition">¶</a></dt>
<dd><p>Produce three-frame translations of forward strand DNA sequences in a
given file, and write to files in a given directory path.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.get_three_frame_transl_string">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">get_three_frame_transl_string</code><span class="sig-paren">(</span><em>inputs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#get_three_frame_transl_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.get_three_frame_transl_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list comprising the header string and sequence string of an
input sequence.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.parse_nhmmer_out">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">parse_nhmmer_out</code><span class="sig-paren">(</span><em>nhmmer_outfilepath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#parse_nhmmer_out"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.parse_nhmmer_out" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens an nhmmer output file (tabular output), extracts the information for location of
domains in the sequences, and returns a list of hit alignment objects.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.run_hmmsearch">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">run_hmmsearch</code><span class="sig-paren">(</span><em>query_path</em>, <em>db_path</em>, <em>hmmsearch_outfilepath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#run_hmmsearch"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.run_hmmsearch" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs hmmsearch (protein queries).</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.run_nhmmer_search">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">run_nhmmer_search</code><span class="sig-paren">(</span><em>query_path</em>, <em>db_path</em>, <em>nhmmer_outfilepath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#run_nhmmer_search"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.run_nhmmer_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs nhmmer.</p>
</dd></dl>

<dl class="function">
<dt id="module_nhmmer_scaffolds.write_nhmmer_hit_fasta">
<code class="descclassname">module_nhmmer_scaffolds.</code><code class="descname">write_nhmmer_hit_fasta</code><span class="sig-paren">(</span><em>nhmmer_hit_objs</em>, <em>fasta_filepath</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_nhmmer_scaffolds.html#write_nhmmer_hit_fasta"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_nhmmer_scaffolds.write_nhmmer_hit_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of nhmmer hit alignment objects, and writes corresponding
fasta records to a fasta file.</p>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="module_get_redun_hits.html"
                        title="previous chapter">module_get_redun_hits module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="module_paralogue_counter.html"
                        title="next chapter">module_paralogue_counter module</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_paralogue_counter.html" title="module_paralogue_counter module"
             >next</a> |</li>
        <li class="right" >
          <a href="module_get_redun_hits.html" title="module_get_redun_hits module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.6.
    </div>
  </body>
</html>