
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_search_scaffolds module &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="module_search_scaffolds_test module" href="module_search_scaffolds_test.html" />
    <link rel="prev" title="module_paralogue_counter module" href="module_paralogue_counter.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_search_scaffolds_test.html" title="module_search_scaffolds_test module"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="module_paralogue_counter.html" title="module_paralogue_counter module"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="module-module_search_scaffolds">
<span id="module-search-scaffolds-module"></span><h1>module_search_scaffolds module<a class="headerlink" href="#module-module_search_scaffolds" title="Permalink to this headline">¶</a></h1>
<p>Contains functions used in search_scaffolds.py.</p>
<dl class="function">
<dt id="module_search_scaffolds.check_for_duplicate_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">check_for_duplicate_ranges</code><span class="sig-paren">(</span><em>range_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#check_for_duplicate_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.check_for_duplicate_ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if the given list of tuples contains duplicates.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.check_if_hsps_overlapping">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">check_if_hsps_overlapping</code><span class="sig-paren">(</span><em>hsp_objs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#check_if_hsps_overlapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.check_if_hsps_overlapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of HSP objects, and returns true if they overlap.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.check_if_two_hsp_ranges_overlap">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">check_if_two_hsp_ranges_overlap</code><span class="sig-paren">(</span><em>lists</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#check_if_two_hsp_ranges_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.check_if_two_hsp_ranges_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes two lists which contain the start and stop positions of two HSPs,
and returns True if they overlap.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.check_if_two_hsps_overlap">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">check_if_two_hsps_overlap</code><span class="sig-paren">(</span><em>hsp_obj1</em>, <em>hsp_obj2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#check_if_two_hsps_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.check_if_two_hsps_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes two HSP objects and returns True if their ranges overlap, and
False if not.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.check_strands">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">check_strands</code><span class="sig-paren">(</span><em>hsp_objs</em>, <em>fwd_strand</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#check_strands"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.check_strands" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if hsps in the list are not all on the same strand.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.clusters_overlap">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">clusters_overlap</code><span class="sig-paren">(</span><em>cluster1</em>, <em>cluster2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#clusters_overlap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.clusters_overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Take two lists of HSPs, and determine if they define overlapping ranges
in the subject sequence.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.find_missed_hsp_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">find_missed_hsp_ranges</code><span class="sig-paren">(</span><em>complete_ranges</em>, <em>final_proximate_ranges</em>, <em>leftover_proximate_ranges</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#find_missed_hsp_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.find_missed_hsp_ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Testable function to test for problems.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_best_proximate_hsps">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_best_proximate_hsps</code><span class="sig-paren">(</span><em>hsp_objs</em>, <em>top_hsp</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_best_proximate_hsps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_best_proximate_hsps" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of HSP objects that are sufficiently proximate to the top
HSP, and among which there is no overlap in the sequence ranges. And, also
return a list of HSP objects that are also proximate to these HSPs, but
were not included in the final list of HSPs because they overlap with the
ones that were. Instead of simply selecting HSPs based on hit range,
consider E-values, favoring HSPs with better E-values if HSPs overlap and
some have to be excluded.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_blastp_hit_seq_obj_and_coord">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_blastp_hit_seq_obj_and_coord</code><span class="sig-paren">(</span><em>hit</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_blastp_hit_seq_obj_and_coord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_blastp_hit_seq_obj_and_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a SearchIO hit object (containing one or more HSPs) and returns a
fasta sequence object likely to correspond to a single gene or the portion
of the subject sequence that is actually homologous to the query sequence.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_cluster_range">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_cluster_range</code><span class="sig-paren">(</span><em>hsp_list</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_cluster_range"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_cluster_range" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a list of SearchIO HSP objects an return a range from the furthest
5’ hsp end to the furthest 3’ hsp end.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_hit_seq_obj">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_hit_seq_obj</code><span class="sig-paren">(</span><em>hit</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_hit_seq_obj"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_hit_seq_obj" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a SearchIO hit object (containing one or more HSPs) and returns a
fasta sequence object likely to correspond to a single gene (the most
similar gene to the query in the nucleotide subject sequence).</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_hit_seq_record_and_coord">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_hit_seq_record_and_coord</code><span class="sig-paren">(</span><em>hit</em>, <em>proximate_hsp_objs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_hit_seq_record_and_coord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_hit_seq_record_and_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a list of SearchIO HSP objects (arguably representing a single
gene), return a Bio.Seq record composed of concatenated HSP sequences.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_hsp_clusters">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_hsp_clusters</code><span class="sig-paren">(</span><em>hit</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_hsp_clusters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_hsp_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a SearchIO Hit object and return a list of lists, where the
sublists are HSPs that cluster together on the subject sequence such that
it is likely that they represent a distinct gene.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_proximate_hsp_objs">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_proximate_hsp_objs</code><span class="sig-paren">(</span><em>hsp_objs</em>, <em>top_hsp</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_proximate_hsp_objs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_proximate_hsp_objs" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of HSP objects that are sufficiently proximate to the top
HSP, and among which there is no overlap in the sequence ranges. And, also
return a list of HSP objects that are also proximate to these HSPs, but
were not included in the final list of HSPs because they overlap with the
ones that were.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_proximate_hsp_objs2">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_proximate_hsp_objs2</code><span class="sig-paren">(</span><em>hsp_objs</em>, <em>top_hsp</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_proximate_hsp_objs2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_proximate_hsp_objs2" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of HSP objects that are sufficiently proximate to the top
HSP, and among which there is no overlap in the sequence ranges. And, also
return a list of HSP objects that are also proximate to these HSPs, but
were not included in the final list of HSPs because they overlap with the
ones that were. Instead of simply selecting HSPs based on hit range,
consider E-values, favoring HSPs with better E-values if HSPs overlap and
some have to be excluded.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_proximate_hsp_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_proximate_hsp_ranges</code><span class="sig-paren">(</span><em>range_list</em>, <em>top_hsp_range</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_proximate_hsp_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_proximate_hsp_ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a list of ranges and returns another list, but corresponding to
the top HSP and proximate HSPs that do not overlap.</p>
<p>Potential problem: Does not take into account the E-value of overlapping
HPSs, and may exclude some HSPs with better E-values because of their
position. It would perhaps be better to take E-values into account?</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_strand_word">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_strand_word</code><span class="sig-paren">(</span><em>hsp_hit_frame</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_strand_word"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_strand_word" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a frame (for a translation of a DNA sequence) and return ‘plus’ or
‘minus’ depending on which strand the frame corresponds to.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.get_tblastn_hit_seq_obj_and_coord">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">get_tblastn_hit_seq_obj_and_coord</code><span class="sig-paren">(</span><em>hit</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#get_tblastn_hit_seq_obj_and_coord"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.get_tblastn_hit_seq_obj_and_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a SearchIO hit object (containing one or more HSPs) and returns a
fasta sequence object likely to correspond to a single gene (the most
similar gene to the query in the nucleotide subject sequence).</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.parse_tblastn">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">parse_tblastn</code><span class="sig-paren">(</span><em>tblastn_out_path</em>, <em>seq_out_path</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#parse_tblastn"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.parse_tblastn" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse a given tblastn output file (XML) and write appropriate sequences
corresponding to hits therein to a given sequence file path. Also, return
relevant info for recording results.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.print_cluster">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">print_cluster</code><span class="sig-paren">(</span><em>clusterplus</em>, <em>hit_num</em>, <em>cluster_num</em>, <em>num_dots</em>, <em>startend=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#print_cluster"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.print_cluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Print visualization of cluster.</p>
<p>startend is a tuple containing the start and end position of the
subsequence of the subject sequence to be visualized.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.recursively_add_clusters">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">recursively_add_clusters</code><span class="sig-paren">(</span><em>hsp_list</em>, <em>hsp_clusters</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#recursively_add_clusters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.recursively_add_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>take a list of SearchIO HSP objects and a list of HSP clusters
(representing distinct genes), and add clusters recursively.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.reduce_left_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">reduce_left_ranges</code><span class="sig-paren">(</span><em>sorted_left_ranges</em>, <em>top_hsp_range</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#reduce_left_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.reduce_left_ranges" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.reduce_right_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">reduce_right_ranges</code><span class="sig-paren">(</span><em>sorted_right_ranges</em>, <em>top_hsp_range</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#reduce_right_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.reduce_right_ranges" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.reduce_to_best_hsps">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">reduce_to_best_hsps</code><span class="sig-paren">(</span><em>sorted_hsps</em>, <em>max_gap</em>, <em>recursion_num</em>, <em>name</em>, <em>leftright</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#reduce_to_best_hsps"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.reduce_to_best_hsps" title="Permalink to this definition">¶</a></dt>
<dd><p>Recursively select HSPs upstream of the top HSP for inclusion in the
cluster.</p>
<p>Note: This function may be inefficient.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.remove_hsps_with_redun_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">remove_hsps_with_redun_ranges</code><span class="sig-paren">(</span><em>range_list</em>, <em>hsp_objs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#remove_hsps_with_redun_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.remove_hsps_with_redun_ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove HSPs with redundant ranges.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.search_scaffolds">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">search_scaffolds</code><span class="sig-paren">(</span><em>query_file</em>, <em>subject_db</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#search_scaffolds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.search_scaffolds" title="Permalink to this definition">¶</a></dt>
<dd><p>Uses functions to search a given nucleotide db with a given protein
query.</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.sort_left_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">sort_left_ranges</code><span class="sig-paren">(</span><em>left_ranges</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#sort_left_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.sort_left_ranges" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.sort_right_ranges">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">sort_right_ranges</code><span class="sig-paren">(</span><em>right_ranges</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#sort_right_ranges"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.sort_right_ranges" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.split_tblastn_hits_into_separate_genes">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">split_tblastn_hits_into_separate_genes</code><span class="sig-paren">(</span><em>query_res_obj</em>, <em>max_gap</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#split_tblastn_hits_into_separate_genes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.split_tblastn_hits_into_separate_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a SearchIO QueryResult object and return a new object with hits
split into groups of HSPs that represent distinct genes. This is important,
because there may be multiple paralogous genes present in a single
nucleotide subject sequence (such as a chromosome or scaffold).</p>
</dd></dl>

<dl class="function">
<dt id="module_search_scaffolds.whether_fwd_strand">
<code class="descclassname">module_search_scaffolds.</code><code class="descname">whether_fwd_strand</code><span class="sig-paren">(</span><em>hsp</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/module_search_scaffolds.html#whether_fwd_strand"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#module_search_scaffolds.whether_fwd_strand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="module_paralogue_counter.html"
                        title="previous chapter">module_paralogue_counter module</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="module_search_scaffolds_test.html"
                        title="next chapter">module_search_scaffolds_test module</a></p>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="module_search_scaffolds_test.html" title="module_search_scaffolds_test module"
             >next</a> |</li>
        <li class="right" >
          <a href="module_paralogue_counter.html" title="module_paralogue_counter module"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">amoebae 0.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>