
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>module_nhmmer_scaffolds &#8212; amoebae 0.0 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for module_nhmmer_scaffolds</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># Copyright 2018 Lael D. Barlow</span>
<span class="c1"># </span>
<span class="c1"># Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);</span>
<span class="c1"># you may not use this file except in compliance with the License.</span>
<span class="c1"># You may obtain a copy of the License at</span>
<span class="c1"># </span>
<span class="c1">#     http://www.apache.org/licenses/LICENSE-2.0</span>
<span class="c1"># </span>
<span class="c1"># Unless required by applicable law or agreed to in writing, software</span>
<span class="c1"># distributed under the License is distributed on an &quot;AS IS&quot; BASIS,</span>
<span class="c1"># WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.</span>
<span class="c1"># See the License for the specific language governing permissions and</span>
<span class="c1"># limitations under the License.</span>
<span class="c1"># </span>
<span class="sd">&quot;&quot;&quot;This module is for using nhmmer to find the location of nucleotide sequences</span>
<span class="sd">matching an HMM, and outputs to a fasta file.</span>

<span class="sd">&quot;&quot;&quot;</span>


<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">from</span> <span class="nn">module_get_fas_from_db_dir</span> <span class="k">import</span> <span class="n">get_seq_obj_from_db_fasta</span>
<span class="kn">import</span> <span class="nn">settings</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span>

<span class="c1"># Maybe it would be better to get the whole gene region and align that to the</span>
<span class="c1"># nucleotide HMM, and then make a translation based on that alignment...</span>


<div class="viewcode-block" id="NhmmerHitAlig"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.NhmmerHitAlig">[docs]</a><span class="k">class</span> <span class="nc">NhmmerHitAlig</span><span class="p">:</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">query_name</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">db_filepath</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hmmfrom</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">hmmto</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alifrom</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alito</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">envfrom</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">envto</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="NhmmerHitAlig.sequ"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.NhmmerHitAlig.sequ">[docs]</a>    <span class="k">def</span> <span class="nf">sequ</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Retrieve object for full hit sequence.</span>
        <span class="c1"># Maybe there is a more efficient way to do this???</span>
        <span class="n">full_seq_obj</span> <span class="o">=</span> <span class="n">get_seq_obj_from_db_fasta</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">db_filepath</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1"># Get appropriate subsequence that aligned to HMM.</span>
        <span class="n">alig_seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
            <span class="n">alig_seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">full_seq_obj</span><span class="o">.</span><span class="n">seq</span><span class="p">)[</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alifrom</span><span class="p">)</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alito</span><span class="p">)]</span>

        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
            <span class="c1">#***********alifrom comes after alito if on neg strand!!!!!!!!!!!!</span>
            <span class="c1">#***********Have to reverse them to get the right subsequence.</span>
            <span class="n">subseqstring</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">full_seq_obj</span><span class="o">.</span><span class="n">seq</span><span class="p">)[</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alito</span><span class="p">)</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alifrom</span><span class="p">)]</span>
            <span class="n">x</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">subseqstring</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">ambiguous_dna</span><span class="p">)</span>
            <span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
            <span class="n">alig_seq</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>

        <span class="c1"># Check that sequence is the right length.</span>
        <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">alig_seq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alito</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Error 1&quot;</span>

        <span class="c1"># Return subsequence.</span>
        <span class="k">return</span> <span class="n">alig_seq</span></div>

<div class="viewcode-block" id="NhmmerHitAlig.sequprot"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.NhmmerHitAlig.sequprot">[docs]</a>    <span class="k">def</span> <span class="nf">sequprot</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Translate the sequence based on how it aligns to the HMM.</span>

<span class="sd">        ***Assumes that the HMM starts with the start codon of the aligned</span>
<span class="sd">        homologues, and that the rest of the alignment is in frame.</span>

<span class="sd">        ***Assumes indels keep the whole hit in frame.</span>

<span class="sd">        ***Assumes standard genetic code.</span>

<span class="sd">        ***Does not always to a good job of translating in the right frame...</span>

<span class="sd">        May not be sufficient. May need to align to whole query sequence to</span>
<span class="sd">        determine frame along length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Determine frame by how it aligns to the HMM, and remove bases</span>
        <span class="c1"># accordingly before translating.</span>
        <span class="n">seq_in_frame</span> <span class="o">=</span> <span class="n">frameify_seq</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequ</span><span class="p">(),</span> <span class="bp">self</span><span class="o">.</span><span class="n">hmmfrom</span><span class="p">)</span>

        <span class="c1"># Make the in frame sequence a Seq object, and translate.</span>
        <span class="n">seq_in_frame_obj</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="n">seq_in_frame</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">ambiguous_dna</span><span class="p">)</span>
        <span class="n">translation</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq_in_frame_obj</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span>

        <span class="c1"># Return translation of sequence.</span>
        <span class="c1">#print(&#39;found seq for &#39; + self.subj_seq_id)</span>
        <span class="k">return</span> <span class="n">translation</span></div>

<div class="viewcode-block" id="NhmmerHitAlig.db_name"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.NhmmerHitAlig.db_name">[docs]</a>    <span class="k">def</span> <span class="nf">db_name</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># Return the name of the db of origin.</span>
        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">db_filepath</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span></div>

<div class="viewcode-block" id="NhmmerHitAlig.hit_range"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.NhmmerHitAlig.hit_range">[docs]</a>    <span class="k">def</span> <span class="nf">hit_range</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">alifrom</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">alito</span><span class="p">]</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span><span class="p">))</span></div></div>


<span class="c1">#class HmmerHitAlig:</span>
<span class="c1">#    def __init__(self):</span>
<span class="c1">#        self.query_name = None</span>
<span class="c1">#        self.db_filepath = None</span>
<span class="c1">#        self.subj_seq_id = None</span>
<span class="c1">#        self.original_strand = None</span>
<span class="c1">#        self.score = None</span>
<span class="c1">#        self.evalue = None</span>
<span class="c1">#        self.hmmfrom = None</span>
<span class="c1">#        self.hmmto = None</span>
<span class="c1">#        self.alifrom = None</span>
<span class="c1">#        self.alito = None</span>
<span class="c1">#        self.envfrom = None</span>
<span class="c1">#        self.envto = None</span>
<span class="c1">#        self.subj_seq_position = None</span>
<span class="c1">#</span>
<span class="c1">#    def position(self):</span>
<span class="c1">#        &quot;&quot;&quot;Returns a string with the position in the subject sequence.</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        return &#39;[&#39; + self.alifrom + &#39;..&#39; + self.alito + &#39;]&#39;</span>
<span class="c1">#</span>
<span class="c1">#    def actual_position(self):</span>
<span class="c1">#        &quot;&quot;&quot;Returns a string with the position in the original db_filepath.</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        subj_seq_start = self.subj_seq_position.strip(&#39;[]&#39;).split(&#39;..&#39;)[0]</span>
<span class="c1">#        subj_seq_end = self.subj_seq_position.strip(&#39;[]&#39;).split(&#39;..&#39;)[1]</span>
<span class="c1">#        # This gets really complex when you have to consider the</span>
<span class="c1">#        # strand!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!</span>
<span class="c1">#        if self.original_strand == &#39;+&#39;:</span>
<span class="c1">#            return &#39;[&#39; + self.alifrom + </span>
<span class="c1">#</span>
<span class="c1">#        elif self.original_strand == &#39;-&#39;:</span>
<span class="c1">#</span>
<span class="c1">#</span>
<span class="c1">#</span>
<span class="c1">#    #def sequ(self):</span>
<span class="c1">#    #    # Retrieve object for full hit sequence.</span>
<span class="c1">#    #    # Maybe there is a more efficient way to do this???</span>
<span class="c1">#    #    full_seq_obj = get_seq_obj_from_db_fasta([self.subj_seq_id],</span>
<span class="c1">#    #            self.db_filepath)[0]</span>
<span class="c1">#</span>
<span class="c1">#    #    # Get appropriate subsequence that aligned to HMM.</span>
<span class="c1">#    #    alig_seq = &#39;&#39;</span>
<span class="c1">#    #    if self.strand == &#39;+&#39;:</span>
<span class="c1">#    #        alig_seq = str(full_seq_obj.seq)[int(self.alifrom) -1: int(self.alito)]</span>
<span class="c1">#</span>
<span class="c1">#    #    elif self.strand == &#39;-&#39;:</span>
<span class="c1">#    #        #***********alifrom comes after alito if on neg strand!!!!!!!!!!!!</span>
<span class="c1">#    #        #***********Have to reverse them to get the right subsequence.</span>
<span class="c1">#    #        subseqstring = str(full_seq_obj.seq)[int(self.alito) -1: int(self.alifrom)]</span>
<span class="c1">#    #        x = Seq(subseqstring, IUPAC.ambiguous_dna)</span>
<span class="c1">#    #        y = x.reverse_complement()</span>
<span class="c1">#    #        alig_seq = str(y)</span>
<span class="c1">#</span>
<span class="c1">#    #    # Check that sequence is the right length.</span>
<span class="c1">#    #    assert len(alig_seq) == abs(int(self.alito) - int(self.alifrom)) + 1, &quot;Error 1&quot;</span>
<span class="c1">#</span>
<span class="c1">#    #    # Return subsequence.</span>
<span class="c1">#    #    return alig_seq</span>
<span class="c1">#</span>
<span class="c1">#    #def sequprot(self):</span>
<span class="c1">#    #    &quot;&quot;&quot;Translate the sequence based on how it aligns to the HMM.</span>
<span class="c1">#</span>
<span class="c1">#    #    ***Assumes that the HMM starts with the start codon of the aligned</span>
<span class="c1">#    #    homologues, and that the rest of the alignment is in frame.</span>
<span class="c1">#</span>
<span class="c1">#    #    ***Assumes indels keep the whole hit in frame.</span>
<span class="c1">#</span>
<span class="c1">#    #    ***Assumes standard genetic code.</span>
<span class="c1">#</span>
<span class="c1">#    #    ***Does not always to a good job of translating in the right frame...</span>
<span class="c1">#</span>
<span class="c1">#    #    May not be sufficient. May need to align to whole query sequence to</span>
<span class="c1">#    #    determine frame along length.</span>
<span class="c1">#    #    &quot;&quot;&quot;</span>
<span class="c1">#    #    # Determine frame by how it aligns to the HMM, and remove bases</span>
<span class="c1">#    #    # accordingly before translating.</span>
<span class="c1">#    #    seq_in_frame = frameify_seq(self.sequ(), self.hmmfrom)</span>
<span class="c1">#</span>
<span class="c1">#    #    # Make the in frame sequence a Seq object, and translate.</span>
<span class="c1">#    #    seq_in_frame_obj = Seq(seq_in_frame, IUPAC.ambiguous_dna)</span>
<span class="c1">#    #    translation = str(seq_in_frame_obj.translate())</span>
<span class="c1">#</span>
<span class="c1">#    #    # Return translation of sequence.</span>
<span class="c1">#    #    #print(&#39;found seq for &#39; + self.subj_seq_id)</span>
<span class="c1">#    #    return translation</span>
<span class="c1">#</span>
<span class="c1">#    #def db_name(self):</span>
<span class="c1">#    #    # Return the name of the db of origin.</span>
<span class="c1">#    #    return os.path.basename(self.db_filepath).rsplit(&#39;_&#39;, 1)[0]</span>
<span class="c1">#</span>
<span class="c1">#    #def hit_range(self):</span>
<span class="c1">#    #    x = [self.alifrom, self.alito]</span>
<span class="c1">#    #    return (min(x), max(x))</span>


<div class="viewcode-block" id="frameify_seq"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.frameify_seq">[docs]</a><span class="k">def</span> <span class="nf">frameify_seq</span><span class="p">(</span><span class="n">inseq</span><span class="p">,</span> <span class="n">hmmfrom</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a string, and return a modified string that is in fram with the</span>
<span class="sd">    HMM.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_in_frame</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">hmmfromposition</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">hmmfrom</span><span class="p">)</span>
    <span class="n">remainder</span> <span class="o">=</span> <span class="n">hmmfromposition</span> <span class="o">%</span> <span class="mi">3</span>
    <span class="k">if</span> <span class="n">remainder</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="c1">#frame = 1</span>
        <span class="n">seq_in_frame</span> <span class="o">=</span> <span class="n">inseq</span> 
    <span class="k">elif</span> <span class="n">remainder</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="c1">#frame = 2</span>
        <span class="n">seq_in_frame</span> <span class="o">=</span> <span class="n">inseq</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
    <span class="k">elif</span> <span class="n">remainder</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1">#frame = 3</span>
        <span class="n">seq_in_frame</span> <span class="o">=</span> <span class="n">inseq</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

    <span class="k">return</span> <span class="n">seq_in_frame</span></div>


<div class="viewcode-block" id="check_if_hits_overlapping"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.check_if_hits_overlapping">[docs]</a><span class="k">def</span> <span class="nf">check_if_hits_overlapping</span><span class="p">(</span><span class="n">hit_objs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list of nhmmer hit objects, and returns true if they overlap.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Compile a list of the ranges.</span>
    <span class="n">ranges</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">hit_objs</span><span class="p">:</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">hit_range</span><span class="p">()</span> 
        <span class="n">ranges</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
    
    <span class="c1"># Generate list of all possible combinations of two ranges.</span>
    <span class="n">combos</span> <span class="o">=</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span><span class="n">ranges</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

    <span class="c1"># Loop over combinations and check for overlap.</span>
    <span class="n">overlap</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">combo</span> <span class="ow">in</span> <span class="n">combos</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">combo</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">y</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">combo</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">intersect</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">intersect</span> <span class="o">!=</span> <span class="nb">set</span><span class="p">():</span>
            <span class="n">overlap</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="n">overlap</span></div>


<div class="viewcode-block" id="get_db_fasta_filepath"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_db_fasta_filepath">[docs]</a><span class="k">def</span> <span class="nf">get_db_fasta_filepath</span><span class="p">(</span><span class="n">dir_string</span><span class="p">,</span> <span class="n">dbname</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the name of the appropriate fasta file.&quot;&quot;&quot;</span>
    <span class="n">subdir_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">subdir</span> <span class="k">for</span> <span class="n">subdir</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">dir_string</span><span class="p">)]</span>
    <span class="n">fasta_filepath</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">for</span> <span class="n">subdir</span> <span class="ow">in</span> <span class="n">subdir_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">subdir</span> <span class="o">==</span> <span class="n">dbname</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
            <span class="n">deep_subdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dir_string</span><span class="p">,</span> <span class="n">subdir</span><span class="p">)</span>
            <span class="n">deep_subdir_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">deepsubdir</span> <span class="k">for</span> <span class="n">deepsubdir</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">deep_subdir</span><span class="p">)</span> <span class="k">if</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">deep_subdir</span><span class="p">,</span> <span class="n">deepsubdir</span><span class="p">))]</span>
            <span class="k">for</span> <span class="n">deeper_subdir</span> <span class="ow">in</span> <span class="n">deep_subdir_list</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;_Nucl&#39;</span><span class="p">,</span> <span class="n">deeper_subdir</span><span class="p">):</span>
                    <span class="n">deeper_subdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">deep_subdir</span><span class="p">,</span> <span class="n">deeper_subdir</span><span class="p">)</span>
                    <span class="n">onlyfiles</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">deeper_subdir</span><span class="p">)</span> <span class="k">if</span>\
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">deeper_subdir</span><span class="p">,</span> <span class="n">f</span><span class="p">))]</span>
                    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">onlyfiles</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.fa&#39;</span><span class="p">):</span>
                            <span class="n">fasta_filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">deeper_subdir</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
                            <span class="k">assert</span> <span class="ow">not</span> <span class="n">fasta_filepath</span> <span class="o">==</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Error: Fasta file not identified.&quot;</span>
    <span class="k">return</span> <span class="n">fasta_filepath</span></div>


<div class="viewcode-block" id="run_nhmmer_search"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.run_nhmmer_search">[docs]</a><span class="k">def</span> <span class="nf">run_nhmmer_search</span><span class="p">(</span><span class="n">query_path</span><span class="p">,</span> <span class="n">db_path</span><span class="p">,</span> <span class="n">nhmmer_outfilepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs nhmmer.&quot;&quot;&quot;</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">([</span><span class="s2">&quot;nhmmer&quot;</span><span class="p">,</span> <span class="s2">&quot;-E&quot;</span><span class="p">,</span> <span class="s2">&quot;0.05&quot;</span><span class="p">,</span> <span class="s2">&quot;--cpu&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;--tblout&quot;</span><span class="p">,</span>\
            <span class="n">nhmmer_outfilepath</span><span class="p">,</span> <span class="n">query_path</span><span class="p">,</span> <span class="n">db_path</span><span class="p">])</span></div>


<div class="viewcode-block" id="run_hmmsearch"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.run_hmmsearch">[docs]</a><span class="k">def</span> <span class="nf">run_hmmsearch</span><span class="p">(</span><span class="n">query_path</span><span class="p">,</span> <span class="n">db_path</span><span class="p">,</span> <span class="n">hmmsearch_outfilepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs hmmsearch (protein queries).&quot;&quot;&quot;</span>
    <span class="c1"># Could use a bit score cutoff instead using the &quot;-T&quot; option.</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">call</span><span class="p">([</span><span class="s2">&quot;hmmsearch&quot;</span><span class="p">,</span> <span class="s2">&quot;-E&quot;</span><span class="p">,</span> <span class="s2">&quot;0.05&quot;</span><span class="p">,</span> <span class="s2">&quot;--cpu&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;--domtblout&quot;</span><span class="p">,</span>\
            <span class="n">hmmsearch_outfilepath</span><span class="p">,</span> <span class="n">query_path</span><span class="p">,</span> <span class="n">db_path</span><span class="p">])</span></div>
    

<div class="viewcode-block" id="parse_nhmmer_out"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.parse_nhmmer_out">[docs]</a><span class="k">def</span> <span class="nf">parse_nhmmer_out</span><span class="p">(</span><span class="n">nhmmer_outfilepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Opens an nhmmer output file (tabular output), extracts the information for location of</span>
<span class="sd">    domains in the sequences, and returns a list of hit alignment objects.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Loop through the lines and identify the relevant info applying to all hits.</span>
    <span class="n">target_filepath</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">hmm_name</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">nhmmer_outfilepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">i</span><span class="p">:</span>
        <span class="c1"># Get info from end of file.</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
            <span class="n">split_mod_line</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39; +&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;# Target file:&#39;</span><span class="p">):</span>
                <span class="n">target_filepath</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;# Query file:&#39;</span><span class="p">):</span>
                <span class="n">hmm_name</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>

    <span class="c1"># Loop through the lines again and make objects for each hit alignement.</span>
    <span class="n">hit_alignments</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">nhmmer_outfilepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">i</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
            <span class="n">write</span> <span class="o">=</span> <span class="kc">None</span>
            
            <span class="c1"># Identify lines with relevant info.</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
                <span class="n">split_mod_line</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39; +&#39;</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>

                <span class="c1"># Construct object.</span>
                <span class="n">a</span> <span class="o">=</span> <span class="n">NhmmerHitAlig</span><span class="p">()</span>
                <span class="n">a</span><span class="o">.</span><span class="n">query_name</span> <span class="o">=</span> <span class="n">hmm_name</span>
                <span class="n">a</span><span class="o">.</span><span class="n">db_filepath</span> <span class="o">=</span> <span class="n">target_filepath</span>
                <span class="n">a</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">strand</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">score</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">evalue</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">hmmfrom</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">hmmto</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">alifrom</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">alito</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">envfrom</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
                <span class="n">a</span><span class="o">.</span><span class="n">envto</span> <span class="o">=</span> <span class="n">split_mod_line</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>

                <span class="c1"># Add object to list.</span>
                <span class="n">hit_alignments</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>

    <span class="c1"># Check that at least one hit was found.</span>
    <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">hit_alignments</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">,</span> <span class="s2">&quot;Error: No hits in file &quot;</span> <span class="o">+</span>\
 <span class="n">nhmmer_outfilepath</span>

    <span class="c1"># Return list of objects.</span>
    <span class="k">return</span> <span class="n">hit_alignments</span></div>


<span class="c1">#def parse_hmmer_out(hmmer_outfilepath):</span>
<span class="c1">#    &quot;&quot;&quot;Opens a hmmer output file (tabular output), extracts the information for location of</span>
<span class="c1">#    domains in the sequences, and returns a list of hit alignment objects.</span>
<span class="c1">#    &quot;&quot;&quot;</span>
<span class="c1">#    # Loop through the lines and identify the relevant info applying to all hits.</span>
<span class="c1">#    target_filepath = None</span>
<span class="c1">#    hmm_name = None</span>
<span class="c1">#    with open(hmmer_outfilepath) as i:</span>
<span class="c1">#        # Get info from end of file.</span>
<span class="c1">#        for line in i:</span>
<span class="c1">#            split_mod_line = re.split(r&#39; +&#39;, line)</span>
<span class="c1">#</span>
<span class="c1">#            if line.startswith(&#39;# Target file:&#39;):</span>
<span class="c1">#                target_filepath = line.replace(&#39; &#39;, &#39;&#39;).split(&#39;:&#39;)[1].strip()</span>
<span class="c1">#</span>
<span class="c1">#            if line.startswith(&#39;# Query file:&#39;):</span>
<span class="c1">#                hmm_name = line.replace(&#39; &#39;, &#39;&#39;).split(&#39;:&#39;)[1].strip()</span>
<span class="c1">#</span>
<span class="c1">#    # Loop through the lines again and make objects for each hit alignement.</span>
<span class="c1">#    hit_alignments = []</span>
<span class="c1">#    with open(hmmer_outfilepath) as i:</span>
<span class="c1">#        for line in i:</span>
<span class="c1">#            write = None</span>
<span class="c1">#            </span>
<span class="c1">#            # Identify lines with relevant info.</span>
<span class="c1">#            if not line.startswith(&#39;#&#39;):</span>
<span class="c1">#                split_mod_line = re.split(r&#39; +&#39;, line)</span>
<span class="c1">#</span>
<span class="c1">#                # Construct object.</span>
<span class="c1">#                a = HmmerHitAlig()</span>
<span class="c1">#                a.query_name = hmm_name a.db_filepath = target_filepath</span>
<span class="c1">#                a.subj_seq_id = split_mod_line[0]</span>
<span class="c1">#                a.original_strand = split_mod_line[24].strip(&#39;()&#39;)</span>
<span class="c1">#                a.score = split_mod_line[13]</span>
<span class="c1">#                a.evalue = split_mod_line[12]</span>
<span class="c1">#                a.hmmfrom = split_mod_line[15]</span>
<span class="c1">#                a.hmmto = split_mod_line[16]</span>
<span class="c1">#                a.alifrom = split_mod_line[17]</span>
<span class="c1">#                a.alito = split_mod_line[18]</span>
<span class="c1">#                a.envfrom = split_mod_line[19]</span>
<span class="c1">#                a.envto = split_mod_line[20]</span>
<span class="c1">#                a.subj_seq_position = split_mod_line[23]</span>
<span class="c1">#</span>
<span class="c1">#                #print(a.db_filepath)</span>
<span class="c1">#                #print(a.subj_seq_id)</span>
<span class="c1">#                #print(a.strand)</span>
<span class="c1">#                #print(a.score)</span>
<span class="c1">#                #print(a.evalue)</span>
<span class="c1">#                #print(a.hmmfrom)</span>
<span class="c1">#                #print(a.hmmto)</span>
<span class="c1">#                #print(a.alifrom)</span>
<span class="c1">#                #print(a.alito)</span>
<span class="c1">#                #print(a.envfrom)</span>
<span class="c1">#                #print(a.envto)</span>
<span class="c1">#                #print(a.pos_in_orig_db_seq)</span>
<span class="c1">#</span>
<span class="c1">#                # Add object to list.</span>
<span class="c1">#                hit_alignments.append(a)</span>
<span class="c1">#</span>
<span class="c1">#    # Check that at least one hit was found.</span>
<span class="c1">#    #assert len(hit_alignments) &gt;= 1, &quot;Error: No hits in file &quot; +\</span>
<span class="c1">#    #hmmer_outfilepath</span>
<span class="c1">#</span>
<span class="c1">#    # Return list of objects.</span>
<span class="c1">#    if len(hit_alignments) &lt; 1:</span>
<span class="c1">#        # Return an empty list if no alignments were in input file.</span>
<span class="c1">#        return []</span>
<span class="c1">#    else:</span>
<span class="c1">#        return hit_alignments</span>


<span class="c1"># Obsolete?</span>
<div class="viewcode-block" id="write_nhmmer_hit_fasta"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.write_nhmmer_hit_fasta">[docs]</a><span class="k">def</span> <span class="nf">write_nhmmer_hit_fasta</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">fasta_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list of nhmmer hit alignment objects, and writes corresponding</span>
<span class="sd">    fasta records to a fasta file.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_filepath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs</span><span class="p">:</span>
            <span class="n">fasta_header</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="o">+</span> <span class="s1">&#39; [&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">alifrom</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span>\
            <span class="n">i</span><span class="o">.</span><span class="n">alito</span> <span class="o">+</span> <span class="s1">&#39;] &#39;</span> <span class="o">+</span> <span class="s1">&#39;(&#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span> <span class="o">+</span> <span class="s1">&#39;) &#39;</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">db_name</span><span class="p">()</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>

            <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">fasta_header</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">sequ</span><span class="p">()</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="cluster_ali_recur"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.cluster_ali_recur">[docs]</a><span class="k">def</span> <span class="nf">cluster_ali_recur</span><span class="p">(</span><span class="n">sublists</span><span class="p">,</span> <span class="n">origlist</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Recursively cluster each object in the list with those that are within</span>
<span class="sd">    10,000 bp. Return a dict.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Problem: cannot assume that the whole gene will be no more than 10kb</span>
    <span class="c1"># long. The furthest exon may have a start position further than 10kb from</span>
    <span class="c1"># the start position of some other exons. This function is a little sloppy,</span>
    <span class="c1"># and may yeild inaccurate results in some cases.</span>
    <span class="c1">#***The assumption should be that individual introns should be no more than</span>
    <span class="c1">#~10 kb.</span>
    
    <span class="c1"># Assume that gene coding sequence does not exceed a certain value.</span>
    <span class="n">max_coding_sequence_len</span> <span class="o">=</span> <span class="mi">1000000</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">origlist</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sublists</span>

    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">origlist</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">sublists</span> <span class="o">+</span> <span class="p">[</span><span class="n">origlist</span><span class="p">]</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">sublist</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">new_orig_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">origlist</span><span class="p">:</span>
            <span class="c1"># If start positions are within 10000 bp, add to sublist.</span>
            <span class="c1"># (The first one will be included as well).</span>
            <span class="k">if</span> <span class="nb">abs</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">origlist</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alifrom</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span> <span class="o">-</span><span class="mi">1</span> <span class="o">&lt;=</span> <span class="n">max_coding_sequence_len</span><span class="p">:</span>
                <span class="n">sublist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">new_orig_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">sublist</span><span class="p">]</span> <span class="o">+</span> <span class="n">cluster_ali_recur</span><span class="p">(</span><span class="n">sublists</span><span class="p">,</span> <span class="n">new_orig_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_concat_seq"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_concat_seq">[docs]</a><span class="k">def</span> <span class="nf">get_concat_seq</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;nucl&#39;</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list of nhmmer hit alignment objects that should correspond to</span>
<span class="sd">    the same gene, and returns a concatenated sequence.</span>

<span class="sd">    May not be sufficient if the hit alignments do overlap.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Make sure the hits aren&#39;t overlapping.</span>
    <span class="k">assert</span> <span class="ow">not</span> <span class="n">check_if_hits_overlapping</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">),</span> <span class="s2">&quot;Error: Overlapping</span><span class="se">\</span>
<span class="s2"> nhmmer hits should not be concatenated.&quot;</span>

    <span class="c1"># Determine strand.</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
    <span class="n">db</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">db_name</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="c1"># Sort by start position.</span>
        <span class="n">nhmmer_hit_objs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="c1"># Sort by start position in reverse.</span>
        <span class="n">nhmmer_hit_objs</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alifrom</span><span class="p">),</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Assemble concatenated file.</span>
    <span class="n">header</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span>
    <span class="n">concat_seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs</span><span class="p">:</span>
        <span class="n">fromtolist</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">alifrom</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">alito</span><span class="p">]</span>
        <span class="n">header</span> <span class="o">=</span> <span class="n">header</span> <span class="o">+</span> <span class="s1">&#39;[&#39;</span> <span class="o">+</span> <span class="nb">min</span><span class="p">(</span><span class="n">fromtolist</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span> <span class="nb">max</span><span class="p">(</span><span class="n">fromtolist</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span> <span class="o">+</span> <span class="s1">&#39;,&#39;</span> 
        <span class="k">if</span> <span class="n">dtype</span> <span class="o">==</span> <span class="s1">&#39;nucl&#39;</span><span class="p">:</span>
            <span class="n">concat_seq</span> <span class="o">=</span> <span class="n">concat_seq</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">sequ</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">dtype</span> <span class="o">==</span> <span class="s1">&#39;prot&#39;</span><span class="p">:</span>
            <span class="n">concat_seq</span> <span class="o">=</span> <span class="n">concat_seq</span> <span class="o">+</span> <span class="n">i</span><span class="o">.</span><span class="n">sequprot</span><span class="p">()</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">header</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; (&#39;</span> <span class="o">+</span> <span class="n">strand</span> <span class="o">+</span> <span class="s1">&#39;)&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> 

    <span class="k">return</span> <span class="n">header</span> <span class="o">+</span> <span class="n">concat_seq</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> </div>


<div class="viewcode-block" id="get_gene_region"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_gene_region">[docs]</a><span class="k">def</span> <span class="nf">get_gene_region</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list of nhmmer hit alignment objects that should correspond to</span>
<span class="sd">    the same gene, and returns a concatenated sequence (of appropriate strand).</span>

<span class="sd">    So, if the gene region is on the negative strand, then it returns the</span>
<span class="sd">    sequence of the negative strand (5&#39; to 3&#39;) by getting the reverse</span>
<span class="sd">    complement of the forward strand.</span>

<span class="sd">    ***Does not account for paralogous genes possibly being very close to each</span>
<span class="sd">    other on the same strand of the same chromosome...</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Determine strand.</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strand</span>
    <span class="n">db</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">db_name</span><span class="p">()</span>

    <span class="c1"># Determine the inclusive range of this hits.</span>
    <span class="n">incl_range</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="c1"># Left to right.</span>
        <span class="n">incl_range</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span><span class="o">.</span><span class="n">alifrom</span><span class="p">),</span>
                <span class="nb">int</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alito</span><span class="p">))</span><span class="o">.</span><span class="n">alito</span><span class="p">))</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="c1"># Right to left.</span>
        <span class="n">incl_range</span> <span class="o">=</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alito</span><span class="p">))</span><span class="o">.</span><span class="n">alito</span><span class="p">),</span>
                <span class="nb">int</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span><span class="o">.</span><span class="n">alifrom</span><span class="p">))</span>

    <span class="c1"># Get full sequence object.</span>
    <span class="n">full_seq_obj</span> <span class="o">=</span> <span class="n">get_seq_obj_from_db_fasta</span><span class="p">([</span><span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">],</span>
            <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">db_filepath</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Define how much extra to include on either side of the inclusive range.</span>
    <span class="n">extra</span> <span class="o">=</span> <span class="mi">100</span>

    <span class="c1"># Define extended range (with extra on either side).</span>
    <span class="n">left_margin</span> <span class="o">=</span> <span class="n">incl_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span><span class="mi">1</span>
    <span class="n">left_extra</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">extra</span><span class="p">,</span> <span class="n">left_margin</span><span class="p">])</span>
    <span class="n">right_margin</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">full_seq_obj</span><span class="p">)</span> <span class="o">-</span> <span class="n">incl_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">right_extra</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">extra</span><span class="p">,</span> <span class="n">right_margin</span><span class="p">])</span>
    <span class="n">exten_range</span> <span class="o">=</span> <span class="p">(</span><span class="n">incl_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">left_extra</span><span class="p">,</span> <span class="n">incl_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">right_extra</span><span class="p">)</span>

    <span class="c1"># Assemble header string. </span>
    <span class="n">header</span> <span class="o">=</span> <span class="s1">&#39;&gt;&#39;</span> <span class="o">+</span> <span class="n">nhmmer_hit_objs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39; [&#39;</span> <span class="o">+</span>\
 <span class="nb">str</span><span class="p">(</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="s1">&#39;] (&#39;</span> <span class="o">+</span> <span class="n">strand</span> <span class="o">+</span> <span class="s1">&#39;)&#39;</span>

    <span class="c1"># Get sequence.</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">full_seq_obj</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">full_seq_obj</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span><span class="n">exten_range</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>

    <span class="c1"># Return a printable fasta record string.</span>
    <span class="c1">#fasta_string = header + &#39;\n&#39; + str(sequence) + &#39;\n&#39; </span>
    <span class="c1">#return fasta_string</span>
    
    <span class="c1"># Return a sequence object.</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">header</span><span class="p">,</span> <span class="n">sequence</span><span class="p">]</span> </div>


<div class="viewcode-block" id="display_dict"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.display_dict">[docs]</a><span class="k">def</span> <span class="nf">display_dict</span><span class="p">(</span><span class="n">nhmmer_hit_objs_dict2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Prints relevant info.&quot;&quot;&quot;</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key1</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">key1</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">key2</span><span class="p">)</span>
            <span class="n">obj_sublists</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span>
            <span class="n">num</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">obj_sublists</span><span class="p">:</span>
                <span class="n">num</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t</span><span class="s1">gene &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t\t\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">j</span><span class="o">.</span><span class="n">alifrom</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span> <span class="n">j</span><span class="o">.</span><span class="n">alito</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_three_frame_transl_string"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_three_frame_transl_string">[docs]</a><span class="k">def</span> <span class="nf">get_three_frame_transl_string</span><span class="p">(</span><span class="n">inputs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Takes a list comprising the header string and sequence string of an</span>
<span class="sd">    input sequence.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Needs to transfer coordinates from original while taking into account the</span>
    <span class="c1"># strand, etc.</span>
    <span class="n">header</span> <span class="o">=</span> <span class="n">inputs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="n">header</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;()&#39;</span><span class="p">)</span>
    <span class="n">position_in_original</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\[\d+\.\.\d+\]&#39;</span><span class="p">)</span>
    <span class="n">start_end</span> <span class="o">=</span> <span class="n">position_in_original</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="n">header</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;[]&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;..&#39;</span><span class="p">)</span>
    <span class="n">start_end</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">inputs</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">ambiguous_dna</span><span class="p">)</span>

    <span class="n">frame1</span> <span class="o">=</span> <span class="n">header</span> <span class="o">+</span> <span class="s1">&#39; frame 1&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span>

    <span class="n">x</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">y</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="n">frame2_start_end</span> <span class="o">=</span> <span class="s1">&#39;[&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start_end</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start_end</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">y</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span>
    <span class="n">frame2_header</span> <span class="o">=</span> <span class="n">position_in_original</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">frame2_start_end</span><span class="p">,</span> <span class="n">header</span><span class="p">)</span>
    <span class="n">frame2</span> <span class="o">=</span> <span class="n">frame2_header</span> <span class="o">+</span> <span class="s1">&#39; frame 2&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> 

    <span class="n">x</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">y</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
        <span class="n">x</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="k">elif</span> <span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
        <span class="n">y</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>

    <span class="n">frame3_start_end</span> <span class="o">=</span> <span class="s1">&#39;[&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start_end</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="n">x</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;..&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">start_end</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">y</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span>
    <span class="n">frame3_header</span> <span class="o">=</span> <span class="n">position_in_original</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="n">frame3_start_end</span><span class="p">,</span> <span class="n">header</span><span class="p">)</span>
    <span class="n">frame3</span> <span class="o">=</span> <span class="n">frame3_header</span> <span class="o">+</span> <span class="s1">&#39; frame 3&#39;</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span><span class="o">.</span><span class="n">translate</span><span class="p">())</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> 

    <span class="n">outstring</span> <span class="o">=</span> <span class="n">frame1</span> <span class="o">+</span> <span class="n">frame2</span> <span class="o">+</span> <span class="n">frame3</span>

    <span class="k">return</span> <span class="n">outstring</span></div>


<span class="c1"># Obsolete?</span>
<div class="viewcode-block" id="get_three_frame_transl"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_three_frame_transl">[docs]</a><span class="k">def</span> <span class="nf">get_three_frame_transl</span><span class="p">(</span><span class="n">seq_filepath</span><span class="p">,</span> <span class="n">three_frame_translation_dirpath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Produce three-frame translations of forward strand DNA sequences in a</span>
<span class="sd">    given file, and write to files in a given directory path.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">pass</span> </div>



<div class="viewcode-block" id="get_nhmmer_hit_gene_region_seqs"><a class="viewcode-back" href="../module_nhmmer_scaffolds.html#module_nhmmer_scaffolds.get_nhmmer_hit_gene_region_seqs">[docs]</a><span class="k">def</span> <span class="nf">get_nhmmer_hit_gene_region_seqs</span><span class="p">(</span><span class="n">nhmmer_hit_objs</span><span class="p">,</span> <span class="n">seq_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Take a list of nhmmer hit alignment objects and return one or more</span>
<span class="sd">    nucleotide subsequences that may correspond to the coding sequence for a</span>
<span class="sd">    single gene.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Cluster the objects into groups that may be from the same gene.</span>
    <span class="c1"># (on same strand, within a certain number of nucleotides of each other)</span>

    <span class="c1"># Split the list of objects into separate sub-lists for each fasta record</span>
    <span class="c1"># for which there are hits in the subject database.</span>
    <span class="n">nhmmer_hit_objs_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">nhmmer_hit_objs_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_ids</span><span class="p">:</span>
            <span class="n">nhmmer_hit_objs_dict</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs_dict</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">nhmmer_hit_objs_dict</span><span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="n">nhmmer_hit_objs_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">.</span><span class="n">subj_seq_id</span><span class="p">)</span>

    <span class="c1"># Split the list of object into those that are on the positive strand, and</span>
    <span class="c1"># those that are on the negative strand.    </span>
    <span class="n">nhmmer_hit_objs_dict2</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict</span><span class="p">:</span>
        <span class="c1"># Identify objects with alignments on either positive or negative strand.</span>
        <span class="n">obj_list</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
        <span class="n">pos</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">neg</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">obj_list</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span> <span class="ow">or</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s2">&quot;Error: Strand not specified.&quot;</span>
            <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">:</span>
                <span class="n">pos</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">neg</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

        <span class="c1"># Make a dict with sub-lists.</span>
        <span class="n">obj_dict</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">obj_dict</span><span class="p">[</span><span class="s1">&#39;positive strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pos</span>
        <span class="n">obj_dict</span><span class="p">[</span><span class="s1">&#39;negative strand&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">neg</span>

        <span class="c1"># Replace original list with dict.</span>
        <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">obj_dict</span>

    <span class="c1"># Further divide the objects based on which are likely to belong to the</span>
    <span class="c1"># same gene.</span>
    <span class="k">for</span> <span class="n">key1</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="k">for</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">obj_list</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span>
            <span class="n">obj_sublists</span> <span class="o">=</span> <span class="n">cluster_ali_recur</span><span class="p">([],</span> <span class="n">obj_list</span><span class="p">)</span>
            <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span> <span class="o">=</span> <span class="n">obj_sublists</span>

    <span class="c1">## Display contents of dictionary structure.</span>
    <span class="n">display_dict</span><span class="p">(</span><span class="n">nhmmer_hit_objs_dict2</span><span class="p">)</span>   

    <span class="c1"># Iterate over all the lists of objects corresponding to potential</span>
    <span class="c1"># individual genes, and write consensus sequences to an output fasta file.</span>
    <span class="c1"># This step seems to take a really long time for some reason, longer than</span>
    <span class="c1"># running the HMMer search in the first place...</span>
    <span class="c1"># (Would be nice to sort the output sequences according to e-value somehow)</span>
    <span class="n">gene_num</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">seq_filepath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">o</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">key1</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">key2</span> <span class="ow">in</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">obj_lists</span> <span class="o">=</span> <span class="n">nhmmer_hit_objs_dict2</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="n">key2</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">obj_lists</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="c1"># Loop over lists of objects corresponding to exons for</span>
                    <span class="c1"># single genes.</span>
                    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">obj_lists</span><span class="p">:</span>
                        <span class="n">gene_num</span> <span class="o">+=</span> <span class="mi">1</span>

                        <span class="c1"># Get concatenated sequence for identified gene.</span>
                        <span class="n">c</span> <span class="o">=</span> <span class="n">get_concat_seq</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="s1">&#39;nucl&#39;</span><span class="p">)</span>
                        <span class="n">o</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">c</span><span class="p">)</span></div>

                        <span class="c1"># Get gene region.</span>
                        <span class="c1">#s = get_gene_region(l)</span>


                        <span class="c1"># Sort objects in hmmer alignment list by start</span>
                        <span class="c1"># alignment.</span>
                        



                        <span class="c1"># Generate temporary 3-frame translation of gene</span>
                        <span class="c1"># region.</span>
                        <span class="c1">#temp_filepath = seq_filepath + &#39;_temp_transl&#39; + str(gene_num) + &#39;.fa&#39;</span>
                        <span class="c1">#with open(temp_filepath, &#39;w&#39;) as tempoutfh:</span>
                        <span class="c1">#    tempoutfh.write(get_three_frame_transl_string(s))</span>

                        <span class="c1"># Run hmmer search on 3-frame translations.</span>
                        <span class="c1">#tab_hmmsearch_outpath = temp_filepath + &#39;_hmmsearchout&#39; + str(gene_num) + &#39;.txt&#39;</span>
                        <span class="c1">#run_hmmsearch(prothmmpath, temp_filepath, tab_hmmsearch_outpath)                           </span>

                        <span class="c1"># Parse the hmmsearch output, and extract the relevant</span>
                        <span class="c1"># gene models.</span>
                        <span class="c1"># ...</span>
                        <span class="c1">#hmmer_objs = parse_hmmer_out(tab_hmmsearch_outpath)</span>

                        <span class="c1"># Remove temporary file with 3-frame translations.</span>
                        <span class="c1">#os.remove(temp_filepath)</span>

                        <span class="c1"># Remove temporary hmmsearch output file.</span>
                        <span class="c1">#os.remove(tab_hmmsearch_outpath)</span>


                        <span class="c1"># get three-frame translations of gene region.</span>
                        <span class="c1">#o.write(get_three_frame_transl_string(s))</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">amoebae 0.0 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Lael D. Barlow.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.5.
    </div>
  </body>
</html>